// rxjs-etc/source/placeholders-spec.ts
/**
 * @license Use of this source code is governed by an MIT-style license that
 * can be found in the LICENSE file at https://github.com/cartant/rxjs-etc
 */

import { Observable } from "rxjs";
import * as p from "ts-snippet/placeholders";

export * from "ts-snippet/placeholders";

export type OT0 = Observable<p.T0>;
export type OT1 = Observable<p.T1>;
export type OT2 = Observable<p.T2>;
export type OT3 = Observable<p.T3>;
export type OT4 = Observable<p.T4>;
export type OT5 = Observable<p.T5>;
export type OT6 = Observable<p.T6>;
export type OT7 = Observable<p.T7>;
export type OT8 = Observable<p.T8>;
export type OT9 = Observable<p.T9>;
export type OT10 = Observable<p.T10>;
export type OT11 = Observable<p.T11>;
export type OT12 = Observable<p.T12>;

export declare const o0: OT0;
export declare const o1: OT1;
export declare const o2: OT2;
export declare const o3: OT3;
export declare const o4: OT4;
export declare const o5: OT5;
export declare const o6: OT6;
export declare const o7: OT7;
export declare const o8: OT8;
export declare const o9: OT9;
export declare const o10: OT10;
export declare const o11: OT11;
export declare const o12: OT12;
