// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/item/ItemRow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ItemRow = void 0;

var _icons = require("@sanity/icons");

var _presence = require("@sanity/base/presence");

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _components = require("@sanity/base/components");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Preview = _interopRequireDefault(require("../../../../Preview"));

var _ConfirmDeleteButton = require("../ConfirmDeleteButton");

var _DragHandle = require("../../common/DragHandle");

var _ItemWithMissingType = require("./ItemWithMissingType");

var _excluded = ["isSortable", "value", "onClick", "onKeyPress", "onFocus", "type", "readOnly", "presence", "onRemove", "validation"];

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  paddingX: 2,
  paddingY: 3
});

var Root = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  transition: 250ms border-color;\n  &[aria-selected='true'] {\n    border-color: var(--card-focus-ring-color);\n  }\n"])));

var ItemRow = /*#__PURE__*/_react.default.forwardRef(function RegularItem(props, ref) {
  var focusRef = _react.default.useRef();

  var isSortable = props.isSortable,
      value = props.value,
      onClick = props.onClick,
      onKeyPress = props.onKeyPress,
      onFocus = props.onFocus,
      type = props.type,
      readOnly = props.readOnly,
      presence = props.presence,
      onRemove = props.onRemove,
      validation = props.validation,
      rest = _objectWithoutProperties(props, _excluded);

  return /*#__PURE__*/_react.default.createElement(Root, _extends({}, rest, {
    border: true,
    radius: 1,
    padding: 1,
    ref: ref
  }), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, isSortable && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginRight: 1
  }, dragHandle), type ? /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    type: "button",
    radius: 2,
    flex: 1,
    onClick: onClick,
    ref: focusRef,
    onKeyPress: onKeyPress,
    onFocus: onFocus
  }, /*#__PURE__*/_react.default.createElement(_Preview.default, {
    layout: "default",
    value: value,
    type: type
  })) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ItemWithMissingType.ItemWithMissingType, {
    value: value,
    onFocus: onFocus
  })), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, !readOnly && presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_components.FormFieldValidationStatus, {
    __unstable_markers: validation,
    __unstable_showSummary: !value._ref
  })), value._ref && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_components.IntentButton, {
    icon: _icons.LinkIcon,
    intent: "edit",
    mode: "bleed",
    params: {
      id: value._ref
    }
  })), !readOnly && onRemove && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ConfirmDeleteButton.ConfirmDeleteButton, {
    placement: "left",
    title: "Remove item",
    onConfirm: onRemove
  })), !value._key && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "This item is missing the required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property.")),
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution"
  }, "Missing key"))))));
});

exports.ItemRow = ItemRow;
