// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/item/ItemWithMissingType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ItemWithMissingType = ItemWithMissingType;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _content = require("@sanity/util/content");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ItemWithMissingType(props) {
  var value = props.value,
      onFocus = props.onFocus;

  var _React$useState = _react.default.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      showDetails = _React$useState2[0],
      setShowDetails = _React$useState2[1];

  var _React$useState3 = _react.default.useState(null),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      popoverRef = _React$useState4[0],
      setPopoverRef = _React$useState4[1];

  (0, _ui.useClickOutside)(() => setShowDetails(false), [popoverRef]);

  var handleKeyDown = _react.default.useCallback(e => {
    if (e.key === 'Escape' || e.key === 'Tab') {
      setShowDetails(false);
    }
  }, []);

  var typeName = (0, _content.resolveTypeName)(value);
  return /*#__PURE__*/_react.default.createElement(_ui.Popover, {
    open: showDetails,
    ref: setPopoverRef,
    onKeyDown: handleKeyDown,
    content: /*#__PURE__*/_react.default.createElement(_ui.Card, {
      margin: 1,
      padding: 1,
      onKeyDown: handleKeyDown,
      tabIndex: 0,
      style: {
        outline: 'none'
      }
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement("b", null, "Why is this happening?")), "The current schema does not declare items of type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " as valid for this list. This could mean that the type has been removed as a valid item type, or that someone else has added it to their own local schema that is not yet deployed.", /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_icons.BulbOutlineIcon, null), " You can still move or delete this item, but it cannot be edited since the schema definition for its type is nowhere to be found.")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, "JSON representation of this item:"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      overflow: "auto",
      border: true
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement("pre", null, JSON.stringify(value, null, 2)))))
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    type: "button",
    radius: 2,
    padding: 2,
    onFocus: onFocus,
    onClick: () => setShowDetails(true),
    onKeyDown: handleKeyDown
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 2,
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, "Item type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " not defined for this list")))));
}
