// @sanity/base/lib/presence/overlay/RegionsWithIntersections.styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BottomRegionWrapper = exports.MiddleRegionWrapper = exports.TopRegionWrapper = exports.RegionWrapper = exports.OverlayWrapper = exports.RootWrapper = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _WithIntersection = require("./WithIntersection");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var RootWrapper = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: relative;\n"])));

exports.RootWrapper = RootWrapper;

var OverlayWrapper = _styledComponents.default.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  pointer-events: none;\n  z-index: 13;\n"])));

exports.OverlayWrapper = OverlayWrapper;
var RegionWrapper = (0, _styledComponents.css)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  overflow: hidden;\n  pointer-events: none;\n  position: absolute;\n"])));
exports.RegionWrapper = RegionWrapper;
var TopRegionWrapper = (0, _styledComponents.default)(_WithIntersection.WithIntersection)(_ref => {
  var debug = _ref.debug,
      margins = _ref.margins;
  return (0, _styledComponents.css)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n    ", "\n\n    z-index: 100;\n    position: sticky;\n    height: 1px;\n    top: ", "px;\n    background-color: ", ";\n  "])), RegionWrapper, margins[0] - 1, debug ? 'red' : 'none');
});
exports.TopRegionWrapper = TopRegionWrapper;
var MiddleRegionWrapper = (0, _styledComponents.default)(_WithIntersection.WithIntersection)(_ref2 => {
  var debug = _ref2.debug;
  return (0, _styledComponents.css)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n    ", "\n\n    visibility: none;\n\n    ", "\n  "])), RegionWrapper, debug && (0, _styledComponents.css)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n      background: rgba(255, 0, 0, 0.25);\n      outline: 1px solid #00b;\n      visibility: visible;\n    "]))));
});
exports.MiddleRegionWrapper = MiddleRegionWrapper;
var BottomRegionWrapper = (0, _styledComponents.default)(_WithIntersection.WithIntersection)(_ref3 => {
  var debug = _ref3.debug;
  return (0, _styledComponents.css)(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n    ", "\n\n    position: sticky;\n    bottom: -1px;\n    height: 1px;\n    background-color: ", ";\n  "])), RegionWrapper, debug ? 'blue' : 'transparent');
});
exports.BottomRegionWrapper = BottomRegionWrapper;
