// @sanity/desk-tool/lib/components/ReferringDocumentsList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReferringDocumentsList = ReferringDocumentsList;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _router = require("@sanity/base/router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ReferringDocumentsList(props) {
  var documents = props.documents;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 1,
    radius: 3,
    shadow: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 1
  }, documents.map(document => /*#__PURE__*/_react.default.createElement(DocumentPreviewLink, {
    document: document,
    key: document._id
  }))));
}

function DocumentPreviewLink(props) {
  var document = props.document;
  var router = (0, _router.useRouter)();
  var intent = (0, _react.useMemo)(() => ({
    action: 'edit',
    params: {
      id: document._id,
      type: document._type
    }
  }), [document]);
  var href = router.resolveIntentLink(intent.action, intent.params);

  var schemaType = _schema.default.get(document._type);

  var handleClick = (0, _react.useCallback)(event => {
    if (event.shiftKey || event.metaKey) return;
    event.preventDefault();
    router.navigateIntent(intent.action, intent.params);
  }, [intent, router]);

  if (!schemaType) {
    return /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, "A document of the unknown type ", /*#__PURE__*/_react.default.createElement("em", null, document._type));
  }

  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "a",
    href: href,
    onClick: handleClick,
    padding: 1,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_preview.default, {
    layout: "default",
    type: schemaType,
    value: document
  })), document._hasDraft && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true
  }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)))));
}
