// @sanity/desk-tool/lib/panes/unknownPane/UnknownPaneType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default = _interopRequireDefault(require("part:@sanity/components/panes/default"));

var _UnknownPaneType = _interopRequireDefault(require("./UnknownPaneType.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class UnknownPaneType extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        isSelected = _this$props.isSelected,
        isCollapsed = _this$props.isCollapsed,
        onCollapse = _this$props.onCollapse,
        onExpand = _this$props.onExpand,
        type = _this$props.type;
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      title: "Unknown pane type",
      index: this.props.index,
      isSelected: isSelected,
      isCollapsed: isCollapsed,
      onCollapse: onCollapse,
      onExpand: onExpand
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _UnknownPaneType.default.root
    }, /*#__PURE__*/_react.default.createElement("p", null, type ? /*#__PURE__*/_react.default.createElement("span", null, "Structure item of type ", /*#__PURE__*/_react.default.createElement("code", null, type), " is not a known entity.") : /*#__PURE__*/_react.default.createElement("span", null, "Structure item is missing required ", /*#__PURE__*/_react.default.createElement("code", null, "type"), " property."))));
  }

}

exports.default = UnknownPaneType;

_defineProperty(UnknownPaneType, "propTypes", {
  type: _propTypes.default.string,
  isSelected: _propTypes.default.bool.isRequired,
  isCollapsed: _propTypes.default.bool.isRequired,
  onExpand: _propTypes.default.func,
  onCollapse: _propTypes.default.func,
  index: _propTypes.default.number
});

_defineProperty(UnknownPaneType, "defaultProps", {
  type: undefined,
  onExpand: undefined,
  onCollapse: undefined
});
