// @sanity/ui/lib/cjs/primitives/textArea/textArea.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  TextArea: () => TextArea
});
var import_react = __toModule(require("react"));
var import_styled_components = __toModule(require("styled-components"));
var import_hooks = __toModule(require("../../hooks"));
var import_internal = __toModule(require("../../styles/internal"));
const Root = import_styled_components.default.span(import_internal.textInputRootStyle);
const InputRoot = import_styled_components.default.span`
  flex: 1;
  min-width: 0;
  display: block;
  position: relative;
`;
const Input = import_styled_components.default.textarea(import_internal.responsiveInputPaddingStyle, import_internal.textInputBaseStyle, import_internal.textInputFontSizeStyle);
const Presentation = import_styled_components.default.div(import_internal.responsiveRadiusStyle, import_internal.textInputRepresentationStyle);
const TextArea = (0, import_react.forwardRef)(function TextArea2(props, forwardedRef) {
  const _a = props, {
    border = true,
    customValidity,
    disabled = false,
    fontSize = 2,
    padding = 3,
    radius = 1
  } = _a, restProps = __objRest(_a, [
    "border",
    "customValidity",
    "disabled",
    "fontSize",
    "padding",
    "radius"
  ]);
  const ref = (0, import_hooks.useForwardedRef)(forwardedRef);
  (0, import_hooks.useCustomValidity)(ref, customValidity);
  return /* @__PURE__ */ import_react.default.createElement(Root, {
    "data-ui": "TextArea"
  }, /* @__PURE__ */ import_react.default.createElement(InputRoot, null, /* @__PURE__ */ import_react.default.createElement(Input, __spreadProps(__spreadValues({
    "data-as": "textarea"
  }, restProps), {
    $fontSize: fontSize,
    $padding: padding,
    disabled,
    ref
  })), /* @__PURE__ */ import_react.default.createElement(Presentation, {
    $border: border,
    $radius: radius
  })));
});
