// @sanity/default-login/lib/legacyParts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BrandLogo", {
  enumerable: true,
  get: function get() {
    return _brandLogo.default;
  }
});
Object.defineProperty(exports, "authenticationFetcher", {
  enumerable: true,
  get: function get() {
    return _authenticationFetcher.default;
  }
});
Object.defineProperty(exports, "pluginConfig", {
  enumerable: true,
  get: function get() {
    return _defaultLogin.default;
  }
});
Object.defineProperty(exports, "LoginDialogContent", {
  enumerable: true,
  get: function get() {
    return _loginDialogContent.default;
  }
});
Object.defineProperty(exports, "userStore", {
  enumerable: true,
  get: function get() {
    return _user.default;
  }
});
Object.defineProperty(exports, "LoginDialog", {
  enumerable: true,
  get: function get() {
    return _loginDialog.default;
  }
});

var _brandLogo = _interopRequireDefault(require("part:@sanity/base/brand-logo?"));

var _authenticationFetcher = _interopRequireDefault(require("part:@sanity/base/authentication-fetcher"));

var _defaultLogin = _interopRequireDefault(require("config:@sanity/default-login"));

var _loginDialogContent = _interopRequireDefault(require("part:@sanity/base/login-dialog-content"));

var _user = _interopRequireDefault(require("part:@sanity/base/user"));

var _loginDialog = _interopRequireDefault(require("part:@sanity/base/login-dialog"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
