// @sanity/default-login/lib/LoginDialogContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _configSanity = _interopRequireDefault(require("config:sanity"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _legacyParts = require("./legacyParts");

var _getProviderLogo = require("./util/getProviderLogo");

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var ProviderLogoWrapper = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  svg,\n  img {\n    border-radius: 50%;\n    height: 1.25em;\n    width: 1.25em;\n  }\n"])));
var BrandLogoWrapper = (0, _styledComponents.default)(_ui.Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  svg {\n    display: block;\n    height: 2rem;\n    width: auto;\n    margin: 0 auto;\n  }\n"])));
var projectName = _configSanity.default.project && _configSanity.default.project.name || ''; // eslint-disable-next-line react/require-optimization

class LoginDialogContent extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleLoginButtonClicked", (provider, event) => {
      var onLoginButtonClick = this.props.onLoginButtonClick;

      if (onLoginButtonClick) {
        this.props.onLoginButtonClick(provider, event);
      } else {
        console.warn('LoginDialogContent is missing the onLoginButtonClick property'); // eslint-disable-line no-console
      }
    });
  }

  render() {
    var _this$props = this.props,
        title = _this$props.title,
        description = _this$props.description,
        providers = _this$props.providers,
        SanityLogo = _this$props.SanityLogo;
    return /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 4,
      paddingY: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, _legacyParts.BrandLogo && projectName && /*#__PURE__*/_react.default.createElement(BrandLogoWrapper, null, /*#__PURE__*/_react.default.createElement(_legacyParts.BrandLogo, {
      projectName: projectName
    })), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, !_legacyParts.BrandLogo && projectName && /*#__PURE__*/_react.default.createElement(_ui.Heading, {
      align: "center",
      as: "h2",
      size: 3
    }, projectName), title && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center",
      weight: "semibold"
    }, title), description && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      muted: true,
      align: "center"
    }, description)), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2,
      as: "ul"
    }, providers === null || providers === void 0 ? void 0 : providers.map(provider => {
      var ProviderLogo = (0, _getProviderLogo.getProviderLogo)(provider);
      var onLoginClick = this.handleLoginButtonClicked.bind(this, provider);
      return /*#__PURE__*/_react.default.createElement(_ui.Card, {
        key: provider === null || provider === void 0 ? void 0 : provider.name,
        radius: 2,
        border: true,
        as: "li"
      }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
        mode: "bleed",
        paddingY: 4,
        onClick: onLoginClick,
        style: {
          width: '100%'
        }
      }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
        justify: "center"
      }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
        space: 2
      }, /*#__PURE__*/_react.default.createElement(ProviderLogoWrapper, null, /*#__PURE__*/_react.default.createElement(ProviderLogo, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, null, provider === null || provider === void 0 ? void 0 : provider.title))))));
    })), SanityLogo && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      justify: "center",
      marginTop: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(SanityLogo, null)))));
  }

}

exports.default = LoginDialogContent;

_defineProperty(LoginDialogContent, "propTypes", {
  title: _propTypes.default.node.isRequired,
  description: _propTypes.default.node,
  // eslint-disable-next-line react/forbid-prop-types
  providers: _propTypes.default.array,
  onLoginButtonClick: _propTypes.default.func,
  SanityLogo: _propTypes.default.func
});

_defineProperty(LoginDialogContent, "defaultProps", {
  description: null,
  providers: null,
  onLoginButtonClick: null,
  SanityLogo: null
});
