// @sanity/base/lib/__legacy/@sanity/components/avatar/Avatar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = Avatar;

var _autoId = require("@reach/auto-id");

var _react = _interopRequireWildcard(require("react"));

var _Avatar = _interopRequireDefault(require("./Avatar.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var SVG_SIZE = 23;
var SVG_RADIUS = SVG_SIZE / 2;

function Avatar(props) {
  var _props$color = props.color,
      color = _props$color === void 0 ? {
    dark: 'currentColor',
    light: 'currentColor'
  } : _props$color,
      src = props.src,
      title = props.title,
      initials = props.initials,
      onImageLoadError = props.onImageLoadError,
      arrowPositionProp = props.arrowPosition,
      animateArrowFrom = props.animateArrowFrom,
      _props$status = props.status,
      status = _props$status === void 0 ? 'online' : _props$status,
      _props$size = props.size,
      size = _props$size === void 0 ? 'small' : _props$size,
      tone = props.tone;
  var backgroundColor = tone === 'navbar' ? color.dark : color.light;
  var elementId = (0, _autoId.useId)();

  var _useState = (0, _react.useState)(animateArrowFrom || arrowPositionProp || 'inside'),
      _useState2 = _slicedToArray(_useState, 2),
      arrowPosition = _useState2[0],
      setArrowPosition = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      imageFailed = _useState4[0],
      setImageFailed = _useState4[1];

  (0, _react.useEffect)(() => {
    if (arrowPosition === arrowPositionProp) return undefined; // Start animation in the next frame

    var raf = requestAnimationFrame(() => setArrowPosition(arrowPositionProp));
    return () => cancelAnimationFrame(raf);
  }, [arrowPosition, arrowPositionProp]);
  (0, _react.useEffect)(() => {
    if (src) setImageFailed(false);
  }, [src]);
  var handleImageError = (0, _react.useCallback)(() => {
    setImageFailed(true);

    if (onImageLoadError) {
      onImageLoadError(new Error('Avatar: the image failed to load'));
    }
  }, [onImageLoadError]);
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-label": title,
    className: _Avatar.default.root,
    "data-arrow-position": arrowPosition,
    "data-size": size,
    "data-status": status,
    "data-tone": tone,
    style: {
      backgroundColor
    },
    title: title
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.arrow
  }, /*#__PURE__*/_react.default.createElement("svg", {
    width: "11",
    height: "7",
    viewBox: "0 0 11 7",
    fill: "none"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "M6.67948 1.50115L11 7L0 7L4.32052 1.50115C4.92109 0.736796 6.07891 0.736795 6.67948 1.50115Z",
    fill: backgroundColor
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.inner
  }, !imageFailed && src && /*#__PURE__*/_react.default.createElement("svg", {
    viewBox: "0 0 ".concat(SVG_SIZE, " ").concat(SVG_SIZE),
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement("pattern", {
    id: "".concat(elementId, "-image-url"),
    patternContentUnits: "objectBoundingBox",
    width: "1",
    height: "1"
  }, /*#__PURE__*/_react.default.createElement("image", {
    href: src,
    width: "1",
    height: "1",
    onError: handleImageError
  }))), /*#__PURE__*/_react.default.createElement("circle", {
    cx: SVG_RADIUS,
    cy: SVG_RADIUS,
    r: SVG_RADIUS,
    fill: "url(#".concat(elementId, "-image-url)")
  }), /*#__PURE__*/_react.default.createElement("ellipse", {
    className: _Avatar.default.bgStroke,
    cx: SVG_RADIUS,
    cy: SVG_RADIUS,
    rx: SVG_RADIUS,
    ry: SVG_RADIUS
  }), /*#__PURE__*/_react.default.createElement("ellipse", {
    className: _Avatar.default.stroke,
    cx: SVG_RADIUS,
    cy: SVG_RADIUS,
    rx: SVG_RADIUS,
    ry: SVG_RADIUS,
    stroke: backgroundColor
  })), (imageFailed || !src) && initials && /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.initials
  }, /*#__PURE__*/_react.default.createElement("span", null, initials))));
}
