// @sanity/base/lib/components/zOffsets/legacyZIndexes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLegacyZIndexes = getLegacyZIndexes;
exports.getCustomCSSPropertyNumber = getCustomCSSPropertyNumber;

var _sanityCssCustomProperties = _interopRequireDefault(require("sanity:css-custom-properties"));

var _defaults = require("./defaults");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />

/**
 * @internal
 */
function getLegacyZIndexes() {
  return {
    navbar: getCustomCSSPropertyNumber('--zindex-navbar') || _defaults.defaults.navbar,
    navbarPopover: getCustomCSSPropertyNumber('--zindex-navbar-popover') || _defaults.defaults.navbarPopover,
    navbarDialog: getCustomCSSPropertyNumber('--zindex-navbar-dialog') || _defaults.defaults.navbarDialog,
    pane: getCustomCSSPropertyNumber('--zindex-pane') || _defaults.defaults.pane,
    paneHeader: _defaults.defaults.paneHeader,
    paneFooter: _defaults.defaults.paneFooter,
    paneResizer: getCustomCSSPropertyNumber('--zindex-pane-resizer') || _defaults.defaults.paneResizer,
    portal: getCustomCSSPropertyNumber('--zindex-portal') || _defaults.defaults.portal,
    popover: getCustomCSSPropertyNumber('--zindex-popover') || _defaults.defaults.popover,
    modal: getCustomCSSPropertyNumber('--zindex-modal') || _defaults.defaults.modal,
    movingItem: getCustomCSSPropertyNumber('--zindex-moving-item') || _defaults.defaults.movingItem,
    drawershade: getCustomCSSPropertyNumber('--zindex-drawershade') || _defaults.defaults.drawershade,
    drawer: getCustomCSSPropertyNumber('--zindex-drawer') || _defaults.defaults.drawer,
    fullscreen: _defaults.defaults.fullscreen,
    toast: _defaults.defaults.toast,
    // THESE ARE NOT IN USE:
    dropdown: getCustomCSSPropertyNumber('--zindex-dropdown') || _defaults.defaults.dropdown,
    navbarFixed: getCustomCSSPropertyNumber('--zindex-navbar-fixed') || _defaults.defaults.navbarFixed,
    fullscreenEdit: getCustomCSSPropertyNumber('--zindex-fullscreen-edit') || _defaults.defaults.fullscreenEdit,
    popoverBackground: getCustomCSSPropertyNumber('--zindex-popover-background') || _defaults.defaults.popoverBackground,
    tooltip: getCustomCSSPropertyNumber('--zindex-tooltip') || _defaults.defaults.tooltip,
    modalBackground: getCustomCSSPropertyNumber('--zindex-modal-background') || _defaults.defaults.modalBackground,
    spinner: getCustomCSSPropertyNumber('--zindex-spinner') || _defaults.defaults.spinner
  };
}
/**
 * @internal
 */


function getCustomCSSPropertyNumber(key) {
  var rawValue = _sanityCssCustomProperties.default[key];

  if (rawValue === undefined) {
    // eslint-disable-next-line no-console
    console.warn("getCustomCSSPropertyNumber: the custom CSS property \"".concat(key, "\" is not defined"));
    return undefined;
  }

  var value = parseInt(rawValue, 10);

  if (isNaN(value)) {
    // eslint-disable-next-line no-console
    console.warn("getCustomCSSPropertyNumber: the custom CSS property \"".concat(key, "\" is not a number"));
    return undefined;
  }

  return value;
}
