// @sanity/form-builder/lib/inputs/InvalidValueInput/UntypedValueInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UntypedValueInput = UntypedValueInput;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _Alert = require("../../components/Alert");

var _Details = require("../../components/Details");

var _legacyParts = require("../../legacyParts");

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function SetMissingTypeButton(_ref) {
  var value = _ref.value,
      targetType = _ref.targetType,
      onChange = _ref.onChange;
  var itemValue = (0, _react.useMemo)(() => _objectSpread(_objectSpread({}, value), {}, {
    _type: targetType
  }), [targetType, value]);
  var handleClick = (0, _react.useCallback)(() => onChange(_PatchEvent.default.from((0, _PatchEvent.setIfMissing)(targetType, ['_type'])), itemValue), [itemValue, onChange, targetType]);
  return /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleClick,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Convert to ", /*#__PURE__*/_react.default.createElement("code", null, targetType))
  });
}

function UnsetItemButton(_ref2) {
  var value = _ref2.value,
      onChange = _ref2.onChange,
      validTypes = _ref2.validTypes;
  // Doesn't matter which `_type` we use as long as it's allowed by the array
  var itemValue = (0, _react.useMemo)(() => _objectSpread(_objectSpread({}, value), {}, {
    _type: validTypes[0]
  }), [validTypes, value]);
  var handleClick = (0, _react.useCallback)(() => onChange(_PatchEvent.default.from((0, _PatchEvent.unset)()), itemValue), [itemValue, onChange]);
  return /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleClick,
    tone: "critical",
    text: "Unset value"
  });
}
/**
 * When the value does not have an `_type` property,
 * but the schema has a named type
 */


function UntypedValueInput(_ref3) {
  var validTypes = _ref3.validTypes,
      value = _ref3.value,
      onChange = _ref3.onChange;
  var isSingleValidType = validTypes.length === 1;

  var isHoistedType = _legacyParts.schema.has(validTypes[0]);

  return /*#__PURE__*/_react.default.createElement(_Alert.Alert, {
    status: "warning",
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Property value missing ", /*#__PURE__*/_react.default.createElement("code", null, "_type"))
  }, /*#__PURE__*/_react.default.createElement(_Details.Details, {
    open: __DEV__,
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Developer info")
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Encountered an object value without a ", /*#__PURE__*/_react.default.createElement("code", null, "_type"), " property."), isSingleValidType && !isHoistedType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Either remove the ", /*#__PURE__*/_react.default.createElement("code", null, "name"), " property of the object declaration, or set", ' ', /*#__PURE__*/_react.default.createElement("code", null, "_type"), " property on items."), !isSingleValidType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "The following types are valid here according to schema:"), !isSingleValidType && /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    as: "ul",
    space: 2
  }, validTypes.map(validType => /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "li",
    key: validType,
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("code", null, validType)))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "h4",
    weight: "semibold",
    size: 1
  }, "Current value (", /*#__PURE__*/_react.default.createElement("code", null, "object"), "):"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    border: true,
    overflow: "auto",
    padding: 2,
    radius: 2,
    tone: "inherit"
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json"
  }, JSON.stringify(value, null, 2)))), /*#__PURE__*/_react.default.createElement(_ui.Grid, {
    columns: [1, 2, 2],
    gap: 1
  }, isSingleValidType && /*#__PURE__*/_react.default.createElement(SetMissingTypeButton, {
    onChange: onChange,
    targetType: validTypes[0],
    value: value
  }), /*#__PURE__*/_react.default.createElement(UnsetItemButton, {
    onChange: onChange,
    validTypes: validTypes,
    value: value
  })))));
}
