// @sanity/ui/lib/cjs/primitives/_selectable/style.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  selectableBaseStyle: () => selectableBaseStyle,
  selectableColorStyle: () => selectableColorStyle
});
var import_styled_components = __toModule(require("styled-components"));
var import_colorVars = __toModule(require("../../styles/colorVars"));
function selectableBaseStyle() {
  return import_styled_components.css`
    background-color: inherit;
    color: inherit;

    &[data-as='button'] {
      -webkit-font-smoothing: inherit;
      appearance: none;
      outline: none;
      font: inherit;
      text-align: inherit;
      border: 0;
      width: stretch;
    }

    /* &:is(a) */
    &[data-as='a'] {
      text-decoration: none;
    }
  `;
}
function selectableColorStyle(props) {
  var _a, _b;
  const { $tone, theme } = props;
  const { base, muted, selectable } = theme.sanity.color;
  const tone = selectable ? selectable[$tone] || selectable.default : muted[$tone] || muted.default;
  return import_styled_components.css`
    ${(0, import_colorVars._colorVarsStyle)(base, tone.enabled)}

    background-color: var(--card-bg-color);
    color: var(--card-fg-color);
    outline: none;

    /* &:is(button) */
    &[data-as='button'] {
      &:disabled {
        ${(0, import_colorVars._colorVarsStyle)(base, tone.disabled)}
      }

      &:not(:disabled) {
        &[aria-pressed='true'] {
          ${(0, import_colorVars._colorVarsStyle)(base, tone.pressed)}
        }

        &[data-selected],
        &[aria-selected='true'] > & {
          ${(0, import_colorVars._colorVarsStyle)(base, tone.selected)}
        }

        @media (hover: hover) {
          &:not([data-selected]) {
            &:hover {
              ${(0, import_colorVars._colorVarsStyle)(base, tone.hovered)}
            }

            &:active {
              ${(0, import_colorVars._colorVarsStyle)(base, tone.pressed)}
            }
          }
        }
      }
    }

    /* &:is(a) */
    &[data-as='a'] {
      &[data-disabled] {
        ${(0, import_colorVars._colorVarsStyle)(base, tone.disabled)}
      }

      &:not([data-disabled]) {
        &[data-pressed] {
          ${(0, import_colorVars._colorVarsStyle)(base, tone.pressed)}
        }

        &[data-selected] {
          ${(0, import_colorVars._colorVarsStyle)(base, tone.selected)}
        }

        @media (hover: hover) {
          &:not([data-selected]) {
            &:hover {
              ${(0, import_colorVars._colorVarsStyle)(base, tone.hovered)}
            }

            &:active {
              ${(0, import_colorVars._colorVarsStyle)(base, tone.pressed)}
            }
          }
        }
      }
    }

    ${(_b = (_a = theme.sanity.styles) == null ? void 0 : _a.card) == null ? void 0 : _b.root}
  `;
}
