// @sanity/structure/lib/views/ComponentView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentViewBuilder = void 0;

var _View = require("./View");

var _SerializeError = require("../SerializeError");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var isComponentSpec = spec => {
  return typeof spec === 'object';
};

class ComponentViewBuilder extends _View.GenericViewBuilder {
  constructor(componentOrSpec) {
    var spec = isComponentSpec(componentOrSpec) ? _objectSpread({}, componentOrSpec) : {
      options: {}
    };
    super();

    _defineProperty(this, "spec", void 0);

    this.spec = spec;
    var userComponent = typeof componentOrSpec === 'function' ? componentOrSpec : this.spec.component;

    if (userComponent) {
      // Because we're cloning, this'll return a new instance, so grab the spec from it
      this.spec = this.component(userComponent).spec;
    }
  }

  component(component) {
    return this.clone({
      component
    });
  }

  getComponent() {
    return this.spec.component;
  }

  options(options) {
    return this.clone({
      options
    });
  }

  getOptions() {
    return this.spec.options || {};
  }

  serialize() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    var base = super.serialize(options);
    var component = this.spec.component;

    if (typeof component !== 'function') {
      throw new _SerializeError.SerializeError('`component` is required and must be a function for `component()` view item', options.path, options.index).withHelpUrl(_SerializeError.HELP_URL.COMPONENT_REQUIRED);
    }

    return _objectSpread(_objectSpread({}, base), {}, {
      component,
      options: this.spec.options || {},
      type: 'component'
    });
  }

  clone(withSpec) {
    var builder = new ComponentViewBuilder();
    builder.spec = _objectSpread(_objectSpread({}, this.spec), withSpec || {});
    return builder;
  }

}

exports.ComponentViewBuilder = ComponentViewBuilder;
