// @sanity/ui/lib/cjs/theme/lib/theme/color/factory.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  createColorTheme: () => createColorTheme
});
var import_createButtonModes = __toModule(require("./button/createButtonModes"));
var import_createCardStates = __toModule(require("./card/createCardStates"));
var import_defaults = __toModule(require("./defaults"));
var import_createInputModes = __toModule(require("./input/createInputModes"));
var import_createMuted = __toModule(require("./muted/createMuted"));
var import_createSelectableTones = __toModule(require("./selectable/createSelectableTones"));
var import_createSolidTones = __toModule(require("./solid/createSolidTones"));
var import_createSpot = __toModule(require("./spot/createSpot"));
function createColorTheme(partialOpts = {}) {
  const builders = __spreadValues(__spreadValues({}, import_defaults.defaultOpts), partialOpts);
  return {
    light: _createColorScheme(builders, false),
    dark: _createColorScheme(builders, true)
  };
}
function _createColorScheme(opts, dark) {
  return {
    default: _createColor(opts, dark, "default"),
    transparent: _createColor(opts, dark, "transparent"),
    primary: _createColor(opts, dark, "primary"),
    positive: _createColor(opts, dark, "positive"),
    caution: _createColor(opts, dark, "caution"),
    critical: _createColor(opts, dark, "critical")
  };
}
function _createColor(opts, dark, name) {
  const base = opts.base({ dark, name });
  const solid = (0, import_createSolidTones.createSolidTones)(opts, base, dark, name);
  const muted = (0, import_createMuted.createMutedTones)(opts, base, dark, name);
  return {
    base,
    button: (0, import_createButtonModes.createButtonModes)(opts, base, dark, solid, muted),
    card: (0, import_createCardStates.createCardStates)(opts, base, dark, name, solid, muted),
    dark,
    input: (0, import_createInputModes.createInputModes)(opts, base, dark, solid, muted),
    selectable: (0, import_createSelectableTones.createSelectableTones)(opts, base, dark, solid, muted),
    spot: (0, import_createSpot.createSpot)(opts, base, dark),
    syntax: opts.syntax({ base, dark }),
    solid,
    muted
  };
}
