// @sanity/desk-tool/lib/panes/documentPane/documentPanel/permissionCheckBanner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PermissionCheckBanner = PermissionCheckBanner;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PermissionCheckBanner(props) {
  var _currentUser$roles, _currentUser$roles2;

  var permission = props.permission,
      requiredPermission = props.requiredPermission,
      currentUser = props.currentUser;
  var plural = (currentUser === null || currentUser === void 0 ? void 0 : (_currentUser$roles = currentUser.roles) === null || _currentUser$roles === void 0 ? void 0 : _currentUser$roles.length) !== 1;
  return permission.granted ? null : /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "transparent",
    padding: 2,
    paddingX: 3,
    shadow: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 2,
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 0
  }, /*#__PURE__*/_react.default.createElement(_icons.ReadOnlyIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    marginLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, "Your role", plural && 's', ' ', join(currentUser === null || currentUser === void 0 ? void 0 : (_currentUser$roles2 = currentUser.roles) === null || _currentUser$roles2 === void 0 ? void 0 : _currentUser$roles2.map(r => /*#__PURE__*/_react.default.createElement("code", {
    key: r.name
  }, r.title)), ', '), ' ', "do", plural || 'es', " not have permissions to ", requiredPermission, " this document")))));
}

function join(array, sep) {
  return array.reduce((result, item) => result === null ? [item] : [...result, sep, item], null);
}
