// @sanity/desk-tool/lib/panes/documentPane/documentPanel/menuItems/menuItems.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMenuItems = exports.getProductionPreviewItem = void 0;

var _binaryIcon = _interopRequireDefault(require("part:@sanity/base/binary-icon"));

var _publicIcon = _interopRequireDefault(require("part:@sanity/base/public-icon"));

var _historyIcon = _interopRequireDefault(require("part:@sanity/base/history-icon"));

var _hotkeys = _interopRequireDefault(require("part:@sanity/components/typography/hotkeys"));

var _react = _interopRequireDefault(require("react"));

var _menuItems = _interopRequireDefault(require("./menuItems.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var getHistoryMenuItem = params => {
  var features = params.features,
      hasValue = params.hasValue,
      isHistoryOpen = params.isHistoryOpen;
  if (!features.reviewChanges) return null;
  return {
    action: 'reviewChanges',
    title: 'Review changes',
    icon: _historyIcon.default,
    isDisabled: isHistoryOpen || !hasValue
  };
};

var getInspectItem = _ref => {
  var hasValue = _ref.hasValue;
  return {
    action: 'inspect',
    title: /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.menuItem
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.menuItemLabel
    }, "Inspect"), /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.hotkey
    }, /*#__PURE__*/_react.default.createElement(_hotkeys.default, {
      keys: ['Ctrl', 'Alt', 'I'],
      size: "small"
    }))),
    icon: _binaryIcon.default,
    isDisabled: !hasValue
  };
};

var getProductionPreviewItem = _ref2 => {
  var previewUrl = _ref2.previewUrl;

  if (!previewUrl) {
    return null;
  }

  return {
    action: 'production-preview',
    title: /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.menuItem
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.menuItemLabel
    }, "Open preview"), /*#__PURE__*/_react.default.createElement("span", {
      className: _menuItems.default.hotkey
    }, /*#__PURE__*/_react.default.createElement(_hotkeys.default, {
      keys: ['Ctrl', 'Alt', 'O'],
      size: "small"
    }))),
    icon: _publicIcon.default,
    url: previewUrl
  };
};

exports.getProductionPreviewItem = getProductionPreviewItem;

var getMenuItems = params => {
  var items = [getProductionPreviewItem, getHistoryMenuItem, getInspectItem].filter(Boolean).map(fn => fn(params));
  return items.filter(i => i !== null);
};

exports.getMenuItems = getMenuItems;
