// @sanity/form-builder/lib/utils/WithMaterializedDocument.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class WithMaterializedDocument extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      materialized: null
    });

    _defineProperty(this, "subscription", void 0);
  }

  componentDidMount() {
    this.setDocId(this.props.documentId);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.documentId !== nextProps.documentId) {
      this.setDocId(nextProps.documentId);
    }
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  unsubscribe() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  setDocId(docId) {
    this.unsubscribe();

    if (!docId) {
      this.setState({
        materialized: null
      });
      return;
    }

    this.subscription = this.props.materialize(docId).subscribe(materialized => this.setState({
      materialized
    }));
  }

  render() {
    var materialized = this.state.materialized;
    var children = this.props.children;
    return materialized ? children(materialized) : null;
  }

}

exports.default = WithMaterializedDocument;
