// @sanity/form-builder/lib/utils/withValuePath.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = withValuePath;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function withValuePath(ComposedComponent) {
  var _class, _temp;

  return _temp = _class = class WithValuePath extends _react.default.PureComponent {
    constructor() {
      super(...arguments);

      _defineProperty(this, "_input", void 0);

      _defineProperty(this, "setInput", input => {
        this._input = input;
      });
    }

    focus() {
      this._input.focus();
    }

    render() {
      var getValuePath = this.context.getValuePath;
      return /*#__PURE__*/_react.default.createElement(ComposedComponent, _extends({
        ref: this.setInput
      }, this.props, {
        getValuePath: getValuePath
      }));
    }

  }, _defineProperty(_class, "displayName", "withValuePath(".concat(ComposedComponent.displayName || ComposedComponent.name, ")")), _defineProperty(_class, "contextTypes", {
    getValuePath: _propTypes.default.func,
    formBuilder: _propTypes.default.any
  }), _temp;
}
