// @sanity/base/lib/__legacy/@sanity/components/dialogs/DefaultDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ui = require("@sanity/ui");

var _classnames = _interopRequireDefault(require("classnames"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _containerQuery = require("part:@sanity/components/container-query");

var _defaultStyle = _interopRequireDefault(require("part:@sanity/components/dialogs/default-style"));

var _react = _interopRequireWildcard(require("react"));

var _components = require("../../../../components");

var _hooks = require("../hooks");

var _DefaultDialogActions = require("./DefaultDialogActions");

var _excluded = ["actions", "actionsAlign", "children", "className", "color", "onAction", "onClose", "onClickOutside", "onEscape", "padding", "showCloseButton", "size", "title"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function DefaultDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Portal, null, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "portal"
  }, /*#__PURE__*/_react.default.createElement(DefaultDialogChildren, props)));
}

var _default = DefaultDialog;
exports.default = _default;

function DefaultDialogChildren(props) {
  var actions = props.actions,
      actionsAlign = props.actionsAlign,
      children = props.children,
      classNameProp = props.className,
      color = props.color,
      onAction = props.onAction,
      onClose = props.onClose,
      onClickOutside = props.onClickOutside,
      onEscape = props.onEscape,
      _props$padding = props.padding,
      padding = _props$padding === void 0 ? 'medium' : _props$padding,
      _props$showCloseButto = props.showCloseButton,
      showCloseButton = _props$showCloseButto === void 0 ? true : _props$showCloseButto,
      _props$size = props.size,
      size = _props$size === void 0 ? 'medium' : _props$size,
      title = props.title,
      restProps = _objectWithoutProperties(props, _excluded);

  var _useLayer = (0, _ui.useLayer)(),
      isTopLayer = _useLayer.isTopLayer,
      zIndex = _useLayer.zIndex;

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      cardElement = _useState2[0],
      setCardElement = _useState2[1];

  var renderCloseButton = onClose && showCloseButton;
  var renderFloatingCloseButton = !title && renderCloseButton;
  var contentClassName = actions && actions.length > 0 ? _defaultStyle.default.content : _defaultStyle.default.contentWithoutFooter;
  (0, _hooks.useClickOutside)((0, _react.useCallback)(() => {
    if (isTopLayer) {
      if (onClickOutside) onClickOutside();
    }
  }, [isTopLayer, onClickOutside]), [cardElement]);
  var closeButtonColor = color && ['danger', 'success', 'warning'].includes(color) ? 'white' : undefined;
  (0, _react.useEffect)(() => {
    var handleKeyDown = event => {
      if ((event.shiftKey || isTopLayer) && event.key === 'Escape') {
        event.stopPropagation();
        if (onEscape) onEscape(event); // NOTE: This code used to be `onEscape || onClose`
        else if (onClose) onClose();
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [isTopLayer, onClose, onEscape]);
  var hasActions = actions && actions.length > 0;
  return /*#__PURE__*/_react.default.createElement(_containerQuery.ContainerQuery, _extends({}, restProps, {
    className: (0, _classnames.default)(_defaultStyle.default.root, hasActions && _defaultStyle.default.hasFunctions, classNameProp),
    "data-dialog-color": color,
    "data-dialog-padding": padding,
    "data-dialog-size": size,
    style: {
      zIndex
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.overlay
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.card,
    ref: setCardElement
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.inner
  }, renderFloatingCloseButton && /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.floatingCloseButtonContainer
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    icon: _closeIcon.default,
    kind: "simple",
    onClick: onClose,
    padding: "small",
    title: "Close"
  })), title && /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.header
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.title
  }, /*#__PURE__*/_react.default.createElement("h1", null, title)), renderCloseButton && /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.closeButtonContainer
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    className: _defaultStyle.default.closeButton,
    color: closeButtonColor,
    icon: _closeIcon.default,
    kind: "simple",
    onClick: onClose,
    padding: "small",
    title: "Close"
  }))), /*#__PURE__*/_react.default.createElement(_components.ScrollContainer, {
    className: contentClassName
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.contentWrapper
  }, children)), actions && actions.length > 0 && /*#__PURE__*/_react.default.createElement("div", {
    className: _defaultStyle.default.footer
  }, /*#__PURE__*/_react.default.createElement(_DefaultDialogActions.DefaultDialogActions, {
    actions: actions,
    actionsAlign: actionsAlign,
    onAction: onAction
  })))));
}
