// @sanity/base/lib/__legacy/@sanity/components/dialogs/FullscreenMessageDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ui = require("@sanity/ui");

var _classnames = _interopRequireDefault(require("classnames"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _react = _interopRequireDefault(require("react"));

var _components = require("../../../../components");

var _FullscreenMessageDialog = _interopRequireDefault(require("./FullscreenMessageDialog.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function FullscreenMessageDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Portal, null, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "portal"
  }, /*#__PURE__*/_react.default.createElement(FullscreenMessageDialogChildren, props)));
}

function FullscreenMessageDialogChildren(props) {
  var _useLayer = (0, _ui.useLayer)(),
      zIndex = _useLayer.zIndex;

  var className = (0, _classnames.default)(_FullscreenMessageDialog.default.root, props.color && _FullscreenMessageDialog.default["color_".concat(props.color)]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    style: {
      zIndex
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _FullscreenMessageDialog.default.card
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _FullscreenMessageDialog.default.cardHeader
  }, /*#__PURE__*/_react.default.createElement("h2", {
    className: _FullscreenMessageDialog.default.cardTitle
  }, props.title), props.onClose && /*#__PURE__*/_react.default.createElement("div", {
    className: _FullscreenMessageDialog.default.cardHeaderActions
  }, /*#__PURE__*/_react.default.createElement("button", {
    onClick: props.onClose,
    type: "button"
  }, /*#__PURE__*/_react.default.createElement(_closeIcon.default, null)))), /*#__PURE__*/_react.default.createElement("div", {
    className: _FullscreenMessageDialog.default.cardContent
  }, props.children), props.buttons && /*#__PURE__*/_react.default.createElement("div", {
    className: _FullscreenMessageDialog.default.cardButtons
  }, props.buttons)));
}

var _default = FullscreenMessageDialog;
exports.default = _default;
