// @sanity/base/lib/__legacy/@sanity/components/dialogs/FullscreenDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _partition4 = _interopRequireDefault(require("lodash/partition"));

var _ui = require("@sanity/ui");

var _classnames = _interopRequireDefault(require("classnames"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _fullscreenStyle = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen-style"));

var _react = _interopRequireWildcard(require("react"));

var _components = require("../../../../components");

var _hooks = require("../hooks");

var _excluded = ["children", "color", "title", "cardClassName", "className", "isOpen", "onAction", "onClickOutside", "onClose", "onEscape", "padding", "actions"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function FullscreenDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Portal, null, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "portal"
  }, /*#__PURE__*/_react.default.createElement(FullscreenDialogChildren, props)));
}

var _default = FullscreenDialog;
exports.default = _default;

function FullscreenDialogChildren(props) {
  var children = props.children,
      color = props.color,
      title = props.title,
      cardClassName = props.cardClassName,
      className = props.className,
      _props$isOpen = props.isOpen,
      isOpen = _props$isOpen === void 0 ? true : _props$isOpen,
      onAction = props.onAction,
      onClickOutside = props.onClickOutside,
      onClose = props.onClose,
      onEscape = props.onEscape,
      padding = props.padding,
      actions = props.actions,
      restProps = _objectWithoutProperties(props, _excluded);

  var _useLayer = (0, _ui.useLayer)(),
      isTopLayer = _useLayer.isTopLayer,
      zIndex = _useLayer.zIndex;

  var _partition2 = (0, _partition4.default)(actions, action => action.secondary),
      _partition3 = _slicedToArray(_partition2, 2),
      secondary = _partition3[0],
      primary = _partition3[1];

  (0, _react.useEffect)(() => {
    var handleKeyDown = event => {
      if ((event.shiftKey || isTopLayer) && event.key === 'Escape') {
        if (onEscape) onEscape(event); // NOTE: This code used to be `onEscape || onClose`
        else if (onClose) onClose();
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [isTopLayer, onClose, onEscape]);

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      cardElement = _useState2[0],
      setCardElement = _useState2[1];

  (0, _hooks.useClickOutside)((0, _react.useCallback)(() => {
    if (!isTopLayer) return;
    if (onClickOutside) onClickOutside();
  }, [isTopLayer, onClickOutside]), [cardElement]);
  return /*#__PURE__*/_react.default.createElement("div", _extends({}, restProps, {
    className: (0, _classnames.default)(_fullscreenStyle.default.root, isOpen ? _fullscreenStyle.default.isOpen : _fullscreenStyle.default.isClosed, className),
    style: {
      zIndex
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_fullscreenStyle.default.card, cardClassName),
    ref: setCardElement
  }, (title || onClose) && /*#__PURE__*/_react.default.createElement("header", {
    className: _fullscreenStyle.default.header
  }, title && /*#__PURE__*/_react.default.createElement("h1", {
    className: _fullscreenStyle.default.title
  }, title), onClose && /*#__PURE__*/_react.default.createElement("div", {
    className: _fullscreenStyle.default.actions
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    className: _fullscreenStyle.default.closeButton,
    icon: _closeIcon.default,
    kind: "simple",
    onClick: onClose,
    padding: "small"
  }))), children && /*#__PURE__*/_react.default.createElement(_components.ScrollContainer, {
    className: _fullscreenStyle.default.content
  }, children), actions && actions.length > 0 && /*#__PURE__*/_react.default.createElement("div", {
    className: _fullscreenStyle.default.actionsWrapper
  }, /*#__PURE__*/_react.default.createElement(_buttonGrid.default, {
    align: "end",
    secondary: secondary.map((action, actionIndex) => /*#__PURE__*/_react.default.createElement(FullscreenDialogActionButton, {
      action: action,
      index: actionIndex // eslint-disable-next-line react/no-array-index-key
      ,
      key: actionIndex,
      onAction: onAction
    }))
  }, primary.map((action, actionIndex) => /*#__PURE__*/_react.default.createElement(FullscreenDialogActionButton, {
    action: action,
    index: actionIndex // eslint-disable-next-line react/no-array-index-key
    ,
    key: actionIndex,
    onAction: onAction
  }))))));
}

function FullscreenDialogActionButton(props) {
  var action = props.action,
      index = props.index,
      onAction = props.onAction;
  var handleClick = (0, _react.useCallback)(() => onAction && onAction(action), [action, onAction]);
  return /*#__PURE__*/_react.default.createElement(_default2.default, {
    autoFocus: action.autoFocus,
    className: action.secondary ? _fullscreenStyle.default.actionSecondary : undefined,
    color: action.color,
    "data-action-index": index,
    disabled: action.disabled,
    inverted: action.inverted,
    kind: action.kind,
    onClick: handleClick
  }, action.title);
}
