// @sanity/base/lib/__legacy/@sanity/components/dialogs/PanePopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _circleCheckIcon = _interopRequireDefault(require("part:@sanity/base/circle-check-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _infoIcon = _interopRequireDefault(require("part:@sanity/base/info-icon"));

var _classnames = _interopRequireDefault(require("classnames"));

var _components = require("../../../../components");

var _PanePopover = _interopRequireDefault(require("./PanePopover.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var DEFAULT_ICONS = {
  info: /*#__PURE__*/_react.default.createElement(_infoIcon.default, null),
  success: /*#__PURE__*/_react.default.createElement(_circleCheckIcon.default, null),
  warning: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null),
  error: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)
};

function PanePopoverDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "portal"
  }, /*#__PURE__*/_react.default.createElement(PanePopoverDialogChildren, props));
}

function PanePopoverDialogChildren(props) {
  var children = props.children,
      icon = props.icon,
      id = props.id,
      _props$kind = props.kind,
      kind = _props$kind === void 0 ? 'info' : _props$kind,
      title = props.title,
      subtitle = props.subtitle;

  var _useLayer = (0, _ui.useLayer)(),
      zIndex = _useLayer.zIndex;

  var iconNode = (0, _react.useMemo)(() => {
    if (kind && typeof icon === 'boolean' && icon) return DEFAULT_ICONS[kind];
    if (typeof icon === 'object') return icon;
    return undefined;
  }, [icon, kind]);
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-label": kind,
    "aria-describedby": "popoverTitle-".concat(kind, "-").concat(id),
    className: (0, _classnames.default)(_PanePopover.default.root, _PanePopover.default.dialog),
    "data-kind": kind,
    style: {
      zIndex
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PanePopover.default.inner
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PanePopover.default.content
  }, /*#__PURE__*/_react.default.createElement("div", {
    id: "popoverTitle-".concat(kind, "-").concat(id),
    className: _PanePopover.default.title
  }, icon && /*#__PURE__*/_react.default.createElement("div", {
    role: "img",
    "aria-hidden": true,
    className: _PanePopover.default.icon
  }, iconNode), title), subtitle && /*#__PURE__*/_react.default.createElement("div", {
    className: _PanePopover.default.subtitle
  }, subtitle), children && /*#__PURE__*/_react.default.createElement("div", {
    className: _PanePopover.default.children
  }, children))));
}

var _default = PanePopoverDialog;
exports.default = _default;
