// @sanity/vision/lib/components/DelayedSpinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _LoadingSpinner = _interopRequireDefault(require("./LoadingSpinner"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Waits for X ms before showing a spinner
class DelayedSpinner extends _react.default.PureComponent {
  constructor() {
    super();
    this.state = {};
  }

  componentDidMount() {
    this.timer = setTimeout(() => this.setState({
      show: true
    }), this.props.delay);
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  render() {
    return this.state.show ? /*#__PURE__*/_react.default.createElement(_LoadingSpinner.default, null) : null;
  }

}

DelayedSpinner.propTypes = {
  delay: _propTypes.default.number
};
DelayedSpinner.defaultProps = {
  delay: 500
};
var _default = DelayedSpinner;
exports.default = _default;
