// @sanity/base/lib/change-indicators/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEBUG_LAYER_BOUNDS = exports.DEBUG = exports.STROKE_WIDTH = exports.ARROW_THRESHOLD = exports.ARROW_SIZE = exports.ARROW_MARGIN_Y = exports.ARROW_MARGIN_X = exports.CONNECTOR_MARGIN = exports.INTERACTIVE_STROKE_WIDTH = exports.CORNER_RADIUS = exports.ENABLED = void 0;
// Toggle tracking of change indicator DOM positions and connectors overlay rendering
var ENABLED = true;
exports.ENABLED = ENABLED;
var CORNER_RADIUS = 4;
exports.CORNER_RADIUS = CORNER_RADIUS;
var INTERACTIVE_STROKE_WIDTH = 16;
exports.INTERACTIVE_STROKE_WIDTH = INTERACTIVE_STROKE_WIDTH;
var CONNECTOR_MARGIN = 8;
exports.CONNECTOR_MARGIN = CONNECTOR_MARGIN;
var ARROW_MARGIN_X = 8;
exports.ARROW_MARGIN_X = ARROW_MARGIN_X;
var ARROW_MARGIN_Y = 2;
exports.ARROW_MARGIN_Y = ARROW_MARGIN_Y;
var ARROW_SIZE = 4;
exports.ARROW_SIZE = ARROW_SIZE;
var ARROW_THRESHOLD = 12;
exports.ARROW_THRESHOLD = ARROW_THRESHOLD;
var STROKE_WIDTH = 2;
exports.STROKE_WIDTH = STROKE_WIDTH;
var DEBUG = false;
exports.DEBUG = DEBUG;
var DEBUG_LAYER_BOUNDS = DEBUG || false;
exports.DEBUG_LAYER_BOUNDS = DEBUG_LAYER_BOUNDS;
