// rxjs-etc/dist/cjs/observable/resettable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resettable = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
function resettable(factory, ...args) {
    const resetter = new rxjs_1.Subject();
    const source = new rxjs_1.Subject();
    let destination = factory(...args);
    let subscription = source.subscribe(destination);
    return {
        reset(...args) {
            subscription.unsubscribe();
            destination = factory(...args);
            subscription = source.subscribe(destination);
            resetter.next();
        },
        unsubscribe() {
            subscription.unsubscribe();
            source.unsubscribe();
        },
        get closed() {
            return subscription.closed;
        },
        get observable() {
            return resetter.asObservable().pipe(operators_1.startWith(null), operators_1.switchMap(() => destination));
        },
        get subject() {
            return source;
        },
    };
}
exports.resettable = resettable;
