// @sanity/base/lib/__legacy/@sanity/components/tags/TextField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _textfieldStyle = _interopRequireDefault(require("part:@sanity/components/tags/textfield-style"));

var _excluded = ["onChange", "value", "readOnly", "markers", "inputId"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function removeAt(array, index) {
  var copy = array ? array.slice() : [];
  copy.splice(index, 1);
  return copy;
} // @todo: refactor to functional component


class TagsTextField extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_input", null);

    _defineProperty(this, "state", {
      inputValue: ''
    });

    _defineProperty(this, "handleRemoveTagClick", event => {
      this.removeTag(Number(event.currentTarget.getAttribute('data-index')));
    });

    _defineProperty(this, "handleKeyDown", event => {
      var _this$props$value = this.props.value,
          value = _this$props$value === void 0 ? [] : _this$props$value;
      var inputValue = this.state.inputValue;

      if (event.key === 'Backspace' && inputValue === '') {
        this.removeTag(value.length - 1);
      }
    });

    _defineProperty(this, "handleKeyPress", event => {
      var inputValue = this.state.inputValue;

      if (inputValue && event.key === 'Enter') {
        this.addAndClearInput(inputValue);
      }
    });

    _defineProperty(this, "handleBlur", event => {
      var inputValue = this.state.inputValue;

      if (inputValue) {
        this.addAndClearInput(inputValue);
      }

      if (this.props.onBlur) {
        this.props.onBlur(event);
      }
    });

    _defineProperty(this, "handleInputChange", event => {
      this.setState({
        inputValue: event.currentTarget.value
      });
    });

    _defineProperty(this, "setInput", el => {
      this._input = el;
    });
  }

  addTag(tagValue) {
    var _this$props = this.props,
        _this$props$value2 = _this$props.value,
        value = _this$props$value2 === void 0 ? [] : _this$props$value2,
        onChange = _this$props.onChange;
    onChange(value.concat(tagValue));
  }

  removeTag(index) {
    var _this$props2 = this.props,
        _this$props2$value = _this$props2.value,
        value = _this$props2$value === void 0 ? [] : _this$props2$value,
        onChange = _this$props2.onChange;
    onChange(removeAt(value, index));
  }

  addAndClearInput(tagValue) {
    this.addTag(tagValue);
    this.setState({
      inputValue: ''
    });
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    var inputValue = this.state.inputValue; // eslint-disable-next-line @typescript-eslint/no-unused-vars

    var _this$props3 = this.props,
        onChange = _this$props3.onChange,
        _this$props3$value = _this$props3.value,
        value = _this$props3$value === void 0 ? [] : _this$props3$value,
        readOnly = _this$props3.readOnly,
        markers = _this$props3.markers,
        inputId = _this$props3.inputId,
        rest = _objectWithoutProperties(_this$props3, _excluded);

    return /*#__PURE__*/_react.default.createElement("div", {
      className: readOnly ? _textfieldStyle.default.rootReadOnly : _textfieldStyle.default.root
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _textfieldStyle.default.inner
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _textfieldStyle.default.content
    }, /*#__PURE__*/_react.default.createElement("ul", {
      className: _textfieldStyle.default.tags
    }, value.map((tag, i) => {
      return /*#__PURE__*/_react.default.createElement("li", {
        key: i,
        className: readOnly ? _textfieldStyle.default.tag : _textfieldStyle.default.tagWithClear
      }, tag, !readOnly && /*#__PURE__*/_react.default.createElement("a", {
        onClick: this.handleRemoveTagClick,
        "data-index": i,
        className: _textfieldStyle.default.clearTag
      }, "\xD7"));
    }), /*#__PURE__*/_react.default.createElement("input", _extends({}, rest, {
      readOnly: readOnly,
      value: inputValue,
      className: _textfieldStyle.default.input,
      onKeyDown: this.handleKeyDown,
      onKeyPress: this.handleKeyPress,
      onChange: this.handleInputChange,
      onBlur: this.handleBlur,
      ref: this.setInput,
      autoComplete: "off",
      id: inputId
    })))), /*#__PURE__*/_react.default.createElement("div", {
      className: _textfieldStyle.default.focusHelper
    })));
  }

}

exports.default = TagsTextField;
