// @sanity/form-builder/lib/inputs/files/common/UploadPlaceholder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styles = require("./styles");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var ClipboardBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  opacity: 0.25;\n  ", ":focus & {\n    opacity: 1;\n  }\n  transition-property: opacity;\n  transition-duration: 100ms;\n"])), _styles.FileTarget);

var _default = /*#__PURE__*/_react.default.memo(function UploadPlaceholder(_ref) {
  var _ref$fileType = _ref.fileType,
      fileType = _ref$fileType === void 0 ? 'file' : _ref$fileType;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    height: "fill",
    align: "center",
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginX: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    align: "center",
    size: 5
  }, /*#__PURE__*/_react.default.createElement(_icons.UploadIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "bold"
  }, "Drop ", fileType)))), /*#__PURE__*/_react.default.createElement(ClipboardBox, {
    marginX: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    align: "center",
    size: 5
  }, /*#__PURE__*/_react.default.createElement(_icons.ClipboardImageIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "bold"
  }, "Paste ", fileType)))));
});

exports.default = _default;
