// @sanity/base/lib/util/draftUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isDraft = isDraft;
exports.isDraftId = isDraftId;
exports.isPublishedId = isPublishedId;
exports.getDraftId = getDraftId;
exports.getPublishedId = getPublishedId;
exports.createDraftFrom = createDraftFrom;
exports.newDraftFrom = newDraftFrom;
exports.createPublishedFrom = createPublishedFrom;
exports.collate = collate;
exports.removeDupes = removeDupes;
exports.DRAFTS_FOLDER = void 0;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DRAFTS_FOLDER = 'drafts';
exports.DRAFTS_FOLDER = DRAFTS_FOLDER;
var DRAFTS_PREFIX = "".concat(DRAFTS_FOLDER, ".");

function isDraft(document) {
  return isDraftId(document._id);
}

function isDraftId(id) {
  return id.startsWith(DRAFTS_PREFIX);
}

function isPublishedId(id) {
  return !isDraftId(id);
}

function getDraftId(id) {
  return isDraftId(id) ? id : DRAFTS_PREFIX + id;
}

function getPublishedId(id) {
  return isDraftId(id) ? id.slice(DRAFTS_PREFIX.length) : id;
}

function createDraftFrom(document) {
  return _objectSpread({
    _id: getDraftId(document._id)
  }, document);
}

function newDraftFrom(document) {
  return _objectSpread({
    _id: DRAFTS_PREFIX
  }, document);
}

function createPublishedFrom(document) {
  return _objectSpread({
    _id: getPublishedId(document._id)
  }, document);
} // Takes a list of documents and collates draft/published pairs into single entries
// {id: <published id>, draft?: <draft document>, published?: <published document>}


function collate(documents) {
  var byId = documents.reduce((res, doc) => {
    var id = getPublishedId(doc._id);
    var entry = res.get(id);

    if (!entry) {
      entry = {
        id
      };
      res.set(id, entry);
    }

    entry[id === doc._id ? 'published' : 'draft'] = doc;
    return res;
  }, new Map());
  return Array.from(byId.values());
} // Removes published documents that also has a draft


function removeDupes(documents) {
  return collate(documents).map(entry => entry.draft || entry.published);
}
