// @sanity/imagetool/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageToolWrapper;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ImageLoader = _interopRequireDefault(require("./ImageLoader"));

var _ImageTool = _interopRequireDefault(require("./ImageTool"));

var _Resize = _interopRequireDefault(require("./Resize"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ImageToolWrapper(props) {
  return /*#__PURE__*/_react.default.createElement(_ImageLoader.default, {
    src: props.src
  }, function (_ref) {
    var isLoading = _ref.isLoading,
        image = _ref.image,
        error = _ref.error;

    if (isLoading) {
      return /*#__PURE__*/_react.default.createElement("div", null, "Loading...");
    }

    if (error) {
      return /*#__PURE__*/_react.default.createElement("div", null, error.message);
    }

    if (image) {
      return /*#__PURE__*/_react.default.createElement(_Resize.default, {
        image: image,
        maxHeight: ImageToolWrapper.maxHeight,
        maxWidth: ImageToolWrapper.maxWidth
      }, function (canvas) {
        return /*#__PURE__*/_react.default.createElement(_ImageTool.default, _extends({
          image: canvas
        }, props));
      });
    }

    return null;
  });
}

ImageToolWrapper.propTypes = {
  src: _propTypes.default.string.isRequired,
  value: _propTypes.default.shape({
    hotspot: _propTypes.default.shape({
      x: _propTypes.default.number,
      y: _propTypes.default.number,
      height: _propTypes.default.number,
      width: _propTypes.default.number
    })
  }),
  image: _propTypes.default.shape({
    height: _propTypes.default.number,
    width: _propTypes.default.number
  }),
  onChange: _propTypes.default.func,
  onChangeEnd: _propTypes.default.func,
  readOnly: _propTypes.default.bool
};
ImageToolWrapper.maxHeight = 500;
ImageToolWrapper.maxWidth = 1000;
