// @sanity/base/lib/theme/legacyTones.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.legacyTones = void 0;

var _helpers = require("./helpers");

var _legacyPalette = require("./legacyPalette");

var legacyTones = {
  state: {
    default: {
      bg: _legacyPalette.legacyPalette.component.bg,
      fg: _legacyPalette.legacyPalette.component.fg,
      dark: (0, _helpers._isDark)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.component.fg),
      default: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.gray.base, _legacyPalette.legacyPalette.component.fg),
      primary: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.state.info.fg, _legacyPalette.legacyPalette.component.fg),
      positive: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.state.success.fg, _legacyPalette.legacyPalette.component.fg),
      caution: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.state.warning.fg, _legacyPalette.legacyPalette.component.fg),
      critical: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.state.danger.fg, _legacyPalette.legacyPalette.component.fg)
    },
    navbar: {
      bg: _legacyPalette.legacyPalette.mainNavigation.bg,
      fg: _legacyPalette.legacyPalette.mainNavigation.fg,
      dark: (0, _helpers._isDark)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.mainNavigation.fg),
      default: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.gray.base, _legacyPalette.legacyPalette.mainNavigation.fg),
      primary: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.state.info.fg, _legacyPalette.legacyPalette.mainNavigation.fg),
      positive: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.state.success.fg, _legacyPalette.legacyPalette.mainNavigation.fg),
      caution: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.state.warning.fg, _legacyPalette.legacyPalette.mainNavigation.fg),
      critical: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.state.danger.fg, _legacyPalette.legacyPalette.mainNavigation.fg)
    }
  },
  button: {
    default: {
      bg: _legacyPalette.legacyPalette.component.bg,
      fg: _legacyPalette.legacyPalette.component.fg,
      dark: (0, _helpers._isDark)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.component.fg),
      default: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.defaultButton.default.base, _legacyPalette.legacyPalette.component.fg),
      primary: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.defaultButton.primary.base, _legacyPalette.legacyPalette.component.fg),
      positive: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.defaultButton.success.base, _legacyPalette.legacyPalette.component.fg),
      caution: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.defaultButton.warning.base, _legacyPalette.legacyPalette.component.fg),
      critical: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.component.bg, _legacyPalette.legacyPalette.defaultButton.danger.base, _legacyPalette.legacyPalette.component.fg)
    },
    navbar: {
      bg: _legacyPalette.legacyPalette.mainNavigation.bg,
      fg: _legacyPalette.legacyPalette.mainNavigation.fg,
      dark: (0, _helpers._isDark)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.mainNavigation.fg),
      default: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.defaultButton.default.base, _legacyPalette.legacyPalette.mainNavigation.fg),
      primary: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.defaultButton.primary.base, _legacyPalette.legacyPalette.mainNavigation.fg),
      positive: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.defaultButton.success.base, _legacyPalette.legacyPalette.mainNavigation.fg),
      caution: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.defaultButton.warning.base, _legacyPalette.legacyPalette.mainNavigation.fg),
      critical: (0, _helpers._buildTints)(_legacyPalette.legacyPalette.mainNavigation.bg, _legacyPalette.legacyPalette.defaultButton.danger.base, _legacyPalette.legacyPalette.mainNavigation.fg)
    }
  }
};
exports.legacyTones = legacyTones;
