// @sanity/base/lib/theme/legacyPalette.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.legacyPalette = void 0;

var _sanityCssCustomProperties = _interopRequireDefault(require("sanity:css-custom-properties"));

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var legacyPalette = {
  black: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--black']),
  component: {
    bg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--component-bg']),
    fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--component-text-color'])
  },
  defaultButton: {
    default: {
      base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--default-button-color'])
    },
    primary: {
      base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--default-button-primary-color'])
    },
    success: {
      base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--default-button-success-color'])
    },
    warning: {
      base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--default-button-warning-color'])
    },
    danger: {
      base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--default-button-danger-color'])
    }
  },
  focus: {
    base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--focus-color'])
  },
  gray: {
    base: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--gray-base'])
  },
  mainNavigation: {
    bg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--main-navigation-color']),
    fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--main-navigation-color--inverted'])
  },
  state: {
    info: {
      fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--state-info-color'])
    },
    success: {
      fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--state-success-color'])
    },
    warning: {
      fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--state-warning-color'])
    },
    danger: {
      fg: (0, _helpers._toHex)(_sanityCssCustomProperties.default['--state-danger-color'])
    }
  }
};
exports.legacyPalette = legacyPalette;
