// @sanity/default-layout/lib/sideMenu/SideMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _signOutIcon = _interopRequireDefault(require("part:@sanity/base/sign-out-icon"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _toolSwitcher = _interopRequireDefault(require("part:@sanity/default-layout/tool-switcher"));

var _components2 = require("../components");

var _spaces = require("../util/spaces");

var _SideMenu = _interopRequireDefault(require("./SideMenu.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function SideMenu(props) {
  var activeToolName = props.activeToolName,
      isOpen = props.isOpen,
      onClose = props.onClose,
      onSignOut = props.onSignOut,
      onSwitchTool = props.onSwitchTool,
      router = props.router,
      tools = props.tools,
      user = props.user;
  var zindex = (0, _components.useZIndex)();
  var className = _SideMenu.default.root;
  if (isOpen) className += " ".concat(_SideMenu.default.isOpen);
  var tabIndex = isOpen ? 0 : -1;
  return /*#__PURE__*/_react.default.createElement(_ui.Layer, {
    className: className,
    zOffset: zindex.drawer
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.backdrop
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.inner
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.header
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.headerMain
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userProfile
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userAvatarContainer
  }, /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
    size: "medium",
    userId: "me"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userProfileText
  }, user.name || user.email)), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.closeButtonContainer
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    icon: _closeIcon.default,
    kind: "simple",
    onClick: onClose,
    padding: "small",
    tabIndex: tabIndex,
    title: "Close menu"
  }))), _spaces.HAS_SPACES && /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.datasetSelectContainer
  }, /*#__PURE__*/_react.default.createElement(_components2.DatasetSelect, {
    isVisible: isOpen
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.toolSwitcher
  }, /*#__PURE__*/_react.default.createElement(_toolSwitcher.default, {
    activeToolName: activeToolName,
    direction: "vertical",
    isVisible: isOpen,
    onSwitchTool: onSwitchTool,
    router: router,
    tools: tools
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.menuBottom
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.signOutButton
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    icon: _signOutIcon.default,
    kind: "simple",
    onClick: onSignOut,
    tabIndex: tabIndex
  }, "Sign out")))));
}

var _default = SideMenu;
exports.default = _default;
