// @sanity/field/lib/types/file/diff/FileFieldDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FileFieldDiff = void 0;

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _diff = require("../../../diff");

var _hooks = require("../../../diff/hooks");

var _helpers = require("./helpers");

var _templateObject;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var SizeDiff = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  ", "\n  &:not([hidden]) {\n    display: inline-block;\n  }\n\n  [data-number='positive'] {\n    color: var(--size-diff-positive);\n  }\n\n  [data-number='negative'] {\n    color: var(--size-diff-negative);\n  }\n"])), _ref => {
  var theme = _ref.theme;
  return "\n    --size-diff-positive: ".concat(theme.sanity.color.solid.positive.enabled.bg, ";\n    --size-diff-negative: ").concat(theme.sanity.color.solid.critical.enabled.bg, ";\n  ");
});

var FileFieldDiff = _ref2 => {
  var diff = _ref2.diff,
      schemaType = _ref2.schemaType;
  var fromValue = diff.fromValue,
      toValue = diff.toValue,
      fields = diff.fields;
  var fromAsset = fromValue === null || fromValue === void 0 ? void 0 : fromValue.asset;
  var toAsset = toValue === null || toValue === void 0 ? void 0 : toValue.asset;
  var prev = (0, _hooks.useRefValue)(fromAsset === null || fromAsset === void 0 ? void 0 : fromAsset._ref);
  var next = (0, _hooks.useRefValue)(toAsset === null || toAsset === void 0 ? void 0 : toAsset._ref);
  var changedFields = Object.keys(fields).filter(name => fields[name].isChanged && name !== '_type');
  var didAssetChange = changedFields.includes('asset');
  var nestedFields = schemaType.fields.filter(field => field.name !== 'asset' && changedFields.includes(field.name)).map(field => field.name); // Sizes in MB TODO: improve. Apple uses 1000^2

  var prevSize = (prev === null || prev === void 0 ? void 0 : prev.size) && prev.size / 1000 / 1000;
  var nextSize = (next === null || next === void 0 ? void 0 : next.size) && next.size / 1000 / 1000;
  var pctDiff = (0, _helpers.getSizeDiff)(prevSize, nextSize);
  var roundedPrevSize = prevSize ? prevSize.toFixed(2) : undefined;
  var roundedNextSize = nextSize ? nextSize.toFixed(2) : undefined;
  var cardStyles = (0, _react.useMemo)(() => ({
    display: 'block',
    flex: 1
  }), []);

  var from = prev && /*#__PURE__*/_react.default.createElement(_diff.DiffCard, {
    as: "del",
    diff: diff,
    path: "asset._ref",
    style: cardStyles
  }, /*#__PURE__*/_react.default.createElement(_diff.MetaInfo, {
    title: prev.originalFilename || 'Untitled',
    icon: _fileIcon.default
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 0,
    style: {
      color: 'inherit'
    }
  }, "".concat(roundedPrevSize, "MB"))));

  var to = next && /*#__PURE__*/_react.default.createElement(_diff.DiffCard, {
    as: "ins",
    diff: diff,
    path: "asset._ref",
    style: cardStyles
  }, /*#__PURE__*/_react.default.createElement(_diff.MetaInfo, {
    title: next.originalFilename || 'Untitled',
    icon: _fileIcon.default
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 0,
    style: {
      color: 'inherit'
    }
  }, "".concat(roundedNextSize, "MB")), pctDiff !== 0 && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    radius: 2,
    padding: 1,
    as: SizeDiff,
    marginLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 0,
    "data-number": pctDiff > 0 ? 'positive' : 'negative'
  }, pctDiff > 0 && '+', pctDiff, "%")))));

  var FileAssetChange = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, from && !to && /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff,
    path: "asset._ref",
    description: "Removed"
  }, from), from && to && /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff,
    path: "asset._ref"
  }, /*#__PURE__*/_react.default.createElement(_diff.FromTo, {
    from: from,
    layout: "grid",
    to: to
  })), !from && to && /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff,
    path: "asset._ref",
    description: "Added"
  }, to));

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, didAssetChange && FileAssetChange, nestedFields.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: didAssetChange ? 4 : 3
  }, /*#__PURE__*/_react.default.createElement(_diff.ChangeList, {
    diff: diff,
    schemaType: schemaType,
    fields: nestedFields
  })));
};

exports.FileFieldDiff = FileFieldDiff;
