// @sanity/base/lib/_exports/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useDocumentPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useDocumentPresence;
  }
});
Object.defineProperty(exports, "useGlobalPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useGlobalPresence;
  }
});
Object.defineProperty(exports, "useUser", {
  enumerable: true,
  get: function get() {
    return _hooks2.useUser;
  }
});
Object.defineProperty(exports, "useCurrentUser", {
  enumerable: true,
  get: function get() {
    return _hooks2.useCurrentUser;
  }
});
Object.defineProperty(exports, "unstable_useCheckDocumentPermission", {
  enumerable: true,
  get: function get() {
    return _hooks3.unstable_useCheckDocumentPermission;
  }
});
Object.defineProperty(exports, "unstable_useCanCreateAnyOf", {
  enumerable: true,
  get: function get() {
    return _hooks3.unstable_useCanCreateAnyOf;
  }
});
Object.defineProperty(exports, "useUserColor", {
  enumerable: true,
  get: function get() {
    return _hooks4.useUserColor;
  }
});
Object.defineProperty(exports, "useTimeAgo", {
  enumerable: true,
  get: function get() {
    return _useTimeAgo.useTimeAgo;
  }
});
Object.defineProperty(exports, "useDocumentValues", {
  enumerable: true,
  get: function get() {
    return _useDocumentValues.useDocumentValues;
  }
});

var _hooks = require("../datastores/presence/hooks");

var _hooks2 = require("../datastores/user/hooks");

var _hooks3 = require("../datastores/grants/hooks");

var _hooks4 = require("../user-color/hooks");

var _useTimeAgo = require("../time/useTimeAgo");

var _useDocumentValues = require("../datastores/document/useDocumentValues");
