// @sanity/base/lib/_exports/structure-builder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.view = exports.orderingMenuItemsForType = exports.orderingMenuItem = exports.menuItemsFromInitialValueTemplateItems = exports.menuItemGroup = exports.menuItem = exports.listItem = exports.list = exports.editor = exports.editorWithInitialValueTemplate = exports.documentWithInitialValueTemplate = exports.documentTypeListItems = exports.documentTypeListItem = exports.documentTypeList = exports.documentListItem = exports.documentList = exports.divider = exports.defaults = exports.defaultDocument = exports.component = exports.document = exports.defaultInitialValueTemplateItems = exports.initialValueTemplateItem = exports.default = void 0;

var _structure = require("@sanity/structure");

var _default = _structure.StructureBuilder;
exports.default = _default;
var initialValueTemplateItem = _structure.StructureBuilder.initialValueTemplateItem,
    defaultInitialValueTemplateItems = _structure.StructureBuilder.defaultInitialValueTemplateItems,
    document = _structure.StructureBuilder.document,
    component = _structure.StructureBuilder.component,
    defaultDocument = _structure.StructureBuilder.defaultDocument,
    defaults = _structure.StructureBuilder.defaults,
    divider = _structure.StructureBuilder.divider,
    documentList = _structure.StructureBuilder.documentList,
    documentListItem = _structure.StructureBuilder.documentListItem,
    documentTypeList = _structure.StructureBuilder.documentTypeList,
    documentTypeListItem = _structure.StructureBuilder.documentTypeListItem,
    documentTypeListItems = _structure.StructureBuilder.documentTypeListItems,
    documentWithInitialValueTemplate = _structure.StructureBuilder.documentWithInitialValueTemplate,
    editorWithInitialValueTemplate = _structure.StructureBuilder.editorWithInitialValueTemplate,
    editor = _structure.StructureBuilder.editor,
    list = _structure.StructureBuilder.list,
    listItem = _structure.StructureBuilder.listItem,
    menuItem = _structure.StructureBuilder.menuItem,
    menuItemGroup = _structure.StructureBuilder.menuItemGroup,
    menuItemsFromInitialValueTemplateItems = _structure.StructureBuilder.menuItemsFromInitialValueTemplateItems,
    orderingMenuItem = _structure.StructureBuilder.orderingMenuItem,
    orderingMenuItemsForType = _structure.StructureBuilder.orderingMenuItemsForType,
    view = _structure.StructureBuilder.view;
exports.view = view;
exports.orderingMenuItemsForType = orderingMenuItemsForType;
exports.orderingMenuItem = orderingMenuItem;
exports.menuItemsFromInitialValueTemplateItems = menuItemsFromInitialValueTemplateItems;
exports.menuItemGroup = menuItemGroup;
exports.menuItem = menuItem;
exports.listItem = listItem;
exports.list = list;
exports.editor = editor;
exports.editorWithInitialValueTemplate = editorWithInitialValueTemplate;
exports.documentWithInitialValueTemplate = documentWithInitialValueTemplate;
exports.documentTypeListItems = documentTypeListItems;
exports.documentTypeListItem = documentTypeListItem;
exports.documentTypeList = documentTypeList;
exports.documentListItem = documentListItem;
exports.documentList = documentList;
exports.divider = divider;
exports.defaults = defaults;
exports.defaultDocument = defaultDocument;
exports.component = component;
exports.document = document;
exports.defaultInitialValueTemplateItems = defaultInitialValueTemplateItems;
exports.initialValueTemplateItem = initialValueTemplateItem;
