// @sanity/form-builder/lib/inputs/PortableText/Objects/EditObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditObject = void 0;

var _debounce2 = _interopRequireDefault(require("lodash/debounce"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireWildcard(require("react"));

var _types = require("@sanity/types");

var _portableTextEditor = require("@sanity/portable-text-editor");

var _simplePatch = require("../../../simplePatch");

var _PatchEvent = require("../../../PatchEvent");

var _DefaultObjectEditing = require("./renderers/DefaultObjectEditing");

var _PopoverObjectEditing = require("./renderers/PopoverObjectEditing");

var _FullscreenObjectEditing = require("./renderers/FullscreenObjectEditing");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var PATCHES = new WeakMap();
var IS_THROTTLING = new WeakMap();
var THROTTLE_MS = 300;

// eslint-disable-next-line complexity
var EditObject = _ref => {
  var focusPath = _ref.focusPath,
      markers = _ref.markers,
      objectEditData = _ref.objectEditData,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      onClose = _ref.onClose,
      onFocus = _ref.onFocus,
      presence = _ref.presence,
      readOnly = _ref.readOnly,
      value = _ref.value;
  var editor = (0, _portableTextEditor.usePortableTextEditor)();

  var ptFeatures = _portableTextEditor.PortableTextEditor.getPortableTextFeatures(editor);

  var _useMemo = (0, _react.useMemo)(() => findObjectAndType(objectEditData, value, ptFeatures), [objectEditData, ptFeatures, value]),
      _useMemo2 = _slicedToArray(_useMemo, 2),
      _object = _useMemo2[0],
      type = _useMemo2[1];

  var _useState = (0, _react.useState)(_object),
      _useState2 = _slicedToArray(_useState, 2),
      object = _useState2[0],
      setObject = _useState2[1];

  var _useState3 = (0, _react.useState)(undefined),
      _useState4 = _slicedToArray(_useState3, 2),
      timeoutInstance = _useState4[0],
      setTimeoutInstance = _useState4[1]; // Initialize weakmaps on mount, and send patches on unmount


  (0, _react.useEffect)(() => {
    PATCHES.set(editor, []);
    IS_THROTTLING.set(editor, false);
    return () => {
      sendPatches();
      PATCHES.delete(editor);
      IS_THROTTLING.delete(editor);
    };
  }, []);
  (0, _react.useLayoutEffect)(() => {
    setObject(_object);
  }, [_object]);

  if (!objectEditData) {
    return null;
  }

  var formBuilderPath = objectEditData.formBuilderPath,
      kind = objectEditData.kind;

  function handleClose() {
    onClose();
  }

  var editModalLayout = (0, _get2.default)(type, 'options.editModal');
  var cancelThrottle = (0, _debounce2.default)(() => {
    IS_THROTTLING.set(editor, false);
  }, THROTTLE_MS);

  function handleChange(patchEvent) {
    setObject((0, _simplePatch.applyAll)(object, patchEvent.patches));
    var patches = PATCHES.get(editor);
    IS_THROTTLING.set(editor, true);

    if (patches) {
      PATCHES.set(editor, PATCHES.get(editor).concat(patchEvent.patches));
      sendPatches();
    }
  }

  function sendPatches() {
    if (IS_THROTTLING.get(editor) === true) {
      cancelThrottle();
      clearInterval(timeoutInstance);
      setTimeoutInstance(setTimeout(sendPatches, THROTTLE_MS + 100));
      return;
    }

    var patches = PATCHES.get(editor);

    if (!patches || patches.length === 0) {
      return;
    }

    var length = patches.length;

    var _patches = (0, _portableTextEditor.compactPatches)(PATCHES.get(editor).slice(0, length));

    PATCHES.set(editor, PATCHES.get(editor).slice(length));
    setTimeout(() => {
      onChange(_PatchEvent.PatchEvent.from(_patches), formBuilderPath);
    });
    cancelThrottle();
  }

  if (!object || !type) {
    return null;
  }

  if (editModalLayout === 'fullscreen') {
    return /*#__PURE__*/_react.default.createElement(_FullscreenObjectEditing.FullscreenObjectEditing, {
      focusPath: focusPath,
      markers: markers,
      object: object,
      onBlur: onBlur,
      onChange: handleChange,
      onClose: handleClose,
      onFocus: onFocus,
      path: formBuilderPath,
      presence: presence,
      readOnly: readOnly,
      type: type
    });
  }

  if (editModalLayout === 'popover' || kind === 'annotation' && typeof editModalLayout === 'undefined') {
    return /*#__PURE__*/_react.default.createElement(_PopoverObjectEditing.PopoverObjectEditing, {
      focusPath: focusPath,
      editorPath: objectEditData.editorPath,
      markers: markers,
      object: object,
      onBlur: onBlur,
      onChange: handleChange,
      onClose: handleClose,
      onFocus: onFocus,
      path: formBuilderPath,
      presence: presence,
      readOnly: readOnly,
      type: type
    });
  }

  return /*#__PURE__*/_react.default.createElement(_DefaultObjectEditing.DefaultObjectEditing, {
    focusPath: focusPath,
    markers: markers,
    object: object,
    onBlur: onBlur,
    onChange: handleChange,
    onClose: handleClose,
    onFocus: onFocus,
    path: formBuilderPath,
    presence: presence,
    readOnly: readOnly,
    type: type
  });
};

exports.EditObject = EditObject;

function findObjectAndType(objectEditData, value, ptFeatures) {
  if (!objectEditData) {
    return [undefined, undefined];
  }

  var editorPath = objectEditData.editorPath,
      formBuilderPath = objectEditData.formBuilderPath,
      kind = objectEditData.kind;
  var object;
  var type; // Try finding the relevant block

  var blockKey = Array.isArray(formBuilderPath) && (0, _types.isKeySegment)(formBuilderPath[0]) && formBuilderPath[0]._key;

  var block = value && blockKey && Array.isArray(value) && value.find(blk => blk._key === blockKey);
  var child = block && block.children && block.children.find(cld => (0, _types.isKeySegment)(editorPath[2]) && cld._key === editorPath[2]._key);

  if (block) {
    // Get object, type, and relevant editor element
    switch (kind) {
      case 'blockObject':
        object = block;
        type = ptFeatures.types.blockObjects.find(t => t.name === block._type);
        break;

      case 'inlineObject':
        object = child; // eslint-disable-next-line max-depth

        if (object) {
          type = ptFeatures.types.inlineObjects.find(t => t.name === child._type);
        }

        break;

      case 'annotation':
        // eslint-disable-next-line max-depth
        if (child) {
          var markDef = child.marks && block.markDefs && block.markDefs.find(def => child.marks.includes(def._key)); // eslint-disable-next-line max-depth

          if (markDef) {
            type = ptFeatures.types.annotations.find(t => t.name === markDef._type);
            object = markDef;
          }
        }

        break;

      default: // Nothing

    }
  }

  return [object, type];
}
