// @sanity/base/lib/change-indicators/overlay/Connector.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Connector = Connector;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("../constants");

var _connectorPath = require("./connectorPath");

var _mapConnectorToLine = require("./mapConnectorToLine");

var _ConnectorModule = _interopRequireDefault(require("./Connector.module.css"));

var _ClampedRect = require("./ClampedRect");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Connector(_ref) {
  var from = _ref.from,
      to = _ref.to,
      hovered = _ref.hovered,
      focused = _ref.focused,
      revertHovered = _ref.revertHovered;
  var line = (0, _mapConnectorToLine.mapConnectorToLine)({
    from,
    to
  }); // If both ends of the connector are out of bounds, then do not render

  if (line.from.outOfBounds && line.to.outOfBounds) {
    return null;
  }

  var linePathDescription = (0, _connectorPath.generateConnectorPath)(line);
  var pathClassName = (0, _classnames.default)(_ConnectorModule.default.path, focused && _ConnectorModule.default.focused, revertHovered && _ConnectorModule.default.revertHovered, hovered && !focused && !revertHovered && _ConnectorModule.default.hovered);
  var rightBarClassName = (0, _classnames.default)(_ConnectorModule.default.rightBar, focused && _ConnectorModule.default.rightBarFocused, revertHovered && _ConnectorModule.default.rightBarRevertHovered, hovered && !focused && !revertHovered && _ConnectorModule.default.rightBarHovered);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("path", {
    className: _ConnectorModule.default.interactivePath,
    d: linePathDescription,
    strokeWidth: _constants.INTERACTIVE_STROKE_WIDTH
  }), /*#__PURE__*/_react.default.createElement("path", {
    className: pathClassName,
    d: linePathDescription,
    strokeWidth: _constants.STROKE_WIDTH
  }), /*#__PURE__*/_react.default.createElement(_ClampedRect.ClampedRect, {
    className: rightBarClassName,
    top: to.rect.top,
    left: to.rect.left,
    height: to.rect.height,
    width: _constants.STROKE_WIDTH,
    bounds: to.bounds
  }), line.from.isAbove && /*#__PURE__*/_react.default.createElement("path", {
    className: pathClassName,
    d: (0, _connectorPath.arrowPath)(line.from.left + _constants.ARROW_MARGIN_X, line.from.bounds.top - _constants.ARROW_THRESHOLD + _constants.ARROW_MARGIN_Y, -1),
    strokeWidth: _constants.STROKE_WIDTH
  }), line.from.isBelow && /*#__PURE__*/_react.default.createElement("path", {
    className: pathClassName,
    d: (0, _connectorPath.arrowPath)(line.from.left + _constants.ARROW_MARGIN_X, line.from.bounds.top + line.from.bounds.height + _constants.ARROW_THRESHOLD - _constants.ARROW_MARGIN_Y, 1),
    strokeWidth: _constants.STROKE_WIDTH
  }), line.to.isAbove && /*#__PURE__*/_react.default.createElement("path", {
    className: pathClassName,
    d: (0, _connectorPath.arrowPath)(line.to.bounds.left + _constants.ARROW_MARGIN_X, line.to.bounds.top - _constants.ARROW_THRESHOLD + _constants.ARROW_MARGIN_Y, -1),
    strokeWidth: _constants.STROKE_WIDTH
  }), line.to.isBelow && /*#__PURE__*/_react.default.createElement("path", {
    className: pathClassName,
    d: (0, _connectorPath.arrowPath)(line.to.bounds.left + _constants.ARROW_MARGIN_X, line.to.bounds.top + line.to.bounds.height + _constants.ARROW_THRESHOLD - _constants.ARROW_MARGIN_Y, 1),
    strokeWidth: _constants.STROKE_WIDTH
  }), _constants.DEBUG && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
    className: _ConnectorModule.default.debugRect,
    x: line.from.bounds.left,
    y: line.from.bounds.top,
    width: line.from.bounds.width,
    height: line.from.bounds.height,
    stroke: "green"
  }), /*#__PURE__*/_react.default.createElement("rect", {
    className: _ConnectorModule.default.debugRect,
    x: line.to.bounds.left,
    y: line.to.bounds.top,
    width: line.to.bounds.width,
    height: line.to.bounds.height,
    stroke: "yellow"
  }), !line.from.outOfBounds && /*#__PURE__*/_react.default.createElement("g", {
    transform: "translate(".concat(line.from.bounds.left, " ").concat(line.from.bounds.top, ")")
  }, /*#__PURE__*/_react.default.createElement("rect", {
    className: _ConnectorModule.default.debugRect,
    width: line.from.bounds.width,
    height: line.from.bounds.height,
    strokeWidth: _constants.STROKE_WIDTH,
    stroke: "green"
  })), !line.to.outOfBounds && /*#__PURE__*/_react.default.createElement("g", {
    transform: "translate(".concat(line.to.bounds.left, " ").concat(line.to.bounds.top, ")")
  }, /*#__PURE__*/_react.default.createElement("rect", {
    className: _ConnectorModule.default.debugRect,
    width: line.to.bounds.width,
    height: line.to.bounds.height,
    strokeWidth: _constants.STROKE_WIDTH,
    stroke: "red"
  }))));
}
