// @sanity/base/lib/change-indicators/overlay/DebugLayers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DebugLayers = DebugLayers;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DebugLayers(_ref) {
  var field = _ref.field,
      change = _ref.change;
  return /*#__PURE__*/_react.default.createElement("g", {
    style: {
      pointerEvents: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("rect", {
    x: field.bounds.left,
    y: field.bounds.top,
    height: field.bounds.height,
    width: field.bounds.width,
    stroke: "green",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    x: field.rect.left,
    y: field.rect.top,
    height: field.rect.height,
    width: field.rect.width,
    stroke: "black",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    x: change.bounds.left,
    y: change.bounds.top,
    height: change.bounds.height,
    width: change.bounds.width,
    stroke: "crimson",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    x: change.rect.left,
    y: change.rect.top,
    height: change.rect.height,
    width: change.rect.width,
    stroke: "black",
    fill: "none",
    strokeWidth: 1
  }));
}
