// @sanity/color/lib/esm/config.js
const black = "#101112";
const white = "#fff";
const gray = {
  title: "Gray",
  darkest: black,
  mid: "#8690A0",
  lightest: white,
  midPoint: 500
};
const red = {
  title: "Red",
  darkest: black,
  mid: "#F03E2F",
  lightest: white,
  midPoint: 500
};
const orange = {
  title: "Orange",
  darkest: black,
  mid: "#FB7D24",
  lightest: white,
  midPoint: 450
};
const yellow = {
  title: "Yellow",
  darkest: black,
  mid: "#FBD024",
  lightest: white,
  midPoint: 300
};
const green = {
  title: "Green",
  darkest: black,
  mid: "#43D675",
  lightest: white,
  midPoint: 400
};
const cyan = {
  title: "Cyan",
  darkest: black,
  mid: "#22DAF4",
  lightest: white,
  midPoint: 400
};
const blue = {
  title: "Blue",
  darkest: black,
  mid: "#2276FC",
  lightest: white,
  midPoint: 500
};
const purple = {
  title: "Purple",
  darkest: black,
  mid: "#C123FC",
  lightest: white,
  midPoint: 500
};
const magenta = {
  title: "Magenta",
  darkest: black,
  mid: "#E5389E",
  lightest: white,
  midPoint: 500
};
export {
  black,
  blue,
  cyan,
  gray,
  green,
  magenta,
  orange,
  purple,
  red,
  white,
  yellow
};
