// @sanity/logos/lib/esm/groqMonogram.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import { hues } from "@sanity/color";
import React, { forwardRef } from "react";
const GroqMonogram = forwardRef(function GroqMonogram2(props, ref) {
  return /* @__PURE__ */ React.createElement("svg", __spreadProps(__spreadValues({
    "data-sanity-icon": "groq-monogram",
    width: "1em",
    height: "1em",
    viewBox: "0 0 128 128",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), {
    ref
  }), /* @__PURE__ */ React.createElement("rect", {
    width: "128",
    height: "128",
    rx: "8",
    fill: hues.green[950].hex
  }), /* @__PURE__ */ React.createElement("rect", {
    y: "32",
    width: "64",
    height: "64",
    fill: hues.green[800].hex
  }), /* @__PURE__ */ React.createElement("path", {
    d: "M64 32L32 64H64V32Z",
    fill: hues.magenta[400].hex
  }), /* @__PURE__ */ React.createElement("path", {
    d: "M64 64H32L64 96V64Z",
    fill: hues.green[300].hex
  }), /* @__PURE__ */ React.createElement("path", {
    d: "M96 64H64V96L96 64Z",
    fill: hues.purple[300].hex
  }));
});
export {
  GroqMonogram
};
