// @sanity/schema/lib/legacy/types/any.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnyType = void 0;

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _pick2 = _interopRequireDefault(require("lodash/pick"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OVERRIDABLE_FIELDS = [..._constants.DEFAULT_OVERRIDEABLE_FIELDS];
var ANY_CORE = {
  name: 'any',
  type: null,
  jsonType: 'any'
};
var AnyType = {
  get() {
    return ANY_CORE;
  },

  extend(subTypeDef, extendMember) {
    var parsed = Object.assign((0, _pick2.default)(ANY_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: ANY_CORE,
      of: subTypeDef.of.map(fieldDef => {
        return {
          name: fieldDef.name,
          type: extendMember((0, _omit2.default)(fieldDef, 'name'))
        };
      })
    });
    return subtype(parsed);

    function subtype(parent) {
      return {
        get() {
          return parent;
        },

        extend: extensionDef => {
          if (extensionDef.of) {
            throw new Error('Cannot override `of` property of subtypes of "array"');
          }

          var current = Object.assign({}, parent, (0, _pick2.default)(extensionDef, OVERRIDABLE_FIELDS), {
            type: parent
          });
          return subtype(current);
        }
      };
    }
  }

};
exports.AnyType = AnyType;
