// @sanity/desk-tool/lib/tool/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _uuid = require("@sanity/uuid");

var _router = require("@sanity/base/router");

var _parsePanesSegment = require("../utils/parsePanesSegment");

var _IntentResolver = require("../components/IntentResolver");

var _constants = require("../constants");

var _features = require("../features");

var _DeskTool = _interopRequireDefault(require("./DeskTool"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function toState(pathSegment) {
  return (0, _parsePanesSegment.parsePanesSegment)(decodeURIComponent(pathSegment));
}

function toPath(panes) {
  return (0, _parsePanesSegment.encodePanesSegment)(panes);
}

function legacyEditParamsToState(params) {
  try {
    return JSON.parse(decodeURIComponent(params));
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn('Failed to parse JSON parameters');
    return {};
  }
}

function legacyEditParamsToPath(params) {
  return JSON.stringify(params);
}

var state = {
  activePanes: []
};

function setActivePanes(panes) {
  state.activePanes = panes;
}

function DeskToolPaneStateSyncer(props) {
  var _useRouterState = (0, _router.useRouterState)(),
      intent = _useRouterState.intent,
      params = _useRouterState.params,
      payload = _useRouterState.payload;

  (0, _react.useEffect)(() => {
    // Set active panes to blank on mount and unmount
    setActivePanes([]);
    return () => setActivePanes([]);
  }, []);
  return intent ? /*#__PURE__*/_react.default.createElement(_IntentResolver.IntentResolver, {
    intent: intent,
    params: params,
    payload: payload
  }) : /*#__PURE__*/_react.default.createElement(_DeskTool.default, _extends({}, props, {
    onPaneChange: setActivePanes
  }));
} // eslint-disable-next-line complexity


function getIntentState(intentName, params, currentState, payload) {
  var paneSegments = currentState && currentState.panes || [];
  var activePanes = state.activePanes || [];
  var editDocumentId = params.id || (0, _uuid.uuid)();
  var isTemplate = intentName === 'create' && params.template; // Loop through open panes and see if any of them can handle the intent

  for (var i = activePanes.length - 1; i >= 0; i--) {
    var pane = activePanes[i];

    if (pane.canHandleIntent && pane.canHandleIntent(intentName, params, {
      pane,
      index: i
    })) {
      var paneParams = isTemplate ? {
        template: params.template
      } : _constants.EMPTY_PARAMS;
      return {
        panes: paneSegments.slice(0, i).concat([[{
          id: editDocumentId,
          params: paneParams,
          payload
        }]])
      };
    }
  }

  return {
    intent: intentName,
    params,
    payload
  };
}

var strokeStyle = {
  stroke: 'currentColor',
  strokeWidth: 1.2
}; // @todo: Move to @sanity/base
// eslint-disable-next-line react/no-multi-comp

function MasterDetailIcon() {
  return /*#__PURE__*/_react.default.createElement("svg", {
    "data-sanity-icon": true,
    width: "1em",
    height: "1em",
    viewBox: "0 0 25 25",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("rect", {
    x: "4.5",
    y: "5.5",
    width: "16",
    height: "14",
    style: strokeStyle
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M10.5 5.5V19.5",
    style: strokeStyle
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M6 8.5H9",
    style: strokeStyle
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M13 8.5H18",
    style: strokeStyle
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M6 11.5H9",
    style: strokeStyle
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M6 14.5H9",
    style: strokeStyle
  }));
}

function DeskTool() {
  return /*#__PURE__*/_react.default.createElement(_features.DeskToolFeaturesProvider, null, /*#__PURE__*/_react.default.createElement(DeskToolPaneStateSyncer, null));
}

var _default = {
  router: (0, _router.route)('/', [// "Asynchronous intent resolving" route
  _router.route.intents('/intent'), // Legacy fallback route, will be redirected to new format
  (0, _router.route)('/edit/:type/:editDocumentId', [(0, _router.route)({
    path: '/:params',
    transform: {
      params: {
        toState: legacyEditParamsToState,
        toPath: legacyEditParamsToPath
      }
    }
  })]), // The regular path - when the intent can be resolved to a specific pane
  (0, _router.route)({
    path: '/:panes',
    // Legacy URLs, used to handle redirects
    children: [(0, _router.route)('/:action', (0, _router.route)('/:legacyEditDocumentId'))],
    transform: {
      panes: {
        toState,
        toPath
      }
    }
  })]),

  canHandleIntent(intentName, params) {
    return Boolean(intentName === 'edit' && params.id || intentName === 'create' && params.type || intentName === 'create' && params.template);
  },

  getIntentState,
  title: 'Desk',
  name: 'desk',
  icon: MasterDetailIcon,
  component: DeskTool
};
exports.default = _default;
