// @sanity/base/lib/datastores/grants/debug/exampleGrants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.restricted = exports.viewer = exports.contributor = exports.developer = exports.editor = exports.administrator = void 0;
var administrator = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.administrator = administrator;
var editor = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.editor = editor;
var developer = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.developer = developer;
var contributor = [{
  filter: '_id in path("**")',
  permissions: ['read']
}, {
  filter: '_id in path("drafts.**")',
  permissions: ['create', 'history', 'update']
}];
exports.contributor = contributor;
var viewer = [{
  filter: '_id in path("**")',
  permissions: ['read', 'history']
}];
exports.viewer = viewer;
var restricted = [{
  filter: '_id in path("drafts.**") && _type in ["stringsTest", "book", "author"]',
  permissions: ['read', 'create', 'history', 'update']
}, {
  filter: '_id in path("**")',
  permissions: ['read']
}];
exports.restricted = restricted;
