// @sanity/base/lib/presence/PresenceTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresenceTooltip = PresenceTooltip;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _UserAvatar = require("../components/UserAvatar");

var _PresenceTooltip = require("./PresenceTooltip.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PresenceTooltip(props) {
  var children = props.children,
      items = props.items,
      placement = props.placement;

  var content = /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 1
  }, items.map(item => /*#__PURE__*/_react.default.createElement(_ui.Box, {
    key: item.user.id,
    padding: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_UserAvatar.UserAvatar, {
    user: item.user,
    status: "online"
  })), /*#__PURE__*/_react.default.createElement(_PresenceTooltip.TextWrapper, null, item.user.displayName)))));

  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: content,
    placement: placement
  }, children);
}
