// @sanity/form-builder/lib/transitional/PopoverDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverDialog = PopoverDialog;

var _react = _interopRequireDefault(require("react"));

var _legacyParts = require("../legacyParts");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This is transitional in order to track usage of the PopoverDialog part from within the form-builder package
// At some point we should offer a PopoverDialog component, either as a part of the future sanity studio components library or @sanity/ui
function PopoverDialog(_ref) {
  var children = _ref.children,
      onClose = _ref.onClose,
      placement = _ref.placement,
      referenceElement = _ref.referenceElement,
      depth = _ref.depth,
      preventOverflow = _ref.preventOverflow,
      fallbackPlacements = _ref.fallbackPlacements,
      portal = _ref.portal,
      title = _ref.title,
      size = _ref.size;
  return /*#__PURE__*/_react.default.createElement(_legacyParts.PopoverDialogPart, {
    title: title,
    fallbackPlacements: fallbackPlacements,
    onClose: onClose,
    onEscape: onClose,
    onClickOutside: onClose,
    referenceElement: referenceElement,
    placement: placement,
    preventOverflow: preventOverflow,
    size: size,
    depth: depth,
    portal: portal
  }, children);
}
