// @sanity/mutator/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Matcher", {
  enumerable: true,
  get: function get() {
    return _jsonpath.Matcher;
  }
});
Object.defineProperty(exports, "Expression", {
  enumerable: true,
  get: function get() {
    return _jsonpath.Expression;
  }
});
Object.defineProperty(exports, "extract", {
  enumerable: true,
  get: function get() {
    return _jsonpath.extract;
  }
});
Object.defineProperty(exports, "extractWithPath", {
  enumerable: true,
  get: function get() {
    return _jsonpath.extractWithPath;
  }
});
Object.defineProperty(exports, "arrayToJSONMatchPath", {
  enumerable: true,
  get: function get() {
    return _jsonpath.arrayToJSONMatchPath;
  }
});
Object.defineProperty(exports, "Patcher", {
  enumerable: true,
  get: function get() {
    return _patch.Patcher;
  }
});
Object.defineProperty(exports, "ImmutableAccessor", {
  enumerable: true,
  get: function get() {
    return _patch.ImmutableAccessor;
  }
});
Object.defineProperty(exports, "Document", {
  enumerable: true,
  get: function get() {
    return _document.Document;
  }
});
Object.defineProperty(exports, "Mutation", {
  enumerable: true,
  get: function get() {
    return _document.Mutation;
  }
});
Object.defineProperty(exports, "BufferedDocument", {
  enumerable: true,
  get: function get() {
    return _document.BufferedDocument;
  }
});

var _jsonpath = require("./jsonpath");

var _patch = require("./patch");

var _document = require("./document");
