// @sanity/server/lib/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ReactDOM", {
  enumerable: true,
  get: function get() {
    return _server.default;
  }
});
Object.defineProperty(exports, "getDevServer", {
  enumerable: true,
  get: function get() {
    return _devServer.default;
  }
});
Object.defineProperty(exports, "getProdServer", {
  enumerable: true,
  get: function get() {
    return _prodServer.default;
  }
});
Object.defineProperty(exports, "getWebpackProdConfig", {
  enumerable: true,
  get: function get() {
    return _webpackConfig.default;
  }
});
Object.defineProperty(exports, "getWebpackDevConfig", {
  enumerable: true,
  get: function get() {
    return _webpackConfig2.default;
  }
});
Object.defineProperty(exports, "getWebpackBaseConfig", {
  enumerable: true,
  get: function get() {
    return _webpack2.default;
  }
});
Object.defineProperty(exports, "applyStaticLoaderFix", {
  enumerable: true,
  get: function get() {
    return _applyStaticLoaderFix.default;
  }
});
Object.defineProperty(exports, "getDocumentElement", {
  enumerable: true,
  get: function get() {
    return _baseServer.getDocumentElement;
  }
});
exports.getWebpackCompiler = void 0;

var _server = _interopRequireDefault(require("react-dom/server"));

var _webpack = _interopRequireDefault(require("webpack"));

var _devServer = _interopRequireDefault(require("./devServer"));

var _prodServer = _interopRequireDefault(require("./prodServer"));

var _webpackConfig = _interopRequireDefault(require("./configs/webpack.config.prod"));

var _webpackConfig2 = _interopRequireDefault(require("./configs/webpack.config.dev"));

var _webpack2 = _interopRequireDefault(require("./configs/webpack.config"));

var _applyStaticLoaderFix = _interopRequireDefault(require("./util/applyStaticLoaderFix"));

var _baseServer = require("./baseServer");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getWebpackCompiler = baseConfig => {
  var config = baseConfig.env === 'production' ? (0, _webpackConfig.default)(baseConfig) : (0, _webpackConfig2.default)(baseConfig);
  return (0, _webpack.default)(config);
}; // exported to allow reuse of same React version from CLI


exports.getWebpackCompiler = getWebpackCompiler;
