// @sanity/structure/lib/GenericList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericListBuilder = exports.shallowIntentChecker = void 0;

var _pickBy2 = _interopRequireDefault(require("lodash/pickBy"));

var _camelCase2 = _interopRequireDefault(require("lodash/camelCase"));

var _initialValueTemplates = require("@sanity/initial-value-templates");

var _Icon = require("./parts/Icon");

var _Layout = require("./Layout");

var _MenuItem = require("./MenuItem");

var _MenuItemGroup = require("./MenuItemGroup");

var _Intent = require("./Intent");

var _SerializeError = require("./SerializeError");

var _InitialValueTemplateItem = require("./InitialValueTemplateItem");

var _validateId = require("./util/validateId");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function noChildResolver() {
  return undefined;
}

var shallowIntentChecker = (intentName, params, _ref) => {
  var pane = _ref.pane,
      index = _ref.index;
  return index <= 1 && (0, _Intent.defaultIntentChecker)(intentName, params, {
    pane,
    index
  });
};

exports.shallowIntentChecker = shallowIntentChecker;

class GenericListBuilder {
  constructor() {
    _defineProperty(this, "initialValueTemplatesSpecified", false);

    _defineProperty(this, "spec", {});
  }

  id(id) {
    return this.clone({
      id
    });
  }

  getId() {
    return this.spec.id;
  }

  title(title) {
    return this.clone({
      title,
      id: this.spec.id || (0, _camelCase2.default)(title)
    });
  }

  getTitle() {
    return this.spec.title;
  }

  defaultLayout(defaultLayout) {
    return this.clone({
      defaultLayout
    });
  }

  getDefaultLayout() {
    return this.spec.defaultLayout;
  }

  menuItems(menuItems) {
    return this.clone({
      menuItems
    });
  }

  getMenuItems() {
    return this.spec.menuItems;
  }

  menuItemGroups(menuItemGroups) {
    return this.clone({
      menuItemGroups
    });
  }

  getMenuItemGroups() {
    return this.spec.menuItemGroups;
  }

  child(child) {
    return this.clone({
      child
    });
  }

  getChild() {
    return this.spec.child;
  }

  canHandleIntent(canHandleIntent) {
    return this.clone({
      canHandleIntent
    });
  }

  getCanHandleIntent() {
    return this.spec.canHandleIntent;
  }

  showIcons(enabled) {
    return this.clone({
      displayOptions: _objectSpread(_objectSpread({}, this.spec.displayOptions || {}), {}, {
        showIcons: enabled
      })
    });
  }

  getShowIcons() {
    return this.spec.displayOptions ? this.spec.displayOptions.showIcons : undefined;
  }

  initialValueTemplates(templates) {
    this.initialValueTemplatesSpecified = true;
    return this.clone({
      initialValueTemplates: Array.isArray(templates) ? templates : [templates]
    });
  }

  getInitialValueTemplates() {
    return this.spec.initialValueTemplates;
  }

  serialize() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    var id = this.spec.id || '';
    var path = options.path;
    var defaultLayout = this.spec.defaultLayout;

    if (defaultLayout && !_Layout.layoutOptions.includes(defaultLayout)) {
      throw new _SerializeError.SerializeError("`layout` must be one of ".concat(_Layout.layoutOptions.map(item => "\"".concat(item, "\"")).join(', ')), path, id || options.index, this.spec.title);
    }

    var initialValueTemplates = (this.spec.initialValueTemplates || []).map((item, i) => (0, _InitialValueTemplateItem.maybeSerializeInitialValueTemplateItem)(item, i, path));
    return {
      id: (0, _validateId.validateId)(id, options.path, id || options.index),
      title: this.spec.title,
      type: 'genericList',
      defaultLayout,
      child: this.spec.child || noChildResolver,
      canHandleIntent: this.spec.canHandleIntent || shallowIntentChecker,
      displayOptions: this.spec.displayOptions,
      initialValueTemplates,
      menuItems: menuItemsWithCreateIntents(this.spec, {
        path,
        initialValueTemplates
      }),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) => (0, _MenuItemGroup.maybeSerializeMenuItemGroup)(item, i, path))
    };
  }

  clone(withSpec) {
    var builder = new this.constructor();
    return builder;
  }

}

exports.GenericListBuilder = GenericListBuilder;

function menuItemsWithCreateIntents(list, options) {
  var path = options.path,
      _options$initialValue = options.initialValueTemplates,
      initialValueTemplates = _options$initialValue === void 0 ? [] : _options$initialValue;
  var items = (list.menuItems || []).map((item, i) => (0, _MenuItem.maybeSerializeMenuItem)(item, i, path));
  var hasCreate = items.some(menuItem => menuItem.intent && menuItem.intent.type === 'create');
  var hasTemplates = initialValueTemplates.length > 0;

  if (hasCreate || !hasTemplates) {
    return items;
  }

  var PlusIcon = (0, _Icon.getPlusIcon)();
  var loneTemplate = initialValueTemplates.length === 1 && (0, _InitialValueTemplateItem.maybeSerializeInitialValueTemplateItem)(initialValueTemplates[0], 0, path);
  var actionButton = new _MenuItem.MenuItemBuilder().title('Create new').icon(PlusIcon).showAsAction({
    whenCollapsed: true
  });

  if (loneTemplate) {
    // If we have a single create item, link to that template directly.
    // Otherwise we'll want to select from a menu
    // Action button
    var template = (0, _initialValueTemplates.getTemplateById)(loneTemplate.templateId);
    var templateTitle = template && template.title;
    items.unshift(actionButton.title("Create new ".concat(loneTemplate.title || templateTitle || '')).intent(getCreateIntent(loneTemplate)).serialize());
  } else {
    // More than one item, so we'll want that dropdown of choices
    items.unshift(actionButton.action('toggleTemplateSelectionMenu').serialize());
  }

  return items;
}

function getCreateIntent(templateItem) {
  var tpl = (0, _initialValueTemplates.getTemplateById)(templateItem.templateId);
  var params = (0, _pickBy2.default)({
    type: tpl && tpl.schemaType,
    template: templateItem.templateId
  }, Boolean);
  var intentParams = templateItem.parameters ? [params, templateItem.parameters] : params;
  return {
    type: 'create',
    params: intentParams
  };
}
