// @sanity/structure/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SerializeError", {
  enumerable: true,
  get: function get() {
    return _SerializeError.SerializeError;
  }
});
exports.StructureBuilder = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _List = require("./List");

var _documentTypeListItems = require("./documentTypeListItems");

var _MenuItem = require("./MenuItem");

var _ListItem = require("./ListItem");

var _MenuItemGroup = require("./MenuItemGroup");

var _DocumentList = require("./DocumentList");

var _SerializeError = require("./SerializeError");

var _Document = require("./Document");

var _Component = require("./Component");

var _DocumentListItem = require("./DocumentListItem");

var views = _interopRequireWildcard(require("./views"));

var _InitialValueTemplateItem = require("./InitialValueTemplateItem");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StructureBuilder = {
  defaults: getDefaultStructure,
  documentTypeList: _documentTypeListItems.getDocumentTypeList,
  documentTypeListItem: _documentTypeListItems.getDocumentTypeListItem,
  documentTypeListItems: _documentTypeListItems.getDocumentTypeListItems,
  document: spec => new _Document.DocumentBuilder(spec),
  documentWithInitialValueTemplate: _Document.documentFromEditorWithInitialValue,
  defaultDocument: _Document.getDefaultDocumentNode,
  list: spec => new _List.ListBuilder(spec),
  listItem: spec => new _ListItem.ListItemBuilder(spec),
  menuItem: spec => new _MenuItem.MenuItemBuilder(spec),
  menuItemGroup: spec => new _MenuItemGroup.MenuItemGroupBuilder(spec),
  menuItemsFromInitialValueTemplateItems: _InitialValueTemplateItem.menuItemsFromInitialValueTemplateItems,
  documentList: spec => new _DocumentList.DocumentListBuilder(spec),
  documentListItem: spec => new _DocumentListItem.DocumentListItemBuilder(spec),
  orderingMenuItem: ordering => (0, _MenuItem.getOrderingMenuItem)(ordering),
  orderingMenuItemsForType: type => (0, _MenuItem.getOrderingMenuItemsForSchemaType)(type),
  editor: _Document.documentFromEditor,
  editorWithInitialValueTemplate: _Document.documentFromEditorWithInitialValue,
  defaultInitialValueTemplateItems: _InitialValueTemplateItem.defaultInitialValueTemplateItems,
  initialValueTemplateItem: (templateId, parameters) => new _InitialValueTemplateItem.InitialValueTemplateItemBuilder({
    id: templateId,
    parameters,
    templateId
  }),
  component: spec => {
    return typeof spec === 'function' ? new _Component.ComponentBuilder().component(spec) : new _Component.ComponentBuilder(spec);
  },
  divider: () => ({
    id: (0, _uniqueId2.default)('__divider__'),
    type: 'divider'
  }),
  view: views
};
exports.StructureBuilder = StructureBuilder;

function hasIcon(schemaType) {
  if (!schemaType || typeof schemaType === 'string') {
    return false;
  }

  return Boolean(schemaType.icon);
}

function getDefaultStructure() {
  var items = (0, _documentTypeListItems.getDocumentTypeListItems)();
  return new _List.ListBuilder().id('__root__').title('Content').items(items).showIcons(items.some(item => hasIcon(item.getSchemaType())));
}
