// @sanity/desk-tool/lib/actions/defaultDocumentActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = defaultResolveDocumentActions;
Object.defineProperty(exports, "PublishAction", {
  enumerable: true,
  get: function get() {
    return _PublishAction.PublishAction;
  }
});
Object.defineProperty(exports, "DiscardChangesAction", {
  enumerable: true,
  get: function get() {
    return _DiscardChangesAction.DiscardChangesAction;
  }
});
Object.defineProperty(exports, "UnpublishAction", {
  enumerable: true,
  get: function get() {
    return _UnpublishAction.UnpublishAction;
  }
});
Object.defineProperty(exports, "DuplicateAction", {
  enumerable: true,
  get: function get() {
    return _DuplicateAction.DuplicateAction;
  }
});
Object.defineProperty(exports, "DeleteAction", {
  enumerable: true,
  get: function get() {
    return _DeleteAction.DeleteAction;
  }
});

var _PublishAction = require("./PublishAction");

var _DiscardChangesAction = require("./DiscardChangesAction");

var _UnpublishAction = require("./UnpublishAction");

var _DuplicateAction = require("./DuplicateAction");

var _DeleteAction = require("./DeleteAction");

function defaultResolveDocumentActions() {
  return [_PublishAction.PublishAction, _DiscardChangesAction.DiscardChangesAction, _UnpublishAction.UnpublishAction, _DuplicateAction.DuplicateAction, _DeleteAction.DeleteAction];
}
