// @sanity/desk-tool/lib/actions/PublishAction.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PublishAction = void 0;

var _reactHooks = require("@sanity/react-hooks");

var _icons = require("@sanity/icons");

var _react = _interopRequireWildcard(require("react"));

var _hooks = require("@sanity/base/hooks");

var _components = require("@sanity/base/components");

var _TimeAgo = require("../components/TimeAgo");

var _documentHistory = require("../panes/documentPane/documentHistory");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var DISABLED_REASON_TITLE = {
  LIVE_EDIT_ENABLED: 'Cannot publish since liveEdit is enabled for this document type',
  ALREADY_PUBLISHED: 'Already published',
  NO_CHANGES: 'No unpublished changes'
};

function getDisabledReason(reason, publishedAt) {
  if (reason === 'ALREADY_PUBLISHED' && publishedAt) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", null, "Published ", /*#__PURE__*/_react.default.createElement(_TimeAgo.TimeAgo, {
      time: publishedAt
    })));
  }

  return DISABLED_REASON_TITLE[reason];
} // eslint-disable-next-line complexity


var PublishAction = props => {
  var id = props.id,
      type = props.type,
      liveEdit = props.liveEdit,
      draft = props.draft,
      published = props.published;

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      publishState = _useState2[0],
      setPublishState = _useState2[1];

  var _useDocumentOperation = (0, _reactHooks.useDocumentOperation)(id, type),
      publish = _useDocumentOperation.publish;

  var validationStatus = (0, _reactHooks.useValidationStatus)(id, type);
  var syncState = (0, _reactHooks.useSyncState)(id);

  var _useDocumentHistory = (0, _documentHistory.useDocumentHistory)(),
      historyOpen = _useDocumentHistory.open,
      historyController = _useDocumentHistory.historyController;

  var hasValidationErrors = validationStatus.markers.some(marker => marker.level === 'error'); // we use this to "schedule" publish after pending tasks (e.g. validation and sync) has completed

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      publishScheduled = _useState4[0],
      setPublishScheduled = _useState4[1];

  var isNeitherSyncingNorValidating = !syncState.isSyncing && !validationStatus.isValidating;
  var publishPermission = (0, _hooks.unstable_useCheckDocumentPermission)(id, type, 'publish');

  var _useCurrentUser = (0, _hooks.useCurrentUser)(),
      currentUser = _useCurrentUser.value; // eslint-disable-next-line no-nested-ternary


  var title = publish.disabled ? getDisabledReason(publish.disabled, (published || {})._updatedAt) || '' : hasValidationErrors ? 'There are validation errors that need to be fixed before this document can be published' : '';
  var hasDraft = Boolean(draft);
  var doPublish = (0, _react.useCallback)(() => {
    publish.execute();
    setPublishState('publishing');
  }, [publish]);
  (0, _react.useEffect)(() => {
    if (publishScheduled && isNeitherSyncingNorValidating) {
      if (!hasValidationErrors) {
        doPublish();
      }

      setPublishScheduled(false);
    }
  }, [isNeitherSyncingNorValidating, doPublish, hasValidationErrors, publishScheduled]);
  (0, _react.useEffect)(() => {
    var didPublish = publishState === 'publishing' && !hasDraft;

    if (didPublish) {
      if (historyController.changesPanelActive()) {
        // Re-open the panel
        historyOpen();
      }
    }

    var nextState = didPublish ? 'published' : null;
    var delay = didPublish ? 200 : 4000;
    var timer = setTimeout(() => {
      setPublishState(nextState);
    }, delay);
    return () => clearTimeout(timer);
  }, [publishState, hasDraft, historyController, historyOpen]);
  var handle = (0, _react.useCallback)(() => {
    if (syncState.isSyncing || validationStatus.isValidating) {
      setPublishScheduled(true);
    } else {
      doPublish();
    }
  }, [syncState.isSyncing, validationStatus.isValidating, doPublish]);

  if (liveEdit) {
    return {
      color: 'success',
      label: 'Publish',
      title: 'Live Edit is enabled for this content type and publishing happens automatically as you make changes',
      disabled: true
    };
  }

  if (!publishPermission.granted) {
    return {
      color: 'success',
      label: 'Publish',
      title: /*#__PURE__*/_react.default.createElement(_components.InsufficientPermissionsMessage, {
        operationLabel: "publish this document",
        currentUser: currentUser
      }),
      disabled: true
    };
  }

  var disabled = Boolean(publishScheduled || publishState === 'publishing' || publishState === 'published' || hasValidationErrors || publish.disabled);
  return {
    disabled,
    color: 'success',
    label: // eslint-disable-next-line no-nested-ternary
    publishState === 'published' ? 'Published' : publishScheduled || publishState === 'publishing' ? 'Publishing…' : 'Publish',
    // @todo: Implement loading state, to show a `<Button loading />` state
    // loading: publishScheduled || publishState === 'publishing',
    icon: publishState === 'published' ? _icons.CheckmarkIcon : _icons.PublishIcon,
    // eslint-disable-next-line no-nested-ternary
    title: publishScheduled ? 'Waiting for tasks to finish before publishing' : publishState === 'published' || publishState === 'publishing' ? null : title,
    shortcut: disabled || publishScheduled ? null : 'Ctrl+Alt+P',
    onHandle: handle
  };
};

exports.PublishAction = PublishAction;
