// @sanity/base/lib/search/weighted/createWeightedSearch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createWeightedSearch = createWeightedSearch;

var _union2 = _interopRequireDefault(require("lodash/union"));

var _sortBy2 = _interopRequireDefault(require("lodash/sortBy"));

var _flow2 = _interopRequireDefault(require("lodash/flow"));

var _uniq2 = _interopRequireDefault(require("lodash/uniq"));

var _flatten2 = _interopRequireDefault(require("lodash/flatten"));

var _toLower2 = _interopRequireDefault(require("lodash/toLower"));

var _compact2 = _interopRequireDefault(require("lodash/compact"));

var _operators = require("rxjs/operators");

var _searchUtils = require("../../util/searchUtils");

var _draftUtils = require("../../util/draftUtils");

var _tokenize = require("../common/tokenize");

var _applyWeights = require("./applyWeights");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var combinePaths = (0, _flow2.default)([_flatten2.default, _union2.default, _compact2.default]);

var toGroqParams = terms => {
  var params = {};
  return terms.reduce((acc, term, i) => {
    acc["t".concat(i)] = "".concat(term, "*"); // "t" is short for term

    return acc;
  }, params);
};

var pathWithMapper = _ref => {
  var mapWith = _ref.mapWith,
      path = _ref.path;
  return mapWith ? "".concat(mapWith, "(").concat(path, ")") : path;
};

function createWeightedSearch(types, client) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var filter = options.filter,
      params = options.params,
      tag = options.tag;
  var searchSpec = types.map(type => ({
    typeName: type.name,
    paths: type.__experimental_search.map(config => ({
      weight: config.weight,
      path: (0, _searchUtils.joinPath)(config.path),
      mapWith: config.mapWith
    }))
  }));
  var combinedSearchPaths = combinePaths(searchSpec.map(configForType => configForType.paths.map(opt => pathWithMapper(opt))));
  var selections = searchSpec.map(spec => {
    var constraint = "_type == \"".concat(spec.typeName, "\" => ");
    var selection = "{ ".concat(spec.paths.map((cfg, i) => "\"w".concat(i, "\": ").concat(pathWithMapper(cfg))), " }");
    return "".concat(constraint).concat(selection);
  }); // this is the actual search function that takes the search string and returns the hits

  return function search(queryString) {
    var searchOpts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var terms = (0, _uniq2.default)((0, _compact2.default)((0, _tokenize.tokenize)((0, _toLower2.default)(queryString))));
    var constraints = terms.map((term, i) => combinedSearchPaths.map(joinedPath => "".concat(joinedPath, " match $t").concat(i))).filter(constraint => constraint.length > 0);
    var filters = ['_type in $__types', searchOpts.includeDrafts === false && "!(_id in path('drafts.**'))", ...constraints.map(constraint => "(".concat(constraint.join('||'), ")")), filter ? "(".concat(filter, ")") : ''].filter(Boolean);
    var selection = selections.length > 0 ? "...select(".concat(selections.join(',\n'), ")") : '';
    var query = "*[".concat(filters.join('&&'), "][0...$__limit]{_type, _id, ").concat(selection, "}");
    return client.observable.fetch(query, _objectSpread(_objectSpread({}, toGroqParams(terms)), {}, {
      __types: searchSpec.map(spec => spec.typeName),
      __limit: 1000
    }, params || {}), {
      tag
    }).pipe((0, _operators.map)(_draftUtils.removeDupes), (0, _operators.map)(hits => (0, _applyWeights.applyWeights)(searchSpec, hits, terms)), (0, _operators.map)(hits => (0, _sortBy2.default)(hits, hit => -hit.score)), (0, _operators.map)(hits => hits.slice(0, 100)));
  };
}
