// @sanity/ui/lib/cjs/components/toast/toast.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  Toast: () => Toast
});
var import_icons = __toModule(require("@sanity/icons"));
var import_react = __toModule(require("react"));
var import_styled_components = __toModule(require("styled-components"));
var import_primitives = __toModule(require("../../primitives"));
const STATUS_CARD_TONE = {
  error: "critical",
  warning: "caution",
  success: "positive",
  info: "primary"
};
const ROLES = {
  error: "alert",
  warning: "alert",
  success: "alert",
  info: "alert"
};
const Root = (0, import_styled_components.default)(import_primitives.Card)`
  pointer-events: all;
`;
const TextBox = (0, import_styled_components.default)(import_primitives.Flex)`
  overflow-x: auto;
`;
function Toast(props) {
  const _a = props, { closable, description, onClose, title, status } = _a, restProps = __objRest(_a, ["closable", "description", "onClose", "title", "status"]);
  const cardTone = status ? STATUS_CARD_TONE[status] : "default";
  const role = status ? ROLES[status] : "status";
  return /* @__PURE__ */ import_react.default.createElement(Root, __spreadProps(__spreadValues({
    "data-ui": "Toast",
    role
  }, restProps), {
    marginTop: 3,
    radius: 2,
    shadow: 2,
    tone: cardTone
  }), /* @__PURE__ */ import_react.default.createElement(import_primitives.Flex, {
    align: "flex-start"
  }, /* @__PURE__ */ import_react.default.createElement(TextBox, {
    flex: 1,
    padding: 3
  }, /* @__PURE__ */ import_react.default.createElement(import_primitives.Stack, {
    space: 3
  }, title && /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, {
    weight: "semibold"
  }, title), description && /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, {
    muted: true,
    size: 1
  }, description))), closable && /* @__PURE__ */ import_react.default.createElement(import_primitives.Box, {
    padding: 1
  }, /* @__PURE__ */ import_react.default.createElement(import_primitives.Button, {
    as: "button",
    icon: import_icons.CloseIcon,
    mode: "bleed",
    padding: 2,
    onClick: onClose,
    style: { verticalAlign: "top" }
  }))));
}
