// rxjs-etc/dist/cjs/operators/deferFinalize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deferFinalize = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
function deferFinalize(callback) {
    return (source) => source.lift(new DeferFinalizeOperator(callback));
}
exports.deferFinalize = deferFinalize;
class DeferFinalizeOperator {
    constructor(callback) {
        this.callback = callback;
    }
    call(subscriber, source) {
        return source.subscribe(new DeferFinalizeSubscriber(subscriber, this.callback));
    }
}
class DeferFinalizeSubscriber extends rxjs_1.Subscriber {
    constructor(destination, callback) {
        super(destination);
        this.callback = callback;
        this.kind = "U";
        this.subscription = undefined;
    }
    complete() {
        this.kind = "C";
        this.defer(() => super.complete());
    }
    error(error) {
        this.kind = "E";
        this.defer(() => super.error(error));
    }
    unsubscribe() {
        this.defer(() => super.unsubscribe());
    }
    defer(func) {
        if (this.subscription) {
            this.subscription.add(func);
            return;
        }
        const subscription = new rxjs_1.Subscription();
        this.subscription = subscription;
        subscription.add(func);
        const result = this.callback(this.kind);
        rxjs_1.from(result)
            .pipe(operators_1.finalize(() => subscription.unsubscribe()))
            .subscribe();
    }
}
