// @sanity/diff/lib/calculate/diffSimple.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.diffNumber = diffNumber;
exports.diffBoolean = diffBoolean;

function diffNumber(fromInput, toInput, options) {
  var fromValue = fromInput.value;
  var toValue = toInput.value;
  var type = fromInput.type;
  if (fromValue === toValue) return {
    type,
    action: 'unchanged',
    fromValue,
    toValue,
    isChanged: false
  };
  return {
    type: fromInput.type,
    action: 'changed',
    isChanged: true,
    fromValue: fromValue,
    toValue: toValue,
    annotation: toInput.annotation
  };
}

function diffBoolean(fromInput, toInput, options) {
  var fromValue = fromInput.value;
  var toValue = toInput.value;
  var type = fromInput.type;
  if (fromValue === toValue) return {
    type,
    action: 'unchanged',
    fromValue,
    toValue,
    isChanged: false
  };
  return {
    type: fromInput.type,
    action: 'changed',
    isChanged: true,
    fromValue: fromValue,
    toValue: toValue,
    annotation: toInput.annotation
  };
}
