// @sanity/default-layout/lib/schemaErrors/SchemaErrors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _errorIcon = _interopRequireDefault(require("part:@sanity/base/error-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _generateHelpUrl = _interopRequireDefault(require("@sanity/generate-help-url"));

var _SchemaErrors = _interopRequireDefault(require("./SchemaErrors.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function renderPath(path) {
  return path.map((segment, i) => {
    var key = "s_".concat(i);

    if (segment.kind === 'type') {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: _SchemaErrors.default.segment,
        key: key
      }, /*#__PURE__*/_react.default.createElement("span", {
        key: "name",
        className: _SchemaErrors.default.pathSegmentTypeName
      }, segment.name), "\u2002", /*#__PURE__*/_react.default.createElement("span", {
        key: "type",
        className: _SchemaErrors.default.pathSegmentTypeType
      }, segment.type));
    }

    if (segment.kind === 'property') {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: _SchemaErrors.default.segment,
        key: key
      }, /*#__PURE__*/_react.default.createElement("span", {
        className: _SchemaErrors.default.pathSegmentProperty
      }, segment.name));
    }

    if (segment.kind === 'type') {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: _SchemaErrors.default.segment,
        key: key
      }, /*#__PURE__*/_react.default.createElement("span", {
        key: "name",
        className: _SchemaErrors.default.pathSegmentTypeName
      }, segment.name), /*#__PURE__*/_react.default.createElement("span", {
        key: "type",
        className: _SchemaErrors.default.pathSegmentTypeType
      }, segment.type));
    }

    return null;
  }).filter(Boolean);
}

function SchemaErrors(props) {
  var problemGroups = props.problemGroups;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _SchemaErrors.default.root
  }, /*#__PURE__*/_react.default.createElement("h2", {
    className: _SchemaErrors.default.title
  }, "Uh oh\u2026 found errors in schema"), /*#__PURE__*/_react.default.createElement("ul", {
    className: _SchemaErrors.default.list
  }, problemGroups.map((group, i) => {
    return /*#__PURE__*/_react.default.createElement("li", {
      key: "g_".concat(i),
      className: _SchemaErrors.default.listItem
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _SchemaErrors.default.path
    }, renderPath(group.path)), /*#__PURE__*/_react.default.createElement("ul", {
      className: _SchemaErrors.default.problems
    }, group.problems.map((problem, j) => /*#__PURE__*/_react.default.createElement("li", {
      key: "g_".concat(i, "_p_").concat(j),
      className: _SchemaErrors.default["problem_".concat(problem.severity)]
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _SchemaErrors.default.problemSeverity
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: _SchemaErrors.default.problemSeverityIcon
    }, problem.severity === 'error' && /*#__PURE__*/_react.default.createElement(_errorIcon.default, null), problem.severity === 'warning' && /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)), /*#__PURE__*/_react.default.createElement("span", {
      className: _SchemaErrors.default.problemSeverityText
    }, problem.severity)), /*#__PURE__*/_react.default.createElement("div", {
      className: _SchemaErrors.default.problemContent
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _SchemaErrors.default.problemMessage
    }, problem.message), problem.helpId && /*#__PURE__*/_react.default.createElement("a", {
      className: _SchemaErrors.default.problemLink,
      href: (0, _generateHelpUrl.default)(problem.helpId),
      target: "_blank",
      rel: "noopener noreferrer"
    }, "View documentation"))))));
  })));
}

SchemaErrors.propTypes = {
  problemGroups: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.arrayOf(_propTypes.default.shape({
      kind: _propTypes.default.string,
      type: _propTypes.default.string,
      name: _propTypes.default.string
    })),
    problems: _propTypes.default.arrayOf(_propTypes.default.shape({
      severity: _propTypes.default.string
    }))
  }).isRequired).isRequired
};
var _default = SchemaErrors;
exports.default = _default;
