// @sanity/form-builder/lib/inputs/DateInputs/base/calendar/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TAIL_WEEKDAYS = exports.DEFAULT_TIME_PRESETS = exports.ARROW_KEYS = exports.HOURS_24 = exports.WEEK_DAY_NAMES = exports.MONTH_NAMES = void 0;

var _range2 = _interopRequireDefault(require("lodash/range"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MONTH_NAMES = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
exports.MONTH_NAMES = MONTH_NAMES;
var WEEK_DAY_NAMES = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
exports.WEEK_DAY_NAMES = WEEK_DAY_NAMES;
var HOURS_24 = (0, _range2.default)(0, 24);
exports.HOURS_24 = HOURS_24;
var ARROW_KEYS = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'];
exports.ARROW_KEYS = ARROW_KEYS;
var DEFAULT_TIME_PRESETS = [[0, 0], [6, 0], [12, 0], [18, 0], [23, 59]]; // all weekdays except first

exports.DEFAULT_TIME_PRESETS = DEFAULT_TIME_PRESETS;
var TAIL_WEEKDAYS = [1, 2, 3, 4, 5, 6];
exports.TAIL_WEEKDAYS = TAIL_WEEKDAYS;
