// @sanity/form-builder/lib/inputs/DateInputs/base/calendar/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatTime = exports.getWeeksOfMonth = exports.getWeekDaysFromWeekStarts = exports.getWeekStartsOfMonth = void 0;

var _dateFns = require("date-fns");

var _constants = require("./constants");

var getWeekStartsOfMonth = date => {
  var firstDay = (0, _dateFns.startOfMonth)(date);
  return (0, _dateFns.eachWeekOfInterval)({
    start: firstDay,
    end: (0, _dateFns.lastDayOfMonth)(firstDay)
  });
};

exports.getWeekStartsOfMonth = getWeekStartsOfMonth;

var getWeekDaysFromWeekStarts = weekStarts => {
  return weekStarts.map(weekStart => [weekStart, ..._constants.TAIL_WEEKDAYS.map(d => (0, _dateFns.addDays)(weekStart, d))]);
};

exports.getWeekDaysFromWeekStarts = getWeekDaysFromWeekStarts;

var getWeeksOfMonth = date => getWeekDaysFromWeekStarts(getWeekStartsOfMonth(date)).map(days => ({
  number: (0, _dateFns.getWeek)(days[0]),
  days
}));

exports.getWeeksOfMonth = getWeeksOfMonth;

var formatTime = (hours, minutes) => "".concat("".concat(hours).padStart(2, '0'), ":").concat("".concat(minutes).padStart(2, '0'));

exports.formatTime = formatTime;
