// @sanity/desk-tool/lib/panes/documentPane/timeline/timelineItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimelineItem = TimelineItem;

var _react = _interopRequireWildcard(require("react"));

var _hooks = require("@sanity/base/hooks");

var _helpers = require("./helpers");

var _userAvatarStack = require("./userAvatarStack");

var _timelineItem = _interopRequireDefault(require("./timelineItem.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function TimelineItem(props) {
  var isSelectionBottom = props.isSelectionBottom,
      isSelectionTop = props.isSelectionTop,
      state = props.state,
      onSelect = props.onSelect,
      timestamp = props.timestamp,
      chunk = props.chunk,
      type = props.type;
  var iconComponent = (0, _helpers.getTimelineEventIconComponent)(type);
  var authorUserIds = Array.from(chunk.authors);
  var timeAgo = (0, _hooks.useTimeAgo)(timestamp, {
    minimal: true
  });
  var handleClick = (0, _react.useCallback)(evt => {
    evt.preventDefault();
    evt.stopPropagation();
    onSelect(chunk);
  }, [onSelect, chunk]);
  return /*#__PURE__*/_react.default.createElement("li", {
    className: _timelineItem.default.root,
    "data-chunk-id": chunk.id,
    "data-state": state,
    "data-selection-bottom": isSelectionBottom,
    "data-selection-top": isSelectionTop,
    "data-type": type
  }, /*#__PURE__*/_react.default.createElement("button", {
    disabled: state === 'disabled' || state === 'selected',
    onClick: handleClick,
    type: "button"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.wrapper
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.iconContainer
  }, iconComponent && /*#__PURE__*/(0, _react.createElement)(iconComponent)), /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.textContainer
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.typeName
  }, (0, _helpers.formatTimelineEventLabel)(type) || /*#__PURE__*/_react.default.createElement("code", null, type)), /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.timestamp
  }, timeAgo)), /*#__PURE__*/_react.default.createElement("div", {
    className: _timelineItem.default.avatarStackContainer
  }, /*#__PURE__*/_react.default.createElement(_userAvatarStack.UserAvatarStack, {
    maxLength: 3,
    userIds: authorUserIds
  })))));
}
