// @sanity/form-builder/lib/inputs/PortableText/expandCollapseButton/expandCollapseButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandCollapseButton = ExpandCollapseButton;

var _fullscreenExitIcon = _interopRequireDefault(require("part:@sanity/base/fullscreen-exit-icon"));

var _fullscreenIcon = _interopRequireDefault(require("part:@sanity/base/fullscreen-icon"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var IS_MAC = typeof window != 'undefined' && /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);

function ExpandCollapseButton(props) {
  var isFullscreen = props.isFullscreen,
      onToggleFullscreen = props.onToggleFullscreen;
  return /*#__PURE__*/_react.default.createElement(_default.default, {
    icon: isFullscreen ? _fullscreenExitIcon.default : _fullscreenIcon.default,
    kind: "simple",
    onClick: onToggleFullscreen,
    padding: "small",
    title: "Open in fullscreen (".concat(IS_MAC ? 'cmd' : 'ctrl', "+enter)")
  });
}
