// @sanity/default-layout/lib/navbar/studioStatus/CurrentVersionsDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _components = require("@sanity/base/components");

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class CurrentVersionsDialog extends _react.default.PureComponent {
  renderTable() {
    var versions = this.props.versions;
    return /*#__PURE__*/_react.default.createElement("table", {
      className: _UpdateNotifierDialog.default.versionsTable
    }, /*#__PURE__*/_react.default.createElement("thead", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("th", null, "Module"), /*#__PURE__*/_react.default.createElement("th", null, "Installed"), /*#__PURE__*/_react.default.createElement("th", null, "Latest"))), /*#__PURE__*/_react.default.createElement("tbody", null, Object.keys(versions).map(pkgName => /*#__PURE__*/_react.default.createElement("tr", {
      key: pkgName
    }, /*#__PURE__*/_react.default.createElement("td", null, pkgName), /*#__PURE__*/_react.default.createElement("td", null, versions[pkgName]), /*#__PURE__*/_react.default.createElement("td", null, versions[pkgName])))));
  }

  render() {
    var onClose = this.props.onClose;
    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarDialog"
    }, /*#__PURE__*/_react.default.createElement(_default2.default, {
      onClose: onClose,
      onClickOutside: onClose,
      size: "medium"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.content
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _UpdateNotifierDialog.default.dialogHeading
    }, "This Studio is up to date"), /*#__PURE__*/_react.default.createElement("p", null, "It was built using the latest versions of all packages."), /*#__PURE__*/_react.default.createElement("details", {
      className: _UpdateNotifierDialog.default.details
    }, /*#__PURE__*/_react.default.createElement("summary", {
      className: _UpdateNotifierDialog.default.summary
    }, "List all installed packages"), this.renderTable()))));
  }

}

_defineProperty(CurrentVersionsDialog, "defaultProps", {
  versions: []
});

var _default = CurrentVersionsDialog;
exports.default = _default;
