// @sanity/default-layout/lib/navbar/studioStatus/SanityStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityStatus;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _packageIcon = _interopRequireDefault(require("part:@sanity/base/package-icon"));

var _autoId = require("@reach/auto-id");

var _CurrentVersionsDialog = _interopRequireDefault(require("./CurrentVersionsDialog"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function formatUpdateLabel(len) {
  if (len === 1) return ' 1 update';
  return "".concat(len, " updates");
}

function SanityStatus(props) {
  var isSupported = props.isSupported,
      isUpToDate = props.isUpToDate,
      level = props.level,
      onHideDialog = props.onHideDialog,
      onShowDialog = props.onShowDialog,
      _props$outdated = props.outdated,
      outdated = _props$outdated === void 0 ? [] : _props$outdated,
      showDialog = props.showDialog,
      versions = props.versions;
  var elementId = (0, _autoId.useId)();
  var currentLevel = outdated.length ? level : 'notice';
  var severity = isSupported ? currentLevel : 'high';
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showDialog && /*#__PURE__*/_react.default.createElement("div", {
    role: "dialog",
    "aria-modal": "true",
    "aria-labelledby": elementId
  }, isUpToDate ? /*#__PURE__*/_react.default.createElement(_CurrentVersionsDialog.default, {
    onClose: onHideDialog,
    versions: versions
  }) : /*#__PURE__*/_react.default.createElement(_UpdateNotifierDialog.default, {
    onClose: onHideDialog,
    outdated: outdated,
    severity: severity
  })), !isUpToDate && /*#__PURE__*/_react.default.createElement(_default.default, {
    "aria-label": "".concat(formatUpdateLabel(outdated.length), ", ").concat(severity, " severity level."),
    icon: _packageIcon.default,
    iconStatus: "primary",
    id: elementId,
    kind: "simple",
    onClick: onShowDialog,
    padding: "small",
    selected: showDialog,
    tone: "navbar"
  }));
}
