// @sanity/default-layout/lib/navbar/studioStatus/UpdateNotifierDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _components = require("@sanity/base/components");

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var upperFirst = str => "".concat(str.slice(0, 1).toUpperCase()).concat(str.slice(1));

class UpdateNotifierDialog extends _react.default.PureComponent {
  renderTable() {
    var outdated = this.props.outdated;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.versionsTable
    }, /*#__PURE__*/_react.default.createElement("table", null, /*#__PURE__*/_react.default.createElement("thead", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("th", null, "Module"), /*#__PURE__*/_react.default.createElement("th", null, "Installed"), /*#__PURE__*/_react.default.createElement("th", null, "Latest"), /*#__PURE__*/_react.default.createElement("th", null, "Importance"))), /*#__PURE__*/_react.default.createElement("tbody", null, outdated.map(pkg => /*#__PURE__*/_react.default.createElement("tr", {
      key: pkg.name
    }, /*#__PURE__*/_react.default.createElement("td", {
      className: _UpdateNotifierDialog.default.npmValue
    }, pkg.name), /*#__PURE__*/_react.default.createElement("td", {
      className: _UpdateNotifierDialog.default.npmValue
    }, pkg.version), /*#__PURE__*/_react.default.createElement("td", {
      className: _UpdateNotifierDialog.default.npmValue
    }, pkg.latest), /*#__PURE__*/_react.default.createElement("td", null, upperFirst(pkg.severity || 'low'))))))), /*#__PURE__*/_react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.textContent
    }, /*#__PURE__*/_react.default.createElement("p", null, "To upgrade, run the ", /*#__PURE__*/_react.default.createElement("a", {
      href: "https://www.sanity.io/docs/reference/cli"
    }, "Sanity CLI"), ' ', "upgrade command in your project folder from a terminal."), /*#__PURE__*/_react.default.createElement("pre", {
      className: _UpdateNotifierDialog.default.code
    }, /*#__PURE__*/_react.default.createElement("code", null, "sanity upgrade"))));
  }

  renderContactDeveloper() {
    var severity = this.props.severity;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.textContent
    }, severity === 'high' ? /*#__PURE__*/_react.default.createElement("p", null, "This Studio should be updated. Please get in touch with the developers and ask them to upgrade it for you.") : /*#__PURE__*/_react.default.createElement("p", null, "This Studio has available upgrades. Consider getting in touch with the developers and ask them to upgrade it for you.")), /*#__PURE__*/_react.default.createElement("details", {
      className: _UpdateNotifierDialog.default.details
    }, /*#__PURE__*/_react.default.createElement("summary", {
      className: _UpdateNotifierDialog.default.summary
    }, "Developer info"), this.renderTable()));
  }

  render() {
    var _this$props = this.props,
        severity = _this$props.severity,
        onClose = _this$props.onClose;
    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarDialog"
    }, /*#__PURE__*/_react.default.createElement(_default2.default, {
      onClose: onClose,
      onClickOutside: onClose,
      title: severity === 'low' ? 'Upgrades available' : 'Studio is outdated'
    }, __DEV__ && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.textContent
    }, /*#__PURE__*/_react.default.createElement("p", null, "This Studio is no longer up to date", ' ', severity === 'high' ? 'and should be upgraded.' : 'and can be upgraded.')), this.renderTable()), !__DEV__ && this.renderContactDeveloper()));
  }

}

_defineProperty(UpdateNotifierDialog, "defaultProps", {
  outdated: []
});

var _default = UpdateNotifierDialog;
exports.default = _default;
