// @sanity/desk-tool/lib/panes/documentPane/documentPanel/header/tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentHeaderTabs = DocumentHeaderTabs;

var _react = _interopRequireWildcard(require("react"));

var _tab = _interopRequireDefault(require("part:@sanity/components/tabs/tab"));

var _tabList = _interopRequireDefault(require("part:@sanity/components/tabs/tab-list"));

var _tabs = _interopRequireDefault(require("./tabs.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function DocumentHeaderTabs(props) {
  var activeViewId = props.activeViewId,
      idPrefix = props.idPrefix,
      onSetActiveView = props.onSetActiveView,
      _props$views = props.views,
      views = _props$views === void 0 ? [] : _props$views;
  var tabPanelId = "".concat(idPrefix, "tabpanel");
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _tabs.default.headerTabsContainer
  }, /*#__PURE__*/_react.default.createElement(_tabList.default, null, views.map((view, index) => /*#__PURE__*/_react.default.createElement(DocumentHeaderTab, {
    icon: view.icon,
    id: "".concat(idPrefix, "tab-").concat(view.id),
    isActive: activeViewId === view.id,
    key: view.id,
    label: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, view.title),
    onSetView: onSetActiveView,
    tabPanelId: tabPanelId,
    viewId: index === 0 ? null : view.id
  }))));
}

function DocumentHeaderTab(props) {
  var handleClick = (0, _react.useCallback)(() => props.onSetView(props.viewId), [props.onSetView, props.viewId]);
  return /*#__PURE__*/_react.default.createElement(_tab.default, {
    icon: props.icon,
    id: props.id,
    isActive: props.isActive,
    label: props.label,
    onClick: handleClick,
    "aria-controls": props.tabPanelId
  });
}
