// @sanity/desk-tool/lib/panes/documentPane/documentPanel/header/validationMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidationMenu = ValidationMenu;

var _components = require("@sanity/base/components");

var _icons = require("@sanity/icons");

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _autoId = require("@reach/auto-id");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function ValidationMenu(props) {
  var boundaryElement = props.boundaryElement,
      isOpen = props.isOpen,
      markers = props.markers,
      schemaType = props.schemaType,
      setFocusPath = props.setFocusPath,
      setOpen = props.setOpen;
  var validationMarkers = markers.filter(marker => marker.type === 'validation');
  var validationErrorMarkers = validationMarkers.filter(marker => marker.level === 'error');
  var validationWarningwarnings = validationMarkers.filter(marker => marker.level === 'warning');
  var id = (0, _autoId.useId)();
  var handleClose = (0, _react.useCallback)(() => setOpen(false), [setOpen]);

  if (validationErrorMarkers.length === 0 && validationWarningwarnings.length === 0) {
    return null;
  }

  var popoverContent = /*#__PURE__*/_react.default.createElement(_components.ValidationList, {
    documentType: schemaType,
    markers: validationMarkers,
    onClose: handleClose,
    onFocus: setFocusPath
  });

  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    id: id || '',
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.ErrorOutlineIcon,
      title: "Show validation issues",
      mode: "bleed",
      tone: "critical"
    }),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, {
      open: isOpen
    }, popoverContent),
    popover: {
      portal: true,
      boundaryElement: boundaryElement,
      constrainSize: true,
      preventOverflow: true,
      width: 0
    },
    placement: "bottom-end"
  });
}
