// @sanity/base/lib/components/AppLoadingScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var AppLoadingScreenStyles = "\n.sanity-app-loading-screen {\n  font-family: system-ui, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen, Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;\n  text-align: center;\n  background-color: #F1F3F6;\n  color: #1C2430;\n  display: flex;\n  width: 100vw;\n  height: 100vh;\n  position: absolute;\n  top: 0;\n  left: 0;\n}\n\n.sanity-app-loading-screen__inner {\n  margin: auto;\n}\n\n.sanity-app-loading-screen__text {\n  margin-top: 2rem;\n  font-size: 13px;\n  font-weight: 600;\n}\n\n.sanity-app-loading-screen__loader {\n  display: block;\n  animation: pulse var(--time) cubic-bezier(.11,0,.27,1) infinite;\n  --time: 3s;\n}\n\n@keyframes pulse {\n  from {\n    transform: scale3d(0.75, 0.75, 0.75);\n  }\n  \n  50% {\n    transform: scale3d(1, 1, 1);\n  }\n  \n  to {\n    transform: scale3d(0.75, 0.75, 0.75);\n  }\n}\n";
var AppLoaderStyles = "\n  .fillShape {\n    stroke-width: 40;\n    stroke: #F03E2F;\n    opacity: 0;\n  }\n  \n  .fillShape--bottom {\n    stroke-dasharray: 90;\n    stroke-dashoffset: 85;\n    animation: bottom var(--time) ease-in infinite;\n  }\n  \n  .fillShape--middle {\n    stroke-dasharray: 115;\n    stroke-dashoffset: 110;\n    animation: middle var(--time) linear infinite;\n  }\n  \n  .fillShape--top {\n    stroke-dasharray: 77;\n    stroke-dashoffset: 72;\n    animation: top var(--time) ease-out infinite;\n  }\n  \n  @keyframes bottom {\n    0%,\n    85% {\n      stroke-dashoffset: 265;\n      opacity: 0;\n    }\n    15%,\n    64% {\n      stroke-dashoffset: 175;\n      opacity: 0.5;\n    }\n  }\n  @keyframes middle {\n    11%,\n    75% {\n      stroke-dashoffset: 100;\n      opacity: 0;\n    }\n    15% {\n      opacity: 1;\n    }\n    25%,\n    63% {\n      stroke-dashoffset: 225;\n      opacity: 1;\n    }\n  }\n  @keyframes top {\n    22%,\n    70% {\n      opacity: 0;\n      stroke-dashoffset: 226;\n    }\n    25% {\n      opacity: 0.5;\n    }\n    35%,\n    54% {\n      stroke-dashoffset: 149;\n      opacity: 0.5;\n    }\n  }\n";

class AppLoadingScreen extends _react.default.PureComponent {
  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "sanity-app-loading-screen"
    }, /*#__PURE__*/_react.default.createElement("style", {
      type: "text/css"
    }, AppLoadingScreenStyles), /*#__PURE__*/_react.default.createElement("div", {
      className: "sanity-app-loading-screen__inner"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "sanity-app-loading-screen__loader"
    }, /*#__PURE__*/_react.default.createElement("svg", {
      width: "73",
      height: "95",
      viewBox: "0 0 73 95",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      shapeRendering: "geometricPrecision"
    }, /*#__PURE__*/_react.default.createElement("style", null, AppLoaderStyles), /*#__PURE__*/_react.default.createElement("mask", {
      id: "bottom",
      "mask-type": "alpha",
      maskUnits: "userSpaceOnUse",
      x: "0",
      y: "62",
      width: "67",
      height: "33"
    }, /*#__PURE__*/_react.default.createElement("path", {
      d: "M56.4905 62.9807C63.6434 67.5637 66.7972 73.9744 66.7972 83.168C60.7987 90.7262 50.4577 95 38.2203 95C17.6963 95 3.17759 84.8102 0.0100098 67.282H19.7851C22.3343 75.3761 29.0748 79.1209 38.0622 79.1209C49.056 79.1209 56.3462 73.2598 56.4905 62.9807",
      fill: "white"
    })), /*#__PURE__*/_react.default.createElement("g", {
      mask: "url(#bottom)"
    }, /*#__PURE__*/_react.default.createElement("path", {
      className: "fillShape fillShape--bottom",
      d: "M8 59C9.35605 77.6466 25.4128 87.4032 37.5 87.5C54 87.6322 62 75 71.5 66.5"
    })), /*#__PURE__*/_react.default.createElement("mask", {
      id: "middle",
      "mask-type": "alpha",
      maskUnits: "userSpaceOnUse",
      x: "2",
      y: "12",
      width: "71",
      height: "72"
    }, /*#__PURE__*/_react.default.createElement("path", {
      d: "M7.73317 12.1206C7.73317 24.8871 15.6418 32.5484 31.4729 36.568L48.2521 40.4571C63.238 43.8926 72.3834 52.4472 72.3834 66.3818C72.493 72.4436 70.5187 78.3594 66.7904 83.1404C66.7904 69.2402 59.5963 61.7232 42.2605 57.2157L25.7698 53.471C12.5842 50.4683 2.40806 43.4323 2.40806 28.2952C2.33733 22.4523 4.21324 16.7521 7.74004 12.0931",
      fill: "white"
    })), /*#__PURE__*/_react.default.createElement("g", {
      mask: "url(#middle)"
    }, /*#__PURE__*/_react.default.createElement("path", {
      className: "fillShape fillShape--middle",
      d: "M-15.5 8.5C-15.5 12 -1 41.7047 38.3457 48C65.7344 52.3822 69 68 62.8457 74.5C56.6914 81 54.5 82 54.5 82"
    })), /*#__PURE__*/_react.default.createElement("mask", {
      id: "top",
      "mask-type": "alpha",
      maskUnits: "userSpaceOnUse",
      x: "7",
      y: "0",
      width: "65",
      height: "32"
    }, /*#__PURE__*/_react.default.createElement("path", {
      d: "M18.0122 31.1124C11.1411 26.7149 7.70557 20.524 7.70557 12.1138C13.4636 4.59677 23.3648 0 35.5198 0C56.4973 0 68.6317 11.09 71.6275 26.6737H52.5945C50.4988 20.5309 45.2562 15.7348 35.6641 15.7348C25.4399 15.7348 18.4863 21.6852 18.0122 31.1124",
      fill: "white"
    })), /*#__PURE__*/_react.default.createElement("g", {
      mask: "url(#top)"
    }, /*#__PURE__*/_react.default.createElement("path", {
      className: "fillShape fillShape--top",
      d: "M9 35C9 21 19 5.49993 40 8.49996C61 11.5 65.3456 29 65.3456 33.5"
    })))), /*#__PURE__*/_react.default.createElement("div", {
      className: "sanity-app-loading-screen__text"
    }, this.props.text)));
  }

}

exports.default = AppLoadingScreen;

_defineProperty(AppLoadingScreen, "propTypes", {
  text: _propTypes.default.string
});

_defineProperty(AppLoadingScreen, "defaultProps", {
  text: 'Loading Content Studio'
});
