// @sanity/base/lib/components/InsufficientPermissionsMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InsufficientPermissionsMessage = InsufficientPermissionsMessage;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function InsufficientPermissionsMessage(props) {
  var currentUser = props.currentUser,
      _props$title = props.title,
      title = _props$title === void 0 ? 'Insufficient permissions' : _props$title,
      _props$operationLabel = props.operationLabel,
      operationLabel = _props$operationLabel === void 0 ? 'access this feature' : _props$operationLabel;
  var roles = (currentUser === null || currentUser === void 0 ? void 0 : currentUser.roles) || [];
  var plural = roles.length !== 1;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.AccessDeniedIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "semibold"
  }, title)), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, "Your role", plural && 's', ' ', join(roles.map(r => /*#__PURE__*/_react.default.createElement("code", {
    key: r.name
  }, r.title)), ', '), ' ', "do", plural || 'es', " not have permissions to ", operationLabel)));
}

function join(array, sep) {
  return array.reduce((result, item) => result === null ? [item] : [...result, sep, item], null);
}
