// @sanity/base/lib/components/SanityRoot.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ui = require("@sanity/ui");

var _configSanity = _interopRequireDefault(require("config:sanity"));

var _root = _interopRequireDefault(require("part:@sanity/base/root"));

var _provider = _interopRequireDefault(require("part:@sanity/components/snackbar/provider"));

var _react = _interopRequireWildcard(require("react"));

var _reactRefractor = _interopRequireDefault(require("react-refractor"));

var _bash = _interopRequireDefault(require("refractor/lang/bash"));

var _javascript = _interopRequireDefault(require("refractor/lang/javascript"));

var _json = _interopRequireDefault(require("refractor/lang/json"));

var _jsx = _interopRequireDefault(require("refractor/lang/jsx"));

var _typescript = _interopRequireDefault(require("refractor/lang/typescript"));

var _sanityCssCustomProperties = _interopRequireDefault(require("sanity:css-custom-properties"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _requiredSanityUiVersion = require("../requiredSanityUiVersion");

var _theme = require("../theme");

var _userColor = require("../user-color");

var _ErrorHandler = _interopRequireDefault(require("./ErrorHandler"));

var _DevServerStatus = _interopRequireDefault(require("./DevServerStatus"));

var _GlobalStyle = require("./GlobalStyle");

var _MissingProjectConfig = _interopRequireDefault(require("./MissingProjectConfig"));

var _VersionChecker = _interopRequireDefault(require("./VersionChecker"));

var _zOffsets = require("./zOffsets");

var _templateObject;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

_reactRefractor.default.registerLanguage(_bash.default);

_reactRefractor.default.registerLanguage(_javascript.default);

_reactRefractor.default.registerLanguage(_json.default);

_reactRefractor.default.registerLanguage(_jsx.default);

_reactRefractor.default.registerLanguage(_typescript.default);

var Root = (0, _styledComponents.default)(_ui.Card).attrs({
  tone: 'transparent'
})(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 100%;\n"])));

function UIErrorMessage() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      WebkitFontSmoothing: 'antialiased',
      padding: 20,
      margin: '0 auto',
      maxWidth: '40rem'
    }
  }, /*#__PURE__*/_react.default.createElement("h1", null, "Error: Missing theme context"), /*#__PURE__*/_react.default.createElement("p", null, "This problem is usually caused by the existence of multiple versions of", ' ', /*#__PURE__*/_react.default.createElement("code", null, "@sanity/ui"), " in the same application."), /*#__PURE__*/_react.default.createElement("p", null, "Make sure to install the same version of ", /*#__PURE__*/_react.default.createElement("code", null, "@sanity/ui"), " as", ' ', /*#__PURE__*/_react.default.createElement("code", null, "@sanity/base"), ":"), /*#__PURE__*/_react.default.createElement("pre", {
    style: {
      padding: 20,
      background: _sanityCssCustomProperties.default['--black'],
      color: _sanityCssCustomProperties.default['--white']
    }
  }, /*#__PURE__*/_react.default.createElement("code", null, "npm install @sanity/ui@", _requiredSanityUiVersion.REQUIRED_UI_VERSION)));
}

function AppProvider() {
  var _useState = (0, _react.useState)(() => document.createElement('div')),
      _useState2 = _slicedToArray(_useState, 2),
      portalElement = _useState2[0],
      setPortalElement = _useState2[1];

  var _useState3 = (0, _react.useState)(null),
      _useState4 = _slicedToArray(_useState3, 2),
      uiError = _useState4[0],
      setUIError = _useState4[1];

  var zIndex = (0, _zOffsets.useZIndex)();

  try {
    (0, _ui.useRootTheme)();
  } catch (_) {
    return /*#__PURE__*/_react.default.createElement(UIErrorMessage, null);
  }

  if (uiError) {
    return /*#__PURE__*/_react.default.createElement(UIErrorMessage, null);
  }

  return /*#__PURE__*/_react.default.createElement(_userColor.UserColorManagerProvider, {
    manager: _userColor.userColorManager
  }, /*#__PURE__*/_react.default.createElement(_ui.PortalProvider, {
    element: portalElement
  }, /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, null, /*#__PURE__*/_react.default.createElement(_ui.ToastProvider, {
    paddingY: 7,
    zOffset: zIndex.toast
  }, /*#__PURE__*/_react.default.createElement(_provider.default, null, /*#__PURE__*/_react.default.createElement(_ui.ThemeColorProvider, {
    tone: "transparent"
  }, /*#__PURE__*/_react.default.createElement(_GlobalStyle.GlobalStyle, null)), /*#__PURE__*/_react.default.createElement(Root, null, /*#__PURE__*/_react.default.createElement(_DevServerStatus.default, null), /*#__PURE__*/_react.default.createElement(_ErrorHandler.default, {
    onUIError: setUIError
  }), /*#__PURE__*/_react.default.createElement(_root.default, null), /*#__PURE__*/_react.default.createElement(_VersionChecker.default, null), /*#__PURE__*/_react.default.createElement("div", {
    "data-portal": "",
    ref: setPortalElement
  })))))));
}

function SanityRoot() {
  var _ref = _configSanity.default.api || {},
      projectId = _ref.projectId,
      dataset = _ref.dataset;

  if (!projectId || !dataset) {
    return /*#__PURE__*/_react.default.createElement(_MissingProjectConfig.default, null);
  }

  return /*#__PURE__*/_react.default.createElement(_zOffsets.ZIndexProvider, null, /*#__PURE__*/_react.default.createElement(_ui.ThemeProvider, {
    theme: _theme.theme
  }, /*#__PURE__*/_react.default.createElement(AppProvider, null)));
}

var _default = SanityRoot;
exports.default = _default;
