// @sanity/ui/lib/esm/styles/box/boxStyle.js
import { getResponsiveProp, responsive } from "../helpers";
const BASE_STYLE = {
  '&[data-as="ul"],&[data-as="ol"]': {
    listStyle: "none"
  }
};
const BOX_SIZING = {
  content: "content-box",
  border: "border-box"
};
const BOX_HEIGHT = {
  stretch: "stretch",
  fill: "100%"
};
function boxStyle() {
  return BASE_STYLE;
}
function responsiveBoxStyle() {
  return [
    responsiveBoxSizingStyle,
    responsiveBoxHeightStyle,
    responsiveBoxOverflowStyle,
    responsiveBoxDisplayStyle
  ];
}
function responsiveBoxDisplayStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$display), (display) => ({
    "&:not([hidden])": { display }
  }));
}
function responsiveBoxSizingStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$sizing), (sizing) => ({
    boxSizing: BOX_SIZING[sizing]
  }));
}
function responsiveBoxHeightStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$height), (height) => ({
    height: BOX_HEIGHT[height]
  }));
}
function responsiveBoxOverflowStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$overflow), (overflow) => ({
    overflow
  }));
}
export {
  boxStyle,
  responsiveBoxStyle
};
