// @sanity/desk-tool/lib/panes/DeskToolPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _documentsListPane = require("./documentsListPane");

var _userComponentPane = require("./userComponentPane");

var _unknownPane = require("./unknownPane");

var _documentPane = require("./documentPane");

var _listPane = require("./listPane");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var paneMap = {
  list: _listPane.ListPane,
  documentList: _documentsListPane.DocumentsListPane,
  document: _documentPane.DocumentPaneProvider,
  component: _userComponentPane.UserComponentPane
}; // eslint-disable-next-line react/prefer-stateless-function

class DeskToolPane extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handlePaneCollapse", () => this.props.onCollapse(this.props.index));

    _defineProperty(this, "handlePaneExpand", () => this.props.onExpand(this.props.index));
  }

  render() {
    var type = this.props.type;
    var PaneComponent = paneMap[type] || _unknownPane.UnknownPane;
    return /*#__PURE__*/_react.default.createElement(PaneComponent, _extends({}, this.props, {
      onExpand: this.handlePaneExpand,
      onCollapse: this.handlePaneCollapse
    }));
  }

}

exports.default = DeskToolPane;

_defineProperty(DeskToolPane, "propTypes", {
  index: _propTypes.default.number,
  title: _propTypes.default.string,
  type: _propTypes.default.string.isRequired,
  onCollapse: _propTypes.default.func,
  onExpand: _propTypes.default.func
});

_defineProperty(DeskToolPane, "defaultProps", {
  title: '',
  index: 0,
  onCollapse: _noop2.default,
  onExpand: _noop2.default
});
