// @sanity/base/lib/__legacy/@sanity/components/utilities/Portal/Portal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Portal = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _Portal = _interopRequireDefault(require("./Portal.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement); // @todo: refactor to functional component

class Portal extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "node", void 0);
  }

  componentWillUnmount() {
    if (this.node) {
      document.body.removeChild(this.node);
    }

    this.node = undefined;
  }

  render() {
    if (!canUseDOM) {
      return null;
    }

    if (!this.node) {
      this.node = document.createElement('div');
      document.body.appendChild(this.node);
    }

    return /*#__PURE__*/(0, _reactDom.createPortal)( /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, this.props.children, /*#__PURE__*/_react.default.createElement("span", {
      tabIndex: 0,
      className: _Portal.default.captureTabFocus
    })), this.node);
  }

}

exports.Portal = Portal;
