// @sanity/base/lib/__legacy/@sanity/components/fileinput/Button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/fileinput/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class FileInputButton extends _react.default.PureComponent {
  render() {
    return /*#__PURE__*/_react.default.createElement(_default.default, null, /*#__PURE__*/_react.default.createElement("div", {
      style: {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        opacity: 0
      }
    }, /*#__PURE__*/_react.default.createElement(_default2.default, _extends({}, this.props, {
      style: {
        height: '100%',
        width: '100%',
        display: 'block'
      }
    }), this.props.children)), this.props.children);
  }

}

exports.default = FileInputButton;
