// @sanity/base/lib/__legacy/@sanity/components/fileinput/DefaultFileInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _DefaultFileInput = _interopRequireDefault(require("./DefaultFileInput.css"));

var _excluded = ["children", "style", "className", "onSelect"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// @todo: refactor to functional component
class DefaultFileInput extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_inputId", (0, _uniqueId2.default)('FileSelect'));

    _defineProperty(this, "handleSelect", event => {
      if (this.props.onSelect) {
        this.props.onSelect(event.target.files);
      }
    });
  }

  componentDidMount() {
    if ('value' in this.props) {
      // eslint-disable-next-line
      console.error(new Error('Warning: A `value` prop was passed to FileSelect. This is most likely not intended as this' + ' component is only for letting users select file(s) from their hard drive'));
    }
  }

  render() {
    var _this$props = this.props,
        children = _this$props.children,
        _this$props$style = _this$props.style,
        style = _this$props$style === void 0 ? {} : _this$props$style,
        className = _this$props.className,
        onSelect = _this$props.onSelect,
        rest = _objectWithoutProperties(_this$props, _excluded);

    return /*#__PURE__*/_react.default.createElement("label", {
      style: style,
      className: (0, _classnames.default)(_DefaultFileInput.default.root, className),
      htmlFor: this._inputId
    }, /*#__PURE__*/_react.default.createElement("input", _extends({}, rest, {
      type: "file",
      value: "",
      id: this._inputId,
      onChange: this.handleSelect,
      style: {
        overflow: 'hidden',
        width: '0.1px',
        height: '0.1px',
        opacity: 0,
        position: 'absolute',
        zIndex: -1
      }
    })), children);
  }

}

exports.default = DefaultFileInput;
