// @sanity/form-builder/lib/legacyParts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BooleanInput", {
  enumerable: true,
  get: function get() {
    return _boolean.default;
  }
});
Object.defineProperty(exports, "DateTimeInput", {
  enumerable: true,
  get: function get() {
    return _datetime.default;
  }
});
Object.defineProperty(exports, "EmailInput", {
  enumerable: true,
  get: function get() {
    return _email.default;
  }
});
Object.defineProperty(exports, "GeoPointInput", {
  enumerable: true,
  get: function get() {
    return _geopoint.default;
  }
});
Object.defineProperty(exports, "NumberInput", {
  enumerable: true,
  get: function get() {
    return _number.default;
  }
});
Object.defineProperty(exports, "ObjectInput", {
  enumerable: true,
  get: function get() {
    return _object.default;
  }
});
Object.defineProperty(exports, "ReferenceInput", {
  enumerable: true,
  get: function get() {
    return _reference.default;
  }
});
Object.defineProperty(exports, "StringInput", {
  enumerable: true,
  get: function get() {
    return _string.default;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function get() {
    return _text.default;
  }
});
Object.defineProperty(exports, "UrlInput", {
  enumerable: true,
  get: function get() {
    return _url.default;
  }
});
Object.defineProperty(exports, "schema", {
  enumerable: true,
  get: function get() {
    return _schema.default;
  }
});
Object.defineProperty(exports, "useRouter", {
  enumerable: true,
  get: function get() {
    return _router.useRouter;
  }
});
Object.defineProperty(exports, "IntentLink", {
  enumerable: true,
  get: function get() {
    return _router.IntentLink;
  }
});
Object.defineProperty(exports, "SearchableSelect", {
  enumerable: true,
  get: function get() {
    return _searchable.default;
  }
});
Object.defineProperty(exports, "CustomMarkers", {
  enumerable: true,
  get: function get() {
    return _blockMarkersCustomDefault.default;
  }
});
Object.defineProperty(exports, "BlockExtras", {
  enumerable: true,
  get: function get() {
    return _blockExtras.default;
  }
});
Object.defineProperty(exports, "Markers", {
  enumerable: true,
  get: function get() {
    return _blockMarkers.default;
  }
});
Object.defineProperty(exports, "client", {
  enumerable: true,
  get: function get() {
    return _client.default;
  }
});
Object.defineProperty(exports, "WithReferringDocuments", {
  enumerable: true,
  get: function get() {
    return _withReferringDocuments.WithReferringDocuments;
  }
});
Object.defineProperty(exports, "DefaultDialog", {
  enumerable: true,
  get: function get() {
    return _default.default;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function get() {
    return _default2.List;
  }
});
Object.defineProperty(exports, "Item", {
  enumerable: true,
  get: function get() {
    return _default2.Item;
  }
});
Object.defineProperty(exports, "createWeightedSearch", {
  enumerable: true,
  get: function get() {
    return _weighted.createWeightedSearch;
  }
});
Object.defineProperty(exports, "SanityPreview", {
  enumerable: true,
  get: function get() {
    return _preview.default;
  }
});
Object.defineProperty(exports, "observePaths", {
  enumerable: true,
  get: function get() {
    return _preview.observePaths;
  }
});
Object.defineProperty(exports, "observeForPreview", {
  enumerable: true,
  get: function get() {
    return _preview.observeForPreview;
  }
});
Object.defineProperty(exports, "ImageLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader.default;
  }
});
Object.defineProperty(exports, "MenuButton", {
  enumerable: true,
  get: function get() {
    return _menuButton.MenuButton;
  }
});
Object.defineProperty(exports, "SnackbarPart", {
  enumerable: true,
  get: function get() {
    return _default3.default;
  }
});
Object.defineProperty(exports, "DropDownButton", {
  enumerable: true,
  get: function get() {
    return _dropdown.default;
  }
});
Object.defineProperty(exports, "StyleSelect", {
  enumerable: true,
  get: function get() {
    return _style.default;
  }
});
Object.defineProperty(exports, "ButtonGrid", {
  enumerable: true,
  get: function get() {
    return _buttonGrid.default;
  }
});
Object.defineProperty(exports, "formBuilderConfig", {
  enumerable: true,
  get: function get() {
    return _formBuilder.default;
  }
});
Object.defineProperty(exports, "userDefinedImageAssetSources", {
  enumerable: true,
  get: function get() {
    return _assetSources.default;
  }
});
Object.defineProperty(exports, "userDefinedFileAssetSources", {
  enumerable: true,
  get: function get() {
    return _assetSources2.default;
  }
});
Object.defineProperty(exports, "FullscreenDialog", {
  enumerable: true,
  get: function get() {
    return _fullscreen.default;
  }
});
Object.defineProperty(exports, "ValidationStatusPart", {
  enumerable: true,
  get: function get() {
    return _status.default;
  }
});
Object.defineProperty(exports, "PopoverDialogPart", {
  enumerable: true,
  get: function get() {
    return _popover.default;
  }
});
Object.defineProperty(exports, "EditItemFoldPart", {
  enumerable: true,
  get: function get() {
    return _fold.default;
  }
});
Object.defineProperty(exports, "documentStore", {
  enumerable: true,
  get: function get() {
    return _document.default;
  }
});
Object.defineProperty(exports, "FieldsetPart", {
  enumerable: true,
  get: function get() {
    return _default4.default;
  }
});
Object.defineProperty(exports, "ActivateOnFocusPart", {
  enumerable: true,
  get: function get() {
    return _activateOnFocus.default;
  }
});
Object.defineProperty(exports, "FormFieldPart", {
  enumerable: true,
  get: function get() {
    return _default5.default;
  }
});
Object.defineProperty(exports, "ProgressCirclePart", {
  enumerable: true,
  get: function get() {
    return _circle.default;
  }
});
Object.defineProperty(exports, "defaultImageAssetSources", {
  enumerable: true,
  get: function get() {
    return _assetSource.default;
  }
});
Object.defineProperty(exports, "defaultFileAssetSources", {
  enumerable: true,
  get: function get() {
    return _assetSource2.default;
  }
});

var _boolean = _interopRequireDefault(require("part:@sanity/form-builder/input/boolean?"));

var _datetime = _interopRequireDefault(require("part:@sanity/form-builder/input/datetime?"));

var _email = _interopRequireDefault(require("part:@sanity/form-builder/input/email?"));

var _geopoint = _interopRequireDefault(require("part:@sanity/form-builder/input/geopoint?"));

var _number = _interopRequireDefault(require("part:@sanity/form-builder/input/number?"));

var _object = _interopRequireDefault(require("part:@sanity/form-builder/input/object?"));

var _reference = _interopRequireDefault(require("part:@sanity/form-builder/input/reference?"));

var _string = _interopRequireDefault(require("part:@sanity/form-builder/input/string?"));

var _text = _interopRequireDefault(require("part:@sanity/form-builder/input/text?"));

var _url = _interopRequireDefault(require("part:@sanity/form-builder/input/url?"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _router = require("@sanity/base/router");

var _searchable = _interopRequireDefault(require("part:@sanity/components/selects/searchable"));

var _blockMarkersCustomDefault = _interopRequireDefault(require("part:@sanity/form-builder/input/block-editor/block-markers-custom-default"));

var _blockExtras = _interopRequireDefault(require("part:@sanity/form-builder/input/block-editor/block-extras"));

var _blockMarkers = _interopRequireDefault(require("part:@sanity/form-builder/input/block-editor/block-markers"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _withReferringDocuments = require("part:@sanity/base/with-referring-documents");

var _default = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _default2 = require("part:@sanity/components/lists/default");

var _weighted = require("part:@sanity/base/search/weighted");

var _preview = _interopRequireWildcard(require("part:@sanity/base/preview"));

var _imageLoader = _interopRequireDefault(require("part:@sanity/components/utilities/image-loader"));

var _menuButton = require("part:@sanity/components/menu-button");

var _default3 = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _style = _interopRequireDefault(require("part:@sanity/components/selects/style"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _formBuilder = _interopRequireDefault(require("config:@sanity/form-builder"));

var _assetSources = _interopRequireDefault(require("part:@sanity/form-builder/input/image/asset-sources?"));

var _assetSources2 = _interopRequireDefault(require("part:@sanity/form-builder/input/file/asset-sources?"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _status = _interopRequireDefault(require("part:@sanity/components/validation/status"));

var _popover = _interopRequireDefault(require("part:@sanity/components/dialogs/popover"));

var _fold = _interopRequireDefault(require("part:@sanity/components/edititem/fold"));

var _document = _interopRequireDefault(require("part:@sanity/base/datastore/document"));

var _default4 = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _activateOnFocus = _interopRequireDefault(require("part:@sanity/components/utilities/activate-on-focus"));

var _default5 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _circle = _interopRequireDefault(require("part:@sanity/components/progress/circle"));

var _assetSource = _interopRequireDefault(require("all:part:@sanity/form-builder/input/image/asset-source"));

var _assetSource2 = _interopRequireDefault(require("all:part:@sanity/form-builder/input/file/asset-source"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
