// @sanity/desk-tool/lib/utils/parsePanesSegment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parsePanesSegment = parsePanesSegment;
exports.encodePanesSegment = encodePanesSegment;
exports.parseOldPanesSegment = parseOldPanesSegment;

var _constants = require("../constants");

var _PaneRouterContext = require("../contexts/PaneRouterContext");

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _toArray(arr) { return _arrayWithHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// old: authors;knut,{"template":"diaryEntry"}
// new: authors;knut,view=diff,eyJyZXYxIjoiYWJjMTIzIiwicmV2MiI6ImRlZjQ1NiJ9|latest-posts
var panePattern = /^([.a-z0-9_-]+),?({.*?})?(?:(;|$))/i;

var isParam = str => /^[a-z0-9]+=[^=]+/i.test(str);

var isPayload = str => /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(str);

function parseChunks(chunks) {
  var initial = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return chunks.reduce((pane, chunk) => {
    if (isParam(chunk)) {
      var key = chunk.slice(0, chunk.indexOf('='));
      var value = chunk.slice(key.length + 1);
      pane.params = _objectSpread(_objectSpread({}, pane.params), {}, {
        [key]: value
      });
    } else if (isPayload(chunk)) {
      pane.payload = tryParseBase64Payload(chunk);
    } else {
      // eslint-disable-next-line no-console
      console.warn('Unknown pane segment: %s - skipping', chunk);
    }

    return pane;
  }, _objectSpread(_objectSpread({}, initial), {}, {
    params: _constants.EMPTY_PARAMS,
    payload: undefined
  }));
}

function encodeChunks(pane, i, group) {
  var payload = pane.payload,
      _pane$params = pane.params,
      params = _pane$params === void 0 ? {} : _pane$params,
      id = pane.id;
  var sameAsFirst = i !== 0 && id === group[0].id;
  var encodedPayload = typeof payload === 'undefined' ? undefined : btoa(JSON.stringify(payload));
  var encodedParams = Object.keys(params).reduce((pairs, key) => {
    if (sameAsFirst && i !== 0 && !_PaneRouterContext.exclusiveParams.includes(key) && group[0].params[key] === params[key]) {
      return pairs;
    }

    return params[key] ? [...pairs, "".concat(key, "=").concat(params[key])] : pairs;
  }, []);
  return [sameAsFirst ? '' : id].concat([encodedParams.length > 0 && encodedParams, encodedPayload].filter(Boolean)).join(',') || ',';
}

function parsePanesSegment(str) {
  if (str.indexOf(',{') !== -1) {
    return parseOldPanesSegment(str);
  }

  return str.split(';').map(group => group.split('|').map(segment => {
    var _segment$split = segment.split(','),
        _segment$split2 = _toArray(_segment$split),
        id = _segment$split2[0],
        chunks = _segment$split2.slice(1);

    return parseChunks(chunks, {
      id
    });
  }).map((pane, i, siblings) => pane.id ? pane : _objectSpread(_objectSpread({}, pane), {}, {
    id: siblings[0].id
  }))).filter(group => group.length > 0);
}

function encodePanesSegment(panes) {
  return (panes || []).map(group => group.map(encodeChunks).join('|')).map(encodeURIComponent).join(';');
}

function parseOldPanesSegment(str) {
  var chunks = [];
  var buffer = str;

  while (buffer.length) {
    var _ref = buffer.match(panePattern) || [],
        _ref2 = _slicedToArray(_ref, 3),
        match = _ref2[0],
        id = _ref2[1],
        payloadChunk = _ref2[2];

    if (!match) {
      buffer = buffer.slice(1);
      continue;
    }

    var payload = payloadChunk && tryParsePayload(payloadChunk);
    chunks.push({
      id,
      payload
    });
    buffer = buffer.slice(match.length);
  }

  return chunks;
}

function tryParsePayload(json) {
  try {
    return JSON.parse(json);
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn("Failed to parse parameters: ".concat(err.message));
    return undefined;
  }
}

function tryParseBase64Payload(data) {
  return data ? tryParsePayload(atob(data)) : undefined;
}
