// @sanity/ui/lib/esm/styles/flex/flexStyle.js
import { getResponsiveProp, rem, responsive } from "../helpers";
const BASE_STYLE = {
  "&&:not([hidden])": {
    display: "flex"
  }
};
function responsiveFlexStyle() {
  return [
    BASE_STYLE,
    responsiveFlexAlignStyle,
    responsiveFlexGapStyle,
    responsiveFlexWrapStyle,
    responsiveFlexJustifyStyle,
    responsiveFlexDirectionStyle
  ];
}
function responsiveFlexAlignStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$align), (align) => {
    return { alignItems: align };
  });
}
function responsiveFlexGapStyle(props) {
  const { theme } = props;
  const { media, space } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$gap), (gap) => ({
    gap: gap ? rem(space[gap]) : void 0
  }));
}
function responsiveFlexWrapStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$wrap), (wrap) => {
    return { flexWrap: wrap };
  });
}
function responsiveFlexJustifyStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$justify), (justify) => {
    return { justifyContent: justify };
  });
}
function responsiveFlexDirectionStyle(props) {
  const { theme } = props;
  const { media } = theme.sanity;
  return responsive(media, getResponsiveProp(props.$direction), (direction) => {
    return { flexDirection: direction };
  });
}
export {
  responsiveFlexAlignStyle,
  responsiveFlexDirectionStyle,
  responsiveFlexJustifyStyle,
  responsiveFlexStyle,
  responsiveFlexWrapStyle
};
