// @sanity/vision/lib/SanityVision.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema?"));

var _Select = _interopRequireDefault(require("./sanity/Select"));

var _Vision = _interopRequireDefault(require("./Vision"));

var _visionGui = _interopRequireDefault(require("./css/visionGui.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var components = {
  Button: _default2.default,
  Select: _Select.default
};
var styles = {
  visionGui: _visionGui.default
};

var client = _client.default.withConfig({
  apiVersion: '1'
}); // Used in Sanity project


function SanityVision() {
  return /*#__PURE__*/_react.default.createElement(_Vision.default, {
    styles: styles,
    components: components,
    client: client,
    schema: _schema.default
  });
}

var _default = SanityVision;
exports.default = _default;
