// @sanity/field/lib/paths/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pathToString = pathToString;
exports.getValueAtPath = getValueAtPath;
exports.findIndex = findIndex;
exports.stringToPath = stringToPath;
exports.normalizePathSegment = normalizePathSegment;
exports.normalizeIndexSegment = normalizeIndexSegment;
exports.normalizeKeySegment = normalizeKeySegment;
exports.normalizeIndexTupleSegment = normalizeIndexTupleSegment;
exports.pathsAreEqual = pathsAreEqual;
exports.getItemKey = getItemKey;
exports.getItemKeySegment = getItemKeySegment;
exports.isEmptyObject = isEmptyObject;

var _types = require("@sanity/types");

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;
var reKeySegment = /_key\s*==\s*['"](.*)['"]/;

function pathToString(path) {
  if (!Array.isArray(path)) {
    throw new Error('Path is not an array');
  }

  return path.reduce((target, segment, i) => {
    var segmentType = typeof segment;

    if (segmentType === 'number') {
      return "".concat(target, "[").concat(segment, "]");
    }

    if (segmentType === 'string') {
      var separator = i === 0 ? '' : '.';
      return "".concat(target).concat(separator).concat(segment);
    }

    if ((0, _types.isKeySegment)(segment) && segment._key) {
      return "".concat(target, "[_key==\"").concat(segment._key, "\"]");
    }

    if (Array.isArray(segment)) {
      var _segment = _slicedToArray(segment, 2),
          from = _segment[0],
          to = _segment[1];

      return "".concat(target, "[").concat(from, ":").concat(to, "]");
    }

    throw new Error("Unsupported path segment `".concat(JSON.stringify(segment), "`"));
  }, '');
}

function getValueAtPath(rootValue, path) {
  var segment = path[0];

  if (!segment) {
    return rootValue;
  }

  var tail = path.slice(1);

  if ((0, _types.isIndexSegment)(segment)) {
    return getValueAtPath(Array.isArray(rootValue) ? rootValue[segment] : undefined, tail);
  }

  if ((0, _types.isKeySegment)(segment)) {
    return getValueAtPath(Array.isArray(rootValue) ? rootValue.find(item => item._key === segment._key) : undefined, tail);
  }

  if (typeof segment === 'string') {
    return getValueAtPath(typeof rootValue === 'object' && rootValue !== null ? rootValue[segment] : undefined, tail);
  }

  throw new Error("Unknown segment type ".concat(JSON.stringify(segment)));
}

function findIndex(array, segment) {
  if (typeof segment !== 'number' && !(0, _types.isKeySegment)(segment)) {
    return -1;
  }

  return typeof segment === 'number' ? segment : array.findIndex(item => (0, _types.isKeyedObject)(item) && item._key === segment._key);
}

function stringToPath(path) {
  var segments = path.match(rePropName);

  if (!segments) {
    throw new Error('Invalid path string');
  }

  return segments.map(normalizePathSegment);
}

function normalizePathSegment(segment) {
  if ((0, _types.isIndexSegment)(segment)) {
    return normalizeIndexSegment(segment);
  }

  if ((0, _types.isKeySegment)(segment)) {
    return normalizeKeySegment(segment);
  }

  if ((0, _types.isIndexTuple)(segment)) {
    return normalizeIndexTupleSegment(segment);
  }

  return segment;
}

function normalizeIndexSegment(segment) {
  return Number(segment.replace(/[^\d]/g, ''));
}

function normalizeKeySegment(segment) {
  var segments = segment.match(reKeySegment);

  if (!segments) {
    throw new Error('Invalid key segment');
  }

  return {
    _key: segments[1]
  };
}

function normalizeIndexTupleSegment(segment) {
  var _segment$split$map = segment.split(':').map(seg => seg === '' ? seg : Number(seg)),
      _segment$split$map2 = _slicedToArray(_segment$split$map, 2),
      from = _segment$split$map2[0],
      to = _segment$split$map2[1];

  return [from, to];
}

function pathsAreEqual(pathA, pathB) {
  if (pathA.length !== pathB.length) {
    return false;
  }

  return pathA.every((segmentA, index) => {
    var segmentB = pathB[index];

    if ((0, _types.isKeySegment)(segmentA) && (0, _types.isKeySegment)(segmentB)) {
      return segmentA._key === segmentB._key;
    }

    if ((0, _types.isIndexSegment)(segmentA)) {
      return Number(segmentA) === Number(segmentB);
    }

    if ((0, _types.isIndexTuple)(segmentA) && (0, _types.isIndexTuple)(segmentB)) {
      return segmentA[0] === segmentB[0] && segmentA[1] === segmentB[1];
    }

    return segmentA === segmentB;
  });
}

function getItemKey(arrayItem) {
  return (0, _types.isKeyedObject)(arrayItem) ? arrayItem._key : undefined;
}

function getItemKeySegment(arrayItem) {
  var key = getItemKey(arrayItem);
  return key ? {
    _key: key
  } : undefined;
}

function isEmptyObject(item) {
  return typeof item === 'object' && item !== null && Object.keys(item).length <= 0;
}
