// @sanity/base/lib/__legacy/@sanity/components/autocomplete/DefaultAutocomplete.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _defaultStyle = _interopRequireDefault(require("part:@sanity/components/autocomplete/default-style"));

var _default = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _default2 = require("part:@sanity/components/lists/default");

var _default3 = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// @todo: refactor to functional component
class DefaultAutocomplete extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_inputId", undefined);

    _defineProperty(this, "handleChange", event => {
      if (this.props.onChange) this.props.onChange(event);
    });
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    this._inputId = this.props.id || (0, _uniqueId2.default)('Autocomplete');
  }

  render() {
    var _this$props = this.props,
        _this$props$suggestio = _this$props.suggestions,
        suggestions = _this$props$suggestio === void 0 ? [] : _this$props$suggestio,
        label = _this$props.label,
        placeholder = _this$props.placeholder,
        _this$props$value = _this$props.value,
        value = _this$props$value === void 0 ? '' : _this$props$value;
    var isOpen = suggestions.length > 1;
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      className: "".concat(isOpen ? _defaultStyle.default.opened : _defaultStyle.default.closed),
      label: label,
      labelFor: this._inputId
    }, /*#__PURE__*/_react.default.createElement(_default3.default, {
      inputId: this._inputId,
      value: value,
      className: _defaultStyle.default.textField,
      placeholder: placeholder,
      onChange: this.handleChange
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: _defaultStyle.default.input
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: _defaultStyle.default.suggestionsContainer
    }, /*#__PURE__*/_react.default.createElement(_default2.List, {
      className: _defaultStyle.default.suggestions
    })));
  }

}

exports.default = DefaultAutocomplete;
