// @sanity/field/lib/types/image/diff/HotspotCropSVG.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HotspotCropSVG = HotspotCropSVG;

var _react = _interopRequireDefault(require("react"));

var _diff = require("../../../diff");

var _helpers = require("./helpers");

var _excluded = ["crop", "diff", "hash", "hotspot", "width", "height"],
    _excluded2 = ["crop", "width", "height"],
    _excluded3 = ["hotspot", "offset", "width", "height"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function HotspotCropSVG(props) {
  var crop = props.crop,
      diff = props.diff,
      hash = props.hash,
      hotspot = props.hotspot,
      _props$width = props.width,
      width = _props$width === void 0 ? 100 : _props$width,
      _props$height = props.height,
      height = _props$height === void 0 ? 100 : _props$height,
      restProps = _objectWithoutProperties(props, _excluded);

  var cropColor = (0, _diff.useDiffAnnotationColor)(diff, 'crop');
  var hotspotColor = (0, _diff.useDiffAnnotationColor)(diff, 'hotspot');
  return /*#__PURE__*/_react.default.createElement("svg", _extends({}, restProps, {
    fill: "none",
    width: width,
    height: height,
    viewBox: "0 0 ".concat(width, " ").concat(height)
  }), /*#__PURE__*/_react.default.createElement("defs", null, crop && hotspot && /*#__PURE__*/_react.default.createElement("mask", {
    id: "mask-hotspot-".concat(hash)
  }, /*#__PURE__*/_react.default.createElement("rect", {
    x: 0,
    y: 0,
    width: width,
    height: height,
    fill: "#fff"
  }), /*#__PURE__*/_react.default.createElement(HotspotSVG, {
    hotspot: hotspot,
    fill: "#000",
    offset: 1,
    width: width,
    height: height,
    stroke: "#000",
    strokeWidth: 3
  }))), crop && /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff,
    path: "crop",
    description: "Crop changed"
  }, /*#__PURE__*/_react.default.createElement("g", null, /*#__PURE__*/_react.default.createElement(CropSVG, {
    crop: crop,
    fill: (0, _helpers.hexToRgba)(cropColor.border, 0.25),
    mask: hotspot ? "url(#mask-hotspot-".concat(hash, ")") : undefined,
    stroke: cropColor.border,
    strokeWidth: 1,
    width: width,
    height: height
  }))), hotspot && /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff,
    path: "hotspot",
    description: "Hotspot changed"
  }, /*#__PURE__*/_react.default.createElement("g", null, /*#__PURE__*/_react.default.createElement(HotspotSVG, {
    hotspot: hotspot,
    fill: (0, _helpers.hexToRgba)(hotspotColor.border, 0.25),
    stroke: hotspotColor.border,
    strokeWidth: 1,
    width: width,
    height: height
  }))));
}

function CropSVG(_ref) {
  var crop = _ref.crop,
      width = _ref.width,
      height = _ref.height,
      restProps = _objectWithoutProperties(_ref, _excluded2);

  var rectProps = {
    x: crop.left * width,
    y: crop.top * height,
    width: (1 - crop.right - crop.left) * width,
    height: (1 - crop.bottom - crop.top) * height
  };
  return /*#__PURE__*/_react.default.createElement("rect", _extends({}, restProps, rectProps, {
    style: {
      vectorEffect: 'non-scaling-stroke'
    }
  }));
}

function HotspotSVG(_ref2) {
  var hotspot = _ref2.hotspot,
      _ref2$offset = _ref2.offset,
      offset = _ref2$offset === void 0 ? 0 : _ref2$offset,
      width = _ref2.width,
      height = _ref2.height,
      restProps = _objectWithoutProperties(_ref2, _excluded3);

  var ellipseProps = {
    cx: hotspot.x * width,
    cy: hotspot.y * height,
    rx: hotspot.width / 2 * width + offset,
    ry: hotspot.height / 2 * height + offset
  };
  return /*#__PURE__*/_react.default.createElement("ellipse", _extends({}, restProps, ellipseProps, {
    style: {
      vectorEffect: 'non-scaling-stroke'
    }
  }));
}
