// @sanity/default-layout/lib/navbar/loginStatus/LoginStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _components = require("@sanity/base/components");

var _chevronDownIcon = _interopRequireDefault(require("part:@sanity/base/chevron-down-icon"));

var _signOutIcon = _interopRequireDefault(require("part:@sanity/base/sign-out-icon"));

var _clickOutside = require("part:@sanity/components/click-outside");

var _default = _interopRequireDefault(require("part:@sanity/components/menus/default"));

var _popover = require("part:@sanity/components/popover");

var _escapable = _interopRequireDefault(require("part:@sanity/components/utilities/escapable"));

var _react = _interopRequireDefault(require("react"));

var _LoginStatus = _interopRequireDefault(require("./LoginStatus.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class LoginStatus extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      isOpen: false
    });

    _defineProperty(this, "handleUserMenuItemClick", item => {
      if (item.action === 'signOut') {
        this.props.onLogout();
      }

      this.setState({
        isOpen: false
      });
    });

    _defineProperty(this, "handleButtonClick", () => {
      this.setState(state => ({
        isOpen: !state.isOpen
      }));
    });

    _defineProperty(this, "handleClose", () => {
      this.setState({
        isOpen: false
      });
    });
  }

  render() {
    var menuItems = [{
      title: "Sign out",
      icon: _signOutIcon.default,
      action: 'signOut'
    }];

    var popoverContent = /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.menuWrapper
    }, /*#__PURE__*/_react.default.createElement(_default.default, {
      onAction: this.handleUserMenuItemClick,
      items: menuItems
    }));

    return /*#__PURE__*/_react.default.createElement(_clickOutside.ClickOutside, {
      onClickOutside: this.handleClose
    }, ref => /*#__PURE__*/_react.default.createElement("button", {
      className: _LoginStatus.default.root,
      onClick: this.handleButtonClick,
      ref: ref,
      title: "Toggle user menu",
      type: "button"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.inner,
      tabIndex: -1
    }, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarPopover"
    }, /*#__PURE__*/_react.default.createElement(_popover.Popover, {
      content: popoverContent,
      open: this.state.isOpen,
      placement: "bottom-end"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.avatarContainer
    }, /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
      size: "medium",
      tone: "navbar",
      userId: "me"
    })))), /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.iconContainer
    }, /*#__PURE__*/_react.default.createElement(_chevronDownIcon.default, null))), this.state.isOpen && /*#__PURE__*/_react.default.createElement(_escapable.default, {
      onEscape: this.handleClose
    })));
  }

}

exports.default = LoginStatus;
