// @sanity/vision/lib/containers/VisionContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _DelayedSpinner = _interopRequireDefault(require("../components/DelayedSpinner"));

var _ErrorDialog = _interopRequireDefault(require("../components/ErrorDialog"));

var _VisionGui = _interopRequireDefault(require("../components/VisionGui"));

var _LoadingContainer = _interopRequireDefault(require("./LoadingContainer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

// Loads the most basic data from a Sanity project
class VisionContainer extends _LoadingContainer.default {
  getSubscriptions() {
    return {
      datasets: {
        uri: '/datasets'
      }
    };
  }

  render() {
    var datasets = this.state.datasets || [];

    if (this.state.error) {
      var defaultDataset = this.context.client.config().dataset;
      datasets[0] = {
        name: defaultDataset
      };
    } else if (!this.hasAllData()) {
      return /*#__PURE__*/_react.default.createElement(_DelayedSpinner.default, null);
    }

    return /*#__PURE__*/_react.default.createElement(_VisionGui.default, _extends({}, this.state, this.props, {
      datasets: datasets
    }));
  }

}

var _default = VisionContainer;
exports.default = _default;
