// @sanity/form-builder/lib/components/tagInput/tagInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TagInput = void 0;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _styles = require("./styles");

var _excluded = ["disabled", "onChange", "onFocus", "placeholder", "readOnly", "value"];

var _templateObject;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Root = (0, _styledComponents.default)(_ui.Box)(props => {
  var theme = props.theme;
  var _theme$sanity = theme.sanity,
      focusRing = _theme$sanity.focusRing,
      input = _theme$sanity.input,
      radius = _theme$sanity.radius;
  var color = theme.sanity.color.input;
  var space = (0, _ui.rem)(theme.sanity.space[1]);
  return {
    position: 'relative',
    borderRadius: "".concat(radius[1], "px"),
    color: color.default.enabled.fg,
    backgroundColor: color.default.enabled.bg,
    boxShadow: (0, _styles.focusRingBorderStyle)({
      color: color.default.enabled.border,
      width: input.border.width
    }),
    '& > .content': {
      position: 'relative',
      lineHeight: 0,
      margin: "-".concat(space, " 0 0 -").concat(space)
    },
    '& > .content > div': {
      display: 'inline-block',
      verticalAlign: 'top',
      padding: "".concat(space, " 0 0 ").concat(space)
    },
    // enabled
    '&:not([data-read-only])': {
      cursor: 'text'
    },
    // hovered
    '@media(hover:hover)': {
      '&:not([data-disabled]):not([data-read-only]):hover': {
        borderColor: color.default.hovered.border
      }
    },
    // focused
    '&:not([data-disabled]):not([data-read-only])[data-focused]': {
      boxShadow: (0, _styles.focusRingStyle)({
        border: {
          color: color.default.enabled.border,
          width: input.border.width
        },
        focusRing
      })
    },
    // disabled
    '*:disabled + &': {
      color: color.default.disabled.fg,
      backgroundColor: color.default.disabled.bg,
      boxShadow: (0, _styles.focusRingBorderStyle)({
        color: color.default.disabled.border,
        width: input.border.width
      })
    }
  };
});

var Input = _styledComponents.default.input(props => {
  var theme = props.theme;
  var font = theme.sanity.fonts.text;
  var color = theme.sanity.color.input;
  var p = theme.sanity.space[2];
  var size = theme.sanity.fonts.text.sizes[2];
  return {
    appearance: 'none',
    background: 'none',
    border: 0,
    borderRadius: 0,
    outline: 'none',
    fontSize: (0, _ui.rem)(size.fontSize),
    lineHeight: size.lineHeight / size.fontSize,
    fontFamily: font.family,
    fontWeight: font.weights.regular,
    margin: 0,
    display: 'block',
    minWidth: '1px',
    maxWidth: '100%',
    boxSizing: 'border-box',
    paddingTop: (0, _ui.rem)(p - size.ascenderHeight),
    paddingRight: (0, _ui.rem)(p),
    paddingBottom: (0, _ui.rem)(p - size.descenderHeight),
    paddingLeft: (0, _ui.rem)(p),
    // enabled
    '&:not(:invalid):not(:disabled)': {
      color: color.default.enabled.fg
    },
    // disabled
    '&:not(:invalid):disabled': {
      color: color.default.disabled.fg
    }
  };
});

var Placeholder = (0, _styledComponents.default)(_ui.Box)(props => {
  var theme = props.theme;
  var color = theme.sanity.color.input;
  return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    position: absolute;\n    top: 0;\n    left: 0;\n    right: 0;\n    pointer-events: none;\n    --card-fg-color: ", ";\n  "])), color.default.enabled.placeholder);
});
var TagInput = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  var disabled = props.disabled,
      onChange = props.onChange,
      onFocus = props.onFocus,
      placeholderProp = props.placeholder,
      readOnly = props.readOnly,
      _props$value = props.value,
      value = _props$value === void 0 ? [] : _props$value,
      restProps = _objectWithoutProperties(props, _excluded);

  var placeholder = (0, _react.useMemo)(() => {
    if (placeholderProp) return placeholderProp;

    if (typeof window !== 'undefined' && 'ontouchstart' in window) {
      return 'Enter tag…';
    }

    return 'Enter tag and press ENTER…';
  }, [placeholderProp]);

  var _useState = (0, _react.useState)(''),
      _useState2 = _slicedToArray(_useState, 2),
      inputValue = _useState2[0],
      setInputValue = _useState2[1];

  var enabled = !disabled && !readOnly;

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      focused = _useState4[0],
      setFocused = _useState4[1];

  var forwardedRef = (0, _ui.useForwardedRef)(ref);
  var rootRef = (0, _react.useRef)(null);
  var handleRootPointerDown = (0, _react.useCallback)(event => {
    var isTagElement = (0, _ui.isHTMLElement)(event.target) && event.target.closest('[data-ui="Tag"]');
    if (isTagElement) return;
    var inputElement = forwardedRef.current;

    if (inputElement) {
      setTimeout(() => inputElement.focus(), 0);
    }
  }, [forwardedRef]);
  var handleInputBlur = (0, _react.useCallback)(() => {
    setFocused(false);
  }, []);
  var handleInputChange = (0, _react.useCallback)(event => {
    setInputValue(event.currentTarget.value);
  }, []);
  var handleInputFocus = (0, _react.useCallback)(() => {
    setFocused(true);
    if (onFocus) onFocus();
  }, [onFocus]);
  var handleInputKeyDown = (0, _react.useCallback)(event => {
    if (event.key === 'Enter') {
      event.preventDefault();
      event.stopPropagation();

      if (onChange && inputValue) {
        var _newValue = value.concat([{
          value: inputValue
        }]);

        setInputValue('');
        if (onChange) onChange(_newValue);
      }
    }
  }, [inputValue, onChange, value]);
  var handleTagRemove = (0, _react.useCallback)(index => {
    if (!onChange) return;
    var newValue = value.slice(0);
    newValue.splice(index, 1);
    onChange(newValue);
  }, [onChange, value]);
  (0, _react.useEffect)(() => {
    var inputElement = forwardedRef.current;

    if (inputElement) {
      inputElement.style.width = '0';
      inputElement.style.width = "".concat(inputElement.scrollWidth, "px");
    }
  }, [forwardedRef, inputValue]);
  return /*#__PURE__*/_react.default.createElement(Root, {
    "data-disabled": disabled ? '' : undefined,
    "data-focused": focused ? '' : undefined,
    "data-read-only": readOnly ? '' : undefined,
    "data-ui": "TagInput",
    onPointerDown: handleRootPointerDown,
    padding: 1,
    ref: rootRef
  }, enabled && /*#__PURE__*/_react.default.createElement(Placeholder, {
    hidden: Boolean(inputValue || value.length),
    padding: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis"
  }, placeholder)), /*#__PURE__*/_react.default.createElement("div", {
    className: "content"
  }, value.map((tag, tagIndex) =>
  /*#__PURE__*/
  // eslint-disable-next-line react/no-array-index-key
  _react.default.createElement("div", {
    key: "tag-".concat(tagIndex)
  }, /*#__PURE__*/_react.default.createElement(Tag, {
    enabled: enabled,
    index: tagIndex,
    muted: disabled,
    onRemove: handleTagRemove,
    tag: tag
  }))), /*#__PURE__*/_react.default.createElement("div", {
    key: "tag-input"
  }, /*#__PURE__*/_react.default.createElement(Input, _extends({}, restProps, {
    disabled: !enabled,
    onBlur: handleInputBlur,
    onChange: handleInputChange,
    onFocus: handleInputFocus,
    onKeyDown: handleInputKeyDown,
    ref: forwardedRef,
    type: "text",
    value: inputValue
  })))));
});
exports.TagInput = TagInput;
TagInput.displayName = 'TagInput';

function Tag(props) {
  var enabled = props.enabled,
      index = props.index,
      muted = props.muted,
      onRemove = props.onRemove,
      tag = props.tag;
  var handleRemoveClick = (0, _react.useCallback)(() => {
    onRemove(index);
  }, [index, onRemove]);
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-ui": "Tag",
    padding: 1,
    radius: 2,
    tone: "transparent"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: muted
  }, tag.value)), enabled && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    icon: _icons.CloseIcon,
    mode: "bleed",
    onClick: handleRemoveClick,
    padding: 1
  }))));
}
