// @sanity/field/lib/types/boolean/preview/BooleanPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = Checkbox;
exports.Switch = Switch;
exports.BooleanPreview = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BooleanPreview = function BooleanPreview(_ref) {
  var _schemaType$options;

  var value = _ref.value,
      schemaType = _ref.schemaType,
      color = _ref.color;
  var Preview = ((_schemaType$options = schemaType.options) === null || _schemaType$options === void 0 ? void 0 : _schemaType$options.layout) === 'checkbox' ? Checkbox : Switch;
  return /*#__PURE__*/_react.default.createElement(Preview, {
    checked: value,
    color: color
  });
};

exports.BooleanPreview = BooleanPreview;

function Checkbox(_ref2) {
  var checked = _ref2.checked,
      color = _ref2.color;
  return /*#__PURE__*/_react.default.createElement("svg", {
    width: "17",
    height: "17",
    viewBox: "0 0 17 17",
    xmlns: "http://www.w3.org/2000/svg",
    fill: color === null || color === void 0 ? void 0 : color.background
  }, /*#__PURE__*/_react.default.createElement("rect", {
    x: "0",
    y: "0",
    width: "17",
    height: "17",
    rx: "2.5"
  }), typeof checked === 'undefined' && /*#__PURE__*/_react.default.createElement("path", {
    d: "M4.07996 8.5H12.92",
    stroke: color === null || color === void 0 ? void 0 : color.text,
    strokeWidth: "2"
  }), checked && /*#__PURE__*/_react.default.createElement("path", {
    d: "M3.5 8L7 11.5L13.5 5",
    stroke: color === null || color === void 0 ? void 0 : color.text,
    strokeWidth: "2"
  }));
}

function Switch(_ref3) {
  var checked = _ref3.checked,
      color = _ref3.color;
  return /*#__PURE__*/_react.default.createElement("svg", {
    width: "38",
    height: "22",
    viewBox: "0 0 38 22",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("rect", {
    width: "38",
    height: "22",
    rx: "11",
    fill: checked ? color === null || color === void 0 ? void 0 : color.border : color === null || color === void 0 ? void 0 : color.background
  }), typeof checked === 'undefined' && /*#__PURE__*/_react.default.createElement("rect", {
    x: "11",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }), checked && /*#__PURE__*/_react.default.createElement("rect", {
    x: "18",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }), typeof checked === 'boolean' && !checked && /*#__PURE__*/_react.default.createElement("rect", {
    x: "4",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }));
}
