// @sanity/form-builder/lib/sanity/DefaultAssetSource/AssetMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AssetMenu = AssetMenu;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function AssetMenu(_ref) {
  var isSelected = _ref.isSelected,
      _ref$border = _ref.border,
      border = _ref$border === void 0 ? true : _ref$border,
      onAction = _ref.onAction;
  var triggerButtonMode = isSelected ? 'default' : 'bleed';
  var triggerButtonTone = isSelected ? 'primary' : 'default';
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      padding: 2,
      mode: border ? 'ghost' : triggerButtonMode,
      icon: _icons.EllipsisVerticalIcon,
      tone: border ? 'default' : triggerButtonTone
    }),
    id: "asset-menu",
    portal: true,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Show uses",
      icon: _icons.LinkIcon,
      onClick: () => {
        onAction({
          type: 'showUsage'
        });
      }
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Delete",
      icon: _icons.TrashIcon,
      tone: "critical",
      onClick: () => {
        onAction({
          type: 'delete'
        });
      }
    })),
    placement: "right"
  });
}
