// @sanity/form-builder/lib/sanity/DefaultAssetSource/AssetUsageDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AssetUsageDialog = AssetUsageDialog;

var _react = _interopRequireWildcard(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _legacyParts = require("../../legacyParts");

var _SpinnerWithText = require("../../components/SpinnerWithText");

var _DocumentList = require("./DocumentList");

var _ConfirmMessage = require("./ConfirmMessage");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var MODE_CONFIRM_DELETE = 'confirmDelete';
var MODE_LIST_USAGE = 'listUsage';

function AssetUsageDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_legacyParts.WithReferringDocuments, {
    id: props.asset._id
  }, _ref => {
    var isLoading = _ref.isLoading,
        referringDocuments = _ref.referringDocuments;
    return /*#__PURE__*/_react.default.createElement(InnerAssetUsageDialog, _extends({}, props, {
      assetIsLoading: isLoading,
      referringDocuments: referringDocuments
    }));
  });
}

var InnerAssetUsageDialog = _ref2 => {
  var asset = _ref2.asset,
      _ref2$assetType = _ref2.assetType,
      assetType = _ref2$assetType === void 0 ? 'image' : _ref2$assetType,
      _ref2$mode = _ref2.mode,
      mode = _ref2$mode === void 0 ? MODE_LIST_USAGE : _ref2$mode,
      _ref2$isDeleting = _ref2.isDeleting,
      isDeleting = _ref2$isDeleting === void 0 ? false : _ref2$isDeleting,
      _ref2$assetIsLoading = _ref2.assetIsLoading,
      assetIsLoading = _ref2$assetIsLoading === void 0 ? false : _ref2$assetIsLoading,
      _ref2$referringDocume = _ref2.referringDocuments,
      referringDocuments = _ref2$referringDocume === void 0 ? [] : _ref2$referringDocume,
      onClose = _ref2.onClose,
      onDelete = _ref2.onDelete;
  var isListMode = mode === MODE_LIST_USAGE;
  var defaultHeaderTitle = isListMode ? "Documents using ".concat(assetType) : "Delete ".concat(assetType);

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      canDelete = _useState2[0],
      setCanDelete = _useState2[1];

  var _useState3 = (0, _react.useState)(true),
      _useState4 = _slicedToArray(_useState3, 2),
      isLoadingParent = _useState4[0],
      setIsLoadingParent = _useState4[1];

  var _useState5 = (0, _react.useState)([]),
      _useState6 = _slicedToArray(_useState5, 2),
      publishedDocuments = _useState6[0],
      setPublishedDocuments = _useState6[1];

  var showActionFooter = mode === MODE_CONFIRM_DELETE;
  var hasResults = publishedDocuments.length > 0;
  var showDocumentList = mode === MODE_LIST_USAGE || hasResults;
  var noPaddingOnStack = mode === MODE_CONFIRM_DELETE && !hasResults;
  var footer = showActionFooter ? /*#__PURE__*/_react.default.createElement(_ui.Grid, {
    padding: 2,
    gap: 2,
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    mode: "bleed",
    text: "Cancel",
    onClick: onClose
  }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Delete",
    tone: "critical",
    icon: _icons.TrashIcon,
    onClick: onDelete,
    loading: isDeleting,
    disabled: !canDelete
  })) : undefined;
  (0, _react.useEffect)(() => {
    var drafts = referringDocuments.reduce((acc, doc) => doc._id.startsWith('drafts.') ? acc.concat(doc._id.slice(7)) : acc, []);
    var documentsWithoutDrafts = referringDocuments.filter(doc => !drafts.includes(doc._id));
    setPublishedDocuments(documentsWithoutDrafts);
    setCanDelete(documentsWithoutDrafts.length === 0 && !assetIsLoading);
    setIsLoadingParent(assetIsLoading);
  }, [assetIsLoading, referringDocuments]);
  return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    id: "asset-dialog",
    header: defaultHeaderTitle,
    width: 1,
    onClose: onClose,
    footer: footer,
    __unstable_autoFocus: !isLoadingParent
  }, isLoadingParent && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_SpinnerWithText.SpinnerWithText, {
    text: "Loading..."
  })), !isLoadingParent && /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    paddingX: noPaddingOnStack ? 0 : [2, 3, 4],
    paddingY: noPaddingOnStack ? 0 : [3, 3, 3, 4],
    space: 1
  }, mode === MODE_CONFIRM_DELETE && /*#__PURE__*/_react.default.createElement(_ConfirmMessage.ConfirmMessage, {
    asset: asset,
    assetType: assetType,
    hasResults: hasResults
  }), showDocumentList && /*#__PURE__*/_react.default.createElement(_DocumentList.DocumentList, {
    asset: asset,
    referringDocuments: publishedDocuments,
    assetType: assetType
  })));
};
