// @sanity/base/lib/__legacy/@sanity/components/previews/BlockImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _blockImageStyle = _interopRequireDefault(require("part:@sanity/components/previews/block-image-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 600,
  height: 600,
  fit: 'fillmax'
};

class BlockImagePreview extends _react.default.PureComponent {
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("BlockImagePreview from `part:@sanity/components/previews/block-image` is deprecated. Please use `import {BlockImagePreview} from '@sanity/base/components'` instead");
  } // eslint-disable-next-line complexity


  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        description = _this$props.description,
        mediaDimensions = _this$props.mediaDimensions,
        media = _this$props.media,
        children = _this$props.children,
        status = _this$props.status;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.root
    }, title && /*#__PURE__*/_react.default.createElement("header", {
      className: _blockImageStyle.default.header
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _blockImageStyle.default.title
    }, title)), /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.preview
    }, media && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions || DEFAULT_MEDIA_DIMENSIONS,
      layout: 'blockImage'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), subtitle || description || status || /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.heading
    }, subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _blockImageStyle.default.subtitle
    }, subtitle), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _blockImageStyle.default.description
    }, description), status && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.status
    }, typeof status === 'function' && status({
      layout: 'default'
    }) || status))), children && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockImageStyle.default.children
    }, children));
  }

}

exports.default = BlockImagePreview;
