// @sanity/base/lib/__legacy/@sanity/components/previews/BlockPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _blockStyle = _interopRequireDefault(require("part:@sanity/components/previews/block-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />

/* eslint-disable complexity */
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 160,
  height: 160,
  aspect: 1,
  fit: 'crop'
};

class BlockPreview extends _react.default.PureComponent {
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("BlockPreview from `part:@sanity/components/previews/block` is deprecated. Please use `import {BlockPreview} from '@sanity/base/components'` instead");
  }

  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        description = _this$props.description,
        mediaDimensions = _this$props.mediaDimensions,
        media = _this$props.media,
        status = _this$props.status,
        children = _this$props.children,
        extendedPreview = _this$props.extendedPreview;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.root
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.header
    }, media && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(_blockStyle.default.media)
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions || DEFAULT_MEDIA_DIMENSIONS,
      layout: 'default'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.heading
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _blockStyle.default.title
    }, title), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _blockStyle.default.subtitle
    }, subtitle), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _blockStyle.default.description
    }, description)), status && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.status
    }, typeof status === 'function' && /*#__PURE__*/(0, _react.createElement)(status, {
      layout: 'default'
    }) || status)), children && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.content
    }, children), extendedPreview && /*#__PURE__*/_react.default.createElement("div", {
      className: _blockStyle.default.extendedPreview
    }, extendedPreview));
  }

}

exports.default = BlockPreview;
