// @sanity/base/lib/__legacy/@sanity/components/previews/DefaultPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _defaultStyle = _interopRequireDefault(require("part:@sanity/components/previews/default-style"));

var _react = _interopRequireDefault(require("react"));

var _Styleable = _interopRequireDefault(require("../utilities/Styleable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />

/** Deprecated. The new DefaultPreview component is located in @sanity/base/components/previews */
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 80,
  height: 80,
  aspect: 1,
  fit: 'crop'
};

class DefaultPreview extends _react.default.PureComponent {
  // eslint-disable-next-line complexity
  // eslint-disable-next-line class-methods-use-this
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("DefaultPreview from `part:@sanity/components/previews/default` is deprecated. Please use `import {DefaultPreview} from '@sanity/base/components'` instead");
  }

  render() {
    var _this$props = this.props,
        _this$props$title = _this$props.title,
        title = _this$props$title === void 0 ? 'Untitled' : _this$props$title,
        subtitle = _this$props.subtitle,
        media = _this$props.media,
        children = _this$props.children,
        status = _this$props.status,
        isPlaceholder = _this$props.isPlaceholder,
        progress = _this$props.progress,
        _this$props$mediaDime = _this$props.mediaDimensions,
        mediaDimensions = _this$props$mediaDime === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _this$props$mediaDime,
        _this$props$styles = _this$props.styles,
        styles = _this$props$styles === void 0 ? {} : _this$props$styles;

    if (isPlaceholder) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: styles.placeholder
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: styles.inner
      }, media !== false && /*#__PURE__*/_react.default.createElement("div", {
        className: styles.media
      }), /*#__PURE__*/_react.default.createElement("div", {
        className: styles.heading
      }, /*#__PURE__*/_react.default.createElement("h2", {
        className: styles.title
      }, "Loading\u2026"), /*#__PURE__*/_react.default.createElement("h3", {
        className: styles.subtitle
      }, "Loading\u2026"))));
    }

    var className = (0, _classnames.default)(styles.root, subtitle !== undefined && styles.hasSubtitle);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: styles.inner
    }, media !== false && media !== undefined && /*#__PURE__*/_react.default.createElement("div", {
      className: styles.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: styles.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("div", {
      className: styles.heading
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: styles.title
    }, typeof title !== 'function' && title, typeof title === 'function' && title({
      layout: 'default'
    })), (subtitle || subtitle === 0) && /*#__PURE__*/_react.default.createElement("h3", {
      className: styles.subtitle
    }, typeof subtitle === 'function' && subtitle({
      layout: 'default'
    }) || subtitle)), status && /*#__PURE__*/_react.default.createElement("div", {
      className: styles.status
    }, typeof status === 'function' && status({
      layout: 'default'
    }) || status), children && /*#__PURE__*/_react.default.createElement("div", {
      className: styles.children
    }, children), typeof progress === 'number' && progress > -1 && /*#__PURE__*/_react.default.createElement("div", {
      className: styles.progress
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: styles.progressBar,
      style: {
        width: "".concat(progress, "%")
      }
    }))));
  }

}

var _default = (0, _Styleable.default)(DefaultPreview, _defaultStyle.default);

exports.default = _default;
