// @sanity/base/lib/__legacy/@sanity/components/previews/DetailPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _detailStyle = _interopRequireDefault(require("part:@sanity/components/previews/detail-style"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var index = 0;
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 120,
  height: 120,
  fit: 'crop',
  aspect: 1
};

class DetailPreview extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "index", index++);
  }

  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("DetailPreview from `part:@sanity/components/previews/detail` is deprecated. Please use `import {DetailPreview} from '@sanity/base/components'` instead");
  } // eslint-disable-next-line complexity


  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        description = _this$props.description,
        _this$props$mediaDime = _this$props.mediaDimensions,
        mediaDimensions = _this$props$mediaDime === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _this$props$mediaDime,
        media = _this$props.media,
        status = _this$props.status,
        children = _this$props.children,
        isPlaceholder = _this$props.isPlaceholder;

    if (isPlaceholder) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _detailStyle.default.placeholder
      }, media !== false && /*#__PURE__*/_react.default.createElement("div", {
        className: _detailStyle.default.media
      }), /*#__PURE__*/_react.default.createElement("div", {
        className: _detailStyle.default.content
      }, /*#__PURE__*/_react.default.createElement("h2", {
        className: _detailStyle.default.title
      }, "Loading"), /*#__PURE__*/_react.default.createElement("h3", {
        className: _detailStyle.default.subtitle
      }, "Loading"), /*#__PURE__*/_react.default.createElement("p", {
        className: _detailStyle.default.description
      }, "Loading")));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.root
    }, media !== false && /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.content
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.top
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.heading
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _detailStyle.default.title
    }, typeof title === 'function' && title({
      layout: 'detail'
    }) || title || /*#__PURE__*/_react.default.createElement("em", null, "Untitled")), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _detailStyle.default.subtitle
    }, typeof subtitle === 'function' && subtitle({
      layout: 'detail'
    }) || subtitle)), status && /*#__PURE__*/_react.default.createElement("div", {
      className: _detailStyle.default.status
    }, typeof status === 'function' && status({
      layout: 'detail'
    }) || status)), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _detailStyle.default.description
    }, typeof description === 'function' && description({
      layout: 'detail'
    }), typeof description === 'string' && description, typeof description === 'object' && description)), children);
  }

}

exports.default = DetailPreview;
