// @sanity/base/lib/__legacy/@sanity/components/previews/InlinePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _inlineStyle = _interopRequireDefault(require("part:@sanity/components/previews/inline-style"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 32,
  height: 32,
  fit: 'crop',
  aspect: 1
};

class InlinePreview extends _react.default.PureComponent {
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("InlinePreview from `part:@sanity/components/previews/inline` is deprecated. Please use `import {InlinePreview} from '@sanity/base/components'` instead");
  } // eslint-disable-next-line complexity


  render() {
    var _this$props = this.props,
        title = _this$props.title,
        media = _this$props.media,
        _this$props$mediaDime = _this$props.mediaDimensions,
        mediaDimensions = _this$props$mediaDime === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _this$props$mediaDime,
        children = _this$props.children;

    if (!title && !media) {
      return /*#__PURE__*/_react.default.createElement("span", null);
    }

    return /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.root
    }, media && /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media !== 'function' && media, /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.title
    }, typeof title === 'function' && title({
      layout: 'inline'
    }) || title), children && /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.children
    }, children));
  }

}

exports.default = InlinePreview;
