// @sanity/base/lib/__legacy/@sanity/components/previews/MediaPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _mediaStyle = _interopRequireDefault(require("part:@sanity/components/previews/media-style"));

var _circle = _interopRequireDefault(require("part:@sanity/components/progress/circle"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var DEFAULT_MEDIA_DIMENSIONS = {
  width: 160,
  height: 160,
  aspect: 1,
  fit: 'crop'
};

class MediaPreview extends _react.default.PureComponent {
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("MediaPreview from `part:@sanity/components/previews/media` is deprecated. Please use `import {MediaPreview} from '@sanity/base/components'` instead");
  } // eslint-disable-next-line complexity


  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        description = _this$props.description,
        media = _this$props.media,
        _this$props$mediaDime = _this$props.mediaDimensions,
        mediaDimensions = _this$props$mediaDime === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _this$props$mediaDime,
        children = _this$props.children,
        isPlaceholder = _this$props.isPlaceholder,
        progress = _this$props.progress;
    var aspect = (mediaDimensions === null || mediaDimensions === void 0 ? void 0 : mediaDimensions.aspect) || DEFAULT_MEDIA_DIMENSIONS.aspect;

    if (isPlaceholder) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _mediaStyle.default.root
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: _mediaStyle.default.padder,
        style: {
          paddingTop: "".concat(100 / aspect, "%")
        }
      }));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.root,
      title: typeof title === 'string' ? title : undefined
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.padder,
      style: {
        paddingTop: "".concat(100 / aspect, "%")
      }
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.mediaContainer
    }, typeof media === 'undefined' && /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.mediaString
    }, title), typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'media'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media, typeof progress === 'number' && progress > -1 && /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.progress
    }, /*#__PURE__*/_react.default.createElement(_circle.default, {
      percent: progress,
      showPercent: true,
      text: "Uploaded"
    }))), /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.meta
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _mediaStyle.default.metaInner
    }, title && /*#__PURE__*/_react.default.createElement("h2", {
      className: _mediaStyle.default.title
    }, title), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _mediaStyle.default.subtitle
    }, subtitle), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _mediaStyle.default.description
    }, description))), children);
  }

}

exports.default = MediaPreview;
