// @sanity/base/lib/__legacy/@sanity/components/toggles/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _autoId = require("@reach/auto-id");

var _Checkbox = _interopRequireDefault(require("./Checkbox.css"));

var _shared = _interopRequireDefault(require("./shared.css"));

var _excluded = ["label", "description", "markers", "checked", "disabled", "readOnly", "children"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var _default = /*#__PURE__*/_react.default.forwardRef(function Checkbox(_ref, ref) {
  var label = _ref.label,
      description = _ref.description,
      markers = _ref.markers,
      checked = _ref.checked,
      disabled = _ref.disabled,
      readOnly = _ref.readOnly,
      children = _ref.children,
      rest = _objectWithoutProperties(_ref, _excluded);

  var elementId = (0, _autoId.useId)();
  var innerRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(() => {
    if (innerRef !== null && innerRef !== void 0 && innerRef.current) {
      innerRef.current.indeterminate = checked === undefined;
    }
  }, [checked]);
  var setRef = (0, _react.useCallback)(el => {
    innerRef.current = el;
    if (typeof ref === 'function') ref(el);else if (ref && typeof ref === 'object') ref.current = el;
  }, [ref]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _Checkbox.default.root
  }, /*#__PURE__*/_react.default.createElement("input", _extends({}, rest, {
    id: "checkbox-".concat(elementId, "-input"),
    "aria-describedby": "checkbox-".concat(elementId, "-description"),
    className: _Checkbox.default.input,
    checked: checked || false,
    type: "checkbox",
    disabled: disabled || readOnly,
    ref: setRef
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _Checkbox.default.checkbox
  }, /*#__PURE__*/_react.default.createElement("svg", {
    className: "".concat(_Checkbox.default.mark, " ").concat(_Checkbox.default.checkmark),
    viewBox: "0 0 16 16",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "M3 7.5L6.5 11L13 4.5",
    stroke: "currentColor",
    strokeWidth: "2"
  })), /*#__PURE__*/_react.default.createElement("svg", {
    className: "".concat(_Checkbox.default.mark, " ").concat(_Checkbox.default.indeterminate),
    viewBox: "0 0 16 16",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "M12 8H4",
    stroke: "currentColor",
    strokeWidth: "2"
  }))), label && /*#__PURE__*/_react.default.createElement("div", {
    className: _shared.default.label
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _shared.default.titleWrapper
  }, /*#__PURE__*/_react.default.createElement("label", {
    className: _shared.default.title,
    htmlFor: "checkbox-".concat(elementId, "-input")
  }, label), children), description && /*#__PURE__*/_react.default.createElement("div", {
    id: "checkbox-".concat(elementId, "-description"),
    className: _shared.default.description
  }, description)));
});

exports.default = _default;
