// @sanity/base/lib/components/previews/blockPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockPreview = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _blockPreview = require("./blockPreview.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_MEDIA_DIMENSIONS = {
  width: 40,
  height: 40,
  aspect: 1,
  fit: 'crop',
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
};

var BlockPreview = props => {
  var title = props.title,
      subtitle = props.subtitle,
      description = props.description,
      mediaDimensions = props.mediaDimensions,
      media = props.media,
      status = props.status,
      children = props.children,
      extendedPreview = props.extendedPreview;
  return /*#__PURE__*/_react.default.createElement(_blockPreview.Root, null, /*#__PURE__*/_react.default.createElement(_blockPreview.Header, {
    padding: 2,
    align: "center"
  }, media && /*#__PURE__*/_react.default.createElement(_blockPreview.MediaWrapper, {
    marginRight: 3
  }, typeof media === 'function' && media({
    dimensions: mediaDimensions || DEFAULT_MEDIA_DIMENSIONS,
    layout: 'block'
  }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement(_ui.Box, null, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement(_blockPreview.ContentWrapper, {
    flex: 1,
    paddingY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis",
    style: {
      color: 'inherit'
    }
  }, title && typeof title === 'function' ? title({
    layout: 'block'
  }) : title, !title && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Untitled")), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, typeof subtitle === 'function' ? subtitle({
    layout: 'block'
  }) : subtitle)), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, typeof description === 'function' ? description({
    layout: 'block'
  }) : description))), status && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 3
  }, typeof status === 'function' ? status({
    layout: 'block'
  }) : status)), children && /*#__PURE__*/_react.default.createElement(_ui.Box, null, children), extendedPreview && /*#__PURE__*/_react.default.createElement(_ui.Box, null, extendedPreview));
};

exports.BlockPreview = BlockPreview;
