// @sanity/base/lib/components/previews/defaultPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultPreview = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Root = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 35px;\n"])));
var MediaWrapper = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: relative;\n  width: 35px;\n  height: 35px;\n  min-width: 35px;\n  border-radius: ", "px;\n\n  & img {\n    position: absolute;\n    left: 0;\n    top: 0;\n    width: 100%;\n    height: 100%;\n    object-fit: contain;\n    border-radius: inherit;\n  }\n\n  & svg {\n    display: block;\n    font-size: calc(21 / 16 * 1em);\n\n    &[data-sanity-icon] {\n      font-size: calc(33 / 16 * 1em);\n      margin: calc(6 / 36 * -1em);\n    }\n  }\n\n  & img + span {\n    display: block;\n    position: absolute;\n    left: 0;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    box-shadow: inset 0 0 0 1px var(--card-fg-color);\n    opacity: 0.2;\n    border-radius: inherit;\n  }\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.radius[2];
});

var DefaultPreview = props => {
  var title = props.title,
      subtitle = props.subtitle,
      media = props.media,
      status = props.status,
      isPlaceholder = props.isPlaceholder,
      children = props.children;
  return /*#__PURE__*/_react.default.createElement(Root, {
    align: "center"
  }, isPlaceholder && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Skeleton, {
    style: {
      width: 35,
      height: 35
    },
    radius: 2,
    marginRight: 2,
    animated: true
  }), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2,
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.TextSkeleton, {
    style: {
      maxWidth: 320
    },
    radius: 1,
    animated: true
  }), /*#__PURE__*/_react.default.createElement(_ui.TextSkeleton, {
    style: {
      maxWidth: 200
    },
    radius: 1,
    size: 1,
    animated: true
  }))), !isPlaceholder && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, media !== false && media !== undefined && /*#__PURE__*/_react.default.createElement(MediaWrapper, {
    align: "center",
    justify: "center",
    marginRight: 2,
    sizing: "border",
    overflow: "hidden"
  }, typeof media === 'function' && media({
    dimensions: {
      width: 35,
      height: 35,
      aspect: 1,
      fit: 'crop',
      dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
    },
    layout: 'default'
  }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", null, media), /*#__PURE__*/_react.default.isValidElement(media) && media, /*#__PURE__*/_react.default.createElement("span", null)), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2,
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis",
    style: {
      color: 'inherit'
    }
  }, title && typeof title === 'function' ? title({
    layout: 'default'
  }) : title, !title && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Untitled")), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, typeof subtitle === 'function' ? subtitle({
    layout: 'default'
  }) : subtitle), children && /*#__PURE__*/_react.default.createElement("div", null, children)), status && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 3
  }, typeof status === 'function' ? status({
    layout: 'default'
  }) : status)));
};

exports.DefaultPreview = DefaultPreview;
