// @sanity/base/lib/components/previews/detailPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DetailPreview = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _detailPreview = require("./detailPreview.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_MEDIA_DIMENSIONS = {
  width: 80,
  height: 80,
  fit: 'crop',
  aspect: 1,
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
};

var DetailPreview = props => {
  var title = props.title,
      subtitle = props.subtitle,
      description = props.description,
      _props$mediaDimension = props.mediaDimensions,
      mediaDimensions = _props$mediaDimension === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _props$mediaDimension,
      media = props.media,
      status = props.status,
      children = props.children,
      isPlaceholder = props.isPlaceholder;
  var fonts = (0, _ui.useTheme)().sanity.fonts;
  var textSize = fonts.text.sizes[1];
  var maxHeight = textSize.lineHeight * 2 - textSize.ascenderHeight - textSize.descenderHeight;

  if (isPlaceholder) {
    return /*#__PURE__*/_react.default.createElement(_detailPreview.Root, {
      align: "center"
    }, /*#__PURE__*/_react.default.createElement(_ui.Skeleton, {
      style: {
        width: 80,
        height: 80
      },
      radius: 2,
      marginRight: 2,
      animated: true
    }), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2,
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.TextSkeleton, {
      style: {
        maxWidth: 320
      },
      radius: 1,
      animated: true
    }), /*#__PURE__*/_react.default.createElement(_ui.TextSkeleton, {
      style: {
        maxWidth: 200
      },
      radius: 1,
      size: 1,
      animated: true
    })));
  }

  return /*#__PURE__*/_react.default.createElement(_detailPreview.Root, {
    align: "center"
  }, media !== false && /*#__PURE__*/_react.default.createElement(_detailPreview.MediaWrapper, {
    align: "center",
    justify: "center",
    marginRight: 2
  }, typeof media === 'function' && media({
    dimensions: mediaDimensions,
    layout: 'detail'
  }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement(_detailPreview.MediaString, null, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement(_detailPreview.Content, {
    justify: "center",
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_detailPreview.Top, {
    align: "center",
    justify: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_detailPreview.Header, {
    space: 2,
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis",
    style: {
      color: 'inherit'
    }
  }, title && typeof title === 'function' ? title({
    layout: 'detail'
  }) : title, !title && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Untitled")), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, typeof subtitle === 'function' ? subtitle({
    layout: 'detail'
  }) : subtitle)), status && /*#__PURE__*/_react.default.createElement(_detailPreview.StatusWrapper, {
    paddingLeft: 1
  }, typeof status === 'function' ? status({
    layout: 'detail'
  }) : status)), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3,
    overflow: "hidden",
    style: {
      maxHeight
    }
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1
  }, typeof description === 'function' ? description({
    layout: 'detail'
  }) : description))), children);
};

exports.DetailPreview = DetailPreview;
