// @sanity/base/lib/components/previews/detailPreview.styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Content = exports.StatusWrapper = exports.MediaString = exports.MediaWrapper = exports.Header = exports.Top = exports.Root = void 0;

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Root = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  user-select: none;\n  height: 5em;\n  box-sizing: border-box;\n  max-width: 100%;\n  position: relative;\n"])));
exports.Root = Root;
var Top = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  min-width: 0;\n"])));
exports.Top = Top;
var Header = (0, _styledComponents.default)(_ui.Stack)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  min-width: 0;\n"])));
exports.Header = Header;
var MediaWrapper = (0, _styledComponents.default)(_ui.Flex)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  flex-basis: 5em;\n  flex-grow: 1;\n  min-width: 5em;\n  width: 5em;\n  max-width: 5em;\n  overflow: hidden;\n  position: relative;\n\n  img {\n    display: block;\n    width: 5em;\n    height: 5em;\n    object-fit: contain;\n  }\n\n  svg {\n    font-size: 3rem;\n    color: inherit;\n  }\n"])));
exports.MediaWrapper = MediaWrapper;
var MediaString = (0, _styledComponents.default)(_ui.Box)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral([""])));
exports.MediaString = MediaString;
var StatusWrapper = (0, _styledComponents.default)(_ui.Box)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n  white-space: nowrap;\n"])));
exports.StatusWrapper = StatusWrapper;
var Content = (0, _styledComponents.default)(_ui.Flex)(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n  flex-grow: 1;\n  min-width: 0;\n\n  > * {\n    min-width: 0;\n  }\n"])));
exports.Content = Content;
