// @sanity/base/lib/__legacy/@sanity/components/selects/RadioSelect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _RadioButtonDefault = _interopRequireDefault(require("../radiobutton/RadioButtonDefault"));

var _RadioSelect = _interopRequireDefault(require("./RadioSelect.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RadioSelect = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  var name = props.name,
      direction = props.direction,
      onChange = props.onChange,
      onFocus = props.onFocus,
      value = props.value,
      readOnly = props.readOnly,
      _props$items = props.items,
      items = _props$items === void 0 ? [] : _props$items,
      inputId = props.inputId;

  var handleRadioChange = item => {
    if (onChange) onChange(item);
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_RadioSelect.default.root, direction == 'vertical' ? _RadioSelect.default.vertical : _RadioSelect.default.horizontal),
    ref: ref
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _RadioSelect.default.radioContainer,
    id: inputId,
    role: "group"
  }, items.map((item, i) => {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _RadioSelect.default.item,
      key: i
    }, /*#__PURE__*/_react.default.createElement(_RadioButtonDefault.default, {
      disabled: readOnly,
      name: name,
      label: item.title,
      item: item,
      onChange: handleRadioChange,
      checked: value === item,
      onFocus: onFocus
    }));
  })));
});

RadioSelect.displayName = 'RadioSelect';
var _default = RadioSelect;
exports.default = _default;
