// @sanity/default-layout/lib/main/ErrorScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _react = _interopRequireDefault(require("react"));

var _ErrorScreen = _interopRequireDefault(require("./ErrorScreen.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function getErrorWithStack(err) {
  var stack = err.stack.toString();
  var message = err.message;
  return stack.indexOf(message) === -1 ? "".concat(message, "\n\n").concat(stack) : stack;
}

function limitStackLength(stack) {
  return stack.split('\n').slice(0, 15).join('\n');
}

function formatStack(stack) {
  return stack // Prettify builder functions
  .replace(/\(\.\.\.\)\./g, '(...)\n  .') // Remove webpack cruft from function names
  .replace(/__WEBPACK_IMPORTED_MODULE_\d+_+/g, '') // Remove default export postfix from function names
  .replace(/___default\./g, '.') // Replace full host path, leave only path to JS-file
  .replace(new RegExp(" \\(https?:\\/\\/".concat(window.location.host), 'g'), ' (');
}

function ErrorScreen(props) {
  var activeTool = props.activeTool,
      error = props.error,
      info = props.info,
      onRetry = props.onRetry,
      onShowDetails = props.onShowDetails,
      showErrorDetails = props.showErrorDetails;
  var toolName = activeTool && (activeTool.title || activeTool.name) || 'active';
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.card
  }, /*#__PURE__*/_react.default.createElement("header", {
    className: _ErrorScreen.default.cardHeader
  }, /*#__PURE__*/_react.default.createElement("h2", {
    className: _ErrorScreen.default.cardTitle
  }, "The \u2018", toolName, "\u2019 tool crashed")), /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.cardContent
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_default2.default, {
    onClick: onRetry
  }, "Retry")), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_default2.default, {
    onClick: onShowDetails,
    disabled: showErrorDetails
  }, "Show details")))), showErrorDetails && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, error.stack ? /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.stack
  }, /*#__PURE__*/_react.default.createElement("h3", null, "Stack trace:"), /*#__PURE__*/_react.default.createElement("pre", null, formatStack(limitStackLength(getErrorWithStack(error))))) : /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.stack
  }, /*#__PURE__*/_react.default.createElement("h3", null, "Error:"), /*#__PURE__*/_react.default.createElement("pre", null, error.message)), info && info.componentStack && /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorScreen.default.stack
  }, /*#__PURE__*/_react.default.createElement("h3", null, "Component stack:"), /*#__PURE__*/_react.default.createElement("pre", null, info.componentStack.replace(/^\s*\n+/, '')))));
}

var _default = ErrorScreen;
exports.default = _default;
