// @sanity/form-builder/lib/inputs/TextInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _autoId = require("@reach/auto-id");

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _PatchEvent = _interopRequireWildcard(require("../PatchEvent"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var StyledTextArea = (0, _styledComponents.default)(_ui.TextArea)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-as='textarea'] {\n    resize: vertical;\n  }\n"])));

var TextInput = /*#__PURE__*/_react.default.forwardRef(function TextInput(props, forwardedRef) {
  var value = props.value,
      markers = props.markers,
      type = props.type,
      readOnly = props.readOnly,
      level = props.level,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      onChange = props.onChange,
      presence = props.presence;
  var inputId = (0, _autoId.useId)();
  var errors = (0, _react.useMemo)(() => markers.filter(marker => marker.type === 'validation' && marker.level === 'error'), [markers]);

  var handleChange = _react.default.useCallback(event => {
    var inputValue = event.currentTarget.value;
    onChange(_PatchEvent.default.from(inputValue ? (0, _PatchEvent.set)(inputValue) : (0, _PatchEvent.unset)()));
  }, [onChange]);

  return /*#__PURE__*/_react.default.createElement(_components.FormField, {
    level: level,
    __unstable_markers: markers,
    title: type.title,
    description: type.description,
    __unstable_presence: presence,
    inputId: inputId
  }, /*#__PURE__*/_react.default.createElement(StyledTextArea, {
    id: inputId,
    customValidity: errors && errors.length > 0 ? errors[0].item.message : '',
    value: value || '',
    readOnly: Boolean(readOnly),
    placeholder: type.placeholder,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    rows: typeof type.rows === 'number' ? type.rows : 10,
    ref: forwardedRef
  }));
});

var _default = TextInput;
exports.default = _default;
