// @sanity/portable-text-editor/lib/editor/DraggableBlock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DraggableBlock = void 0;

var _react = _interopRequireWildcard(require("react"));

var _slate = require("slate");

var _slateReact = require("@sanity/slate-react");

var _debug = require("../utils/debug");

var _weakMaps = require("../utils/weakMaps");

var _nodes = require("./nodes");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var debug = (0, _debug.debugWithName)('components:DraggableBlock');
var debugRenders = false;

function useForceUpdate() {
  var _useState = (0, _react.useState)(0),
      _useState2 = _slicedToArray(_useState, 2),
      setValue = _useState2[1];

  return () => setValue(value => ++value);
}

var DraggableBlock = _ref => {
  var children = _ref.children,
      element = _ref.element,
      readOnly = _ref.readOnly;
  var editor = (0, _slateReact.useEditor)();
  var dragGhostRef = (0, _react.useRef)();
  var forceUpdate = useForceUpdate();

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isDragOver = _useState4[0],
      setIsDragOver = _useState4[1];

  var isVoid = (0, _react.useMemo)(() => _slate.Editor.isVoid(editor, element), []);
  var isInline = (0, _react.useMemo)(() => _slate.Editor.isInline(editor, element), []);

  if (readOnly) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
  }

  if (debugRenders) {
    debug('render');
  } // Note: this is called not for the dragging block, but for the targets when the block is dragged over them


  var handleDragOver = event => {
    var isMyDragOver = _weakMaps.IS_DRAGGING_BLOCK_ELEMENT.get(editor); // debug('Drag over', isMyDragOver)


    if (!isMyDragOver) {
      return;
    }

    event.preventDefault();
    event.dataTransfer.dropEffect = 'move';

    _weakMaps.IS_DRAGGING_ELEMENT_TARGET.set(editor, element);

    var blockElement = _slateReact.ReactEditor.toDOMNode(editor, element);

    var elementRect = blockElement.getBoundingClientRect();
    var offset = elementRect.top;
    var height = elementRect.height;
    var Y = event.pageY;
    var loc = Math.abs(offset - Y);
    var position = 'bottom';

    if (element === editor.children[0]) {
      position = 'top';
    } else if (loc < height / 2) {
      position = 'top';

      _weakMaps.IS_DRAGGING_BLOCK_TARGET_POSITION.set(editor, position);
    } else {
      position = 'bottom';

      _weakMaps.IS_DRAGGING_BLOCK_TARGET_POSITION.set(editor, position);
    }

    if (isMyDragOver === element) {
      event.dataTransfer.dropEffect = 'none';
      return;
    }

    forceUpdate();
    setIsDragOver(true);
  }; // Note: this is called not for the dragging block, but for the targets when the block is dragged over them


  var handleDragLeave = event => {
    setIsDragOver(false);
  }; // Note: this is called for the dragging block


  var handleDragEnd = event => {
    _weakMaps.IS_DRAGGING.set(editor, false);

    event.preventDefault();
    event.stopPropagation();

    var targetBlock = _weakMaps.IS_DRAGGING_ELEMENT_TARGET.get(editor);

    _weakMaps.IS_DRAGGING_ELEMENT_TARGET.delete(editor);

    if (dragGhostRef.current) {
      debug('Removing drag ghost');
      document.body.removeChild(dragGhostRef.current);
    }

    if (targetBlock) {
      var _dragPosition = _weakMaps.IS_DRAGGING_BLOCK_TARGET_POSITION.get(editor);

      _weakMaps.IS_DRAGGING_BLOCK_TARGET_POSITION.delete(editor);

      var targetPath = _slateReact.ReactEditor.findPath(editor, targetBlock);

      var myPath = _slateReact.ReactEditor.findPath(editor, element);

      var isBefore = _slate.Path.isBefore(myPath, targetPath);

      if (_dragPosition === 'bottom' && !isBefore) {
        // If it is already at the bottom, don't do anything.
        if (targetPath[0] >= editor.children.length - 1) {
          debug('target is already at the bottom, not moving');
          return;
        }

        var originalPath = targetPath;
        targetPath = _slate.Path.next(targetPath);
        debug("Adjusting targetPath from ".concat(JSON.stringify(originalPath), " to ").concat(JSON.stringify(targetPath)));
      }

      if (_dragPosition === 'top' && isBefore && targetPath[0] !== editor.children.length - 1) {
        var _originalPath = targetPath;
        targetPath = _slate.Path.previous(targetPath);
        debug("Adjusting targetPath from ".concat(JSON.stringify(_originalPath), " to ").concat(JSON.stringify(targetPath)));
      }

      if (_slate.Path.equals(targetPath, myPath)) {
        event.preventDefault();
        debug('targetPath and myPath is the same, not moving');
        return;
      }

      debug("Moving element ".concat(element._key, " from path ").concat(JSON.stringify(myPath), " to ").concat(JSON.stringify(targetPath), " (").concat(_dragPosition, ")"));

      _slate.Transforms.moveNodes(editor, {
        at: myPath,
        to: targetPath
      });

      editor.onChange();
      return;
    }

    debug('No target element, not doing anything');
  }; // Note: this is called not for the dragging block, but for the drop target


  var handleDrop = event => {
    if (_weakMaps.IS_DRAGGING_BLOCK_ELEMENT.get(editor)) {
      debug('On drop (prevented)', element);
      event.preventDefault();
      event.stopPropagation();
      setIsDragOver(false);
    }
  }; // Note: this is called for the dragging block


  var handleDrag = event => {
    if (!isVoid) {
      _weakMaps.IS_DRAGGING_BLOCK_ELEMENT.delete(editor);

      return;
    }

    _weakMaps.IS_DRAGGING.set(editor, true);

    _weakMaps.IS_DRAGGING_BLOCK_ELEMENT.set(editor, element);

    event.stopPropagation(); // Stop propagation so that leafs don't get this and take focus/selection!
  }; // Note: this is called for the dragging block


  var handleDragStart = event => {
    if (!isVoid || isInline) {
      debug('Not dragging block');

      _weakMaps.IS_DRAGGING_BLOCK_ELEMENT.delete(editor);

      _weakMaps.IS_DRAGGING.set(editor, false);

      return;
    }

    debug('Drag start');

    _weakMaps.IS_DRAGGING.set(editor, true);

    event.dataTransfer.setData('application/portable-text', 'something');
    event.dataTransfer.effectAllowed = 'move none'; // Specify dragImage so that single elements in the preview will not be the drag image,
    // but always the whole block preview itself.
    // Also clone it so that it will not be visually clipped by scroll-containers etc.

    var _element = event.currentTarget;

    if (_element && _element instanceof HTMLElement) {
      var dragGhost = _element.cloneNode(true);

      dragGhostRef.current = dragGhost;
      dragGhost.style.width = "".concat(element.clientWidth, "px");
      dragGhost.style.height = "".concat(element.clientHeight, "px");
      dragGhost.style.position = 'absolute';
      dragGhost.style.top = '-99999px';
      dragGhost.style.left = '-99999px';

      if (document.body) {
        document.body.appendChild(dragGhost);

        var rect = _element.getBoundingClientRect();

        var x = event.clientX - rect.left;
        var y = event.clientY - rect.top;
        dragGhost.style.width = "".concat(rect.width, "px");
        dragGhost.style.height = "".concat(rect.height, "px");
        event.dataTransfer.setDragImage(dragGhost, x, y);
      }
    }

    handleDrag(event);
  };

  var isDraggingOverFirstBlock = isDragOver && editor.children[0] === _weakMaps.IS_DRAGGING_ELEMENT_TARGET.get(editor);

  var isDraggingOverLastBlock = isDragOver && editor.children[editor.children.length - 1] === _weakMaps.IS_DRAGGING_ELEMENT_TARGET.get(editor);

  var dragPosition = _weakMaps.IS_DRAGGING_BLOCK_TARGET_POSITION.get(editor);

  return /*#__PURE__*/_react.default.createElement("div", {
    draggable: isVoid,
    onDragStart: handleDragStart,
    onDrag: handleDrag,
    onDragOver: handleDragOver,
    onDragLeave: handleDragLeave,
    onDragEnd: handleDragEnd,
    onDrop: handleDrop
  }, /*#__PURE__*/_react.default.createElement(_nodes.DraggableBlockWrapper, {
    isDraggingOverTop: isDraggingOverFirstBlock || isDragOver && !isDraggingOverFirstBlock && !isDraggingOverLastBlock && dragPosition === 'top',
    isDraggingOverBottom: isDraggingOverLastBlock || isDragOver && !isDraggingOverFirstBlock && !isDraggingOverLastBlock && dragPosition === 'bottom'
  }, children));
};

exports.DraggableBlock = DraggableBlock;
