// @sanity/base/lib/preview/prepareForPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.invokePrepare = invokePrepare;
exports.default = prepareForPreview;

var _uniqBy2 = _interopRequireDefault(require("lodash/uniqBy"));

var _pick2 = _interopRequireDefault(require("lodash/pick"));

var _isPlainObject2 = _interopRequireDefault(require("lodash/isPlainObject"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _flatten2 = _interopRequireDefault(require("lodash/flatten"));

var _debounce2 = _interopRequireDefault(require("lodash/debounce"));

var _types = require("@sanity/types");

var _constants = require("./constants");

var _portableText = require("./utils/portableText");

var _keysOf = require("./utils/keysOf");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PRESERVE_KEYS = ['_id', '_type', '_upload'];
var EMPTY = [];

var errorCollector = (() => {
  var errorsByType = {};
  return {
    add: (type, value, error) => {
      if (!errorsByType[type.name]) {
        errorsByType[type.name] = [];
      }

      errorsByType[type.name].push({
        error: error,
        type: type,
        value
      });
    },

    getAll() {
      return errorsByType;
    },

    clear() {
      errorsByType = {};
    }

  };
})();

var reportErrors = (0, _debounce2.default)(() => {
  /* eslint-disable no-console */
  var errorsByType = errorCollector.getAll();
  var uniqueErrors = (0, _flatten2.default)(Object.keys(errorsByType).map(typeName => {
    var entries = errorsByType[typeName];
    return (0, _uniqBy2.default)(entries, entry => entry.error.message);
  }));
  var errorCount = uniqueErrors.length;

  if (errorCount === 0) {
    return;
  }

  console.groupCollapsed("%cHeads up! Got ".concat(errorCount === 1 ? 'error' : "".concat(errorCount, " errors"), " while preparing data for preview. Click for details."), 'color: #ff7e7c');
  Object.keys(errorsByType).forEach(typeName => {
    var entries = errorsByType[typeName];
    var first = entries[0];
    console.group("Check the preview config for schema type \"".concat(typeName, "\": %o"), first.type.preview);
    var uniqued = (0, _uniqBy2.default)(entries, entry => entry.error.message);
    uniqued.forEach(entry => {
      if (entry.error.type === 'returnValueError') {
        var hasPrepare = typeof entry.type.preview.prepare === 'function';
        var value = entry.value,
            error = entry.error;
        console.log("Encountered an invalid ".concat(hasPrepare ? 'return value when calling prepare(%o)' : 'value targeted by preview.select', ":"), value);
        console.error(error);
      }

      if (entry.error.type === 'prepareError') {
        var _value = entry.value,
            _error = entry.error;
        console.log('Encountered an error when calling prepare(%o):', _value);
        console.error(_error);
      }
    });
    console.groupEnd();
  });
  console.groupEnd();
  errorCollector.clear();
  /* eslint-enable no-console */
}, 1000);

var isRenderable = fieldName => value => {
  var type = typeof value;

  if (value === null || type === 'undefined' || type === 'string' || type === 'number' || type === 'boolean') {
    return EMPTY;
  }

  return [assignType('returnValueError', new Error("The \"".concat(fieldName, "\" field should be a string, number, boolean, undefined or null, instead saw ").concat(inspect(value))))];
};

var FIELD_NAME_VALIDATORS = {
  media: () => {
    // not sure how to validate media as it would  possibly involve executing a function and check the
    // return value
    return EMPTY;
  },
  title: isRenderable('title'),
  subtitle: isRenderable('subtitle'),
  description: isRenderable('description'),
  imageUrl: isRenderable('imageUrl'),
  date: isRenderable('date')
};

function inspect(val) {
  var prefixType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

  if ((0, _isPlainObject2.default)(val)) {
    var keys = Object.keys(val);
    var ellipse = keys.length > 3 ? '...' : '';
    var prefix = "object with keys ";
    return "".concat(prefixType ? prefix : '', "{").concat(keys.slice(0, 3).join(', ')).concat(ellipse, "}");
  }

  if (Array.isArray(val)) {
    var _ellipse = val.length > 3 ? '...' : '';

    var _prefix = "array with ";
    return "".concat(prefixType ? _prefix : '', "[").concat(val.map(v => inspect(v, false))).concat(_ellipse, "]");
  }

  return "the ".concat(typeof val, " ").concat(val);
}

function validateFieldValue(fieldName, fieldValue) {
  if (typeof fieldValue === 'undefined') {
    return EMPTY;
  }

  var validator = FIELD_NAME_VALIDATORS[fieldName];
  return validator && validator(fieldValue) || EMPTY;
}

function assignType(type, error) {
  return Object.assign(error, {
    type
  });
}

function validatePreparedValue(preparedValue) {
  if (!(0, _isPlainObject2.default)(preparedValue)) {
    return [assignType('returnValueError', new Error("Invalid return value. Expected a plain object with at least a 'title' field, instead saw ".concat(inspect(preparedValue))))];
  }

  return Object.keys(preparedValue).reduce((acc, fieldName) => {
    return [...acc, ...validateFieldValue(fieldName, preparedValue[fieldName])];
  }, EMPTY);
}

function validateReturnedPreview(result) {
  return _objectSpread(_objectSpread({}, result), {}, {
    errors: [...(result.errors || []), ...validatePreparedValue(result.returnValue)]
  });
}

function defaultPrepare(value) {
  return (0, _keysOf.keysOf)(value).reduce((acc, fieldName) => {
    var val = value[fieldName];
    return _objectSpread(_objectSpread({}, acc), {}, {
      [fieldName]: (0, _portableText.isPortableTextArray)(val) ? (0, _portableText.extractTextFromBlocks)(val) : value[fieldName]
    });
  }, {});
}

function invokePrepare(type, value) {
  var _type$preview;

  var viewOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var prepare = (_type$preview = type.preview) === null || _type$preview === void 0 ? void 0 : _type$preview.prepare;

  try {
    return {
      returnValue: prepare ? prepare(value, viewOptions) : defaultPrepare(value),
      errors: EMPTY
    };
  } catch (error) {
    return {
      returnValue: null,
      errors: [assignType('prepareError', error)]
    };
  }
}

function withErrors(result, type, selectedValue) {
  result.errors.forEach(error => errorCollector.add(type, selectedValue, error));
  reportErrors();
  return _constants.INVALID_PREVIEW_FALLBACK;
}

function hasEnumListOptions(type) {
  var options = type.options && typeof type.options === 'object' ? type.options : false;

  if (!options || !('list' in options)) {
    return false;
  }

  var listOptions = options.list;
  return Array.isArray(listOptions);
}

function getListOptions(type) {
  if (!hasEnumListOptions(type)) {
    return undefined;
  }

  var listOptions = type.options.list;
  return listOptions.map(option => (0, _types.isTitledListValue)(option) ? option : {
    title: option,
    value: option
  });
}

function prepareForPreview(rawValue, type) {
  var _type$preview2, _type$preview3;

  var viewOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var hasCustomPrepare = typeof ((_type$preview2 = type.preview) === null || _type$preview2 === void 0 ? void 0 : _type$preview2.prepare) === 'function';
  var selection = ((_type$preview3 = type.preview) === null || _type$preview3 === void 0 ? void 0 : _type$preview3.select) || {};
  var targetKeys = Object.keys(selection);
  var selectedValue = targetKeys.reduce((acc, key) => {
    var _typeWithFields$field;

    // Find the field the value belongs to
    var typeWithFields = 'fields' in type ? type : null;
    var targetFieldName = selection[key];
    var valueField = typeWithFields === null || typeWithFields === void 0 ? void 0 : (_typeWithFields$field = typeWithFields.fields) === null || _typeWithFields$field === void 0 ? void 0 : _typeWithFields$field.find(f => f.name === targetFieldName);
    var listOptions = valueField && getListOptions(valueField.type); // If the user has _not_ specified a `prepare()` function for the preview, and the
    // field type has an `options.list`, we want to use the title of the selected item
    // as the preview value. If, however, there _is_ a custom `prepare()`, we leave this
    // mapping up to the user to perform should they want to. This is both to maintain
    // backwards compatiblity, but also to allow using the raw value for prepare operations

    if (!hasCustomPrepare && listOptions) {
      // Find the selected option that matches the raw value
      var selectedOption = listOptions && listOptions.find(opt => opt.value === (0, _get2.default)(rawValue, selection[key]));
      acc[key] = selectedOption ? selectedOption.title : (0, _get2.default)(rawValue, selection[key]);
    } else {
      acc[key] = (0, _get2.default)(rawValue, selection[key]);
    }

    return acc;
  }, {});
  var prepareResult = invokePrepare(type, selectedValue, viewOptions);

  if (prepareResult.errors.length > 0) {
    return withErrors(prepareResult, type, selectedValue);
  }

  var returnValueResult = validateReturnedPreview(invokePrepare(type, selectedValue, viewOptions));
  return returnValueResult.errors.length > 0 ? withErrors(returnValueResult, type, selectedValue) : _objectSpread(_objectSpread({}, (0, _pick2.default)(rawValue, PRESERVE_KEYS)), prepareResult.returnValue);
}
