// @sanity/field/lib/types/portableText/diff/components/Header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Header;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var headingSizes = {
  h1: 2,
  h2: 1,
  h3: 0,
  h4: 0,
  h5: 0,
  h6: 0
};
var StyledHeading = (0, _styledComponents.default)(_ui.Heading)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &:not([hidden]) {\n    display: inline;\n    text-transform: none;\n    margin: 0;\n\n    &::before,\n    &::after {\n      content: unset;\n    }\n  }\n"])));

function Header(_ref) {
  var style = _ref.style,
      children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(StyledHeading, {
    size: headingSizes[style]
  }, children);
}
