// @sanity/field/lib/types/portableText/diff/components/PortableText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PortableText;

var _xor2 = _interopRequireDefault(require("lodash/xor"));

var _uniq2 = _interopRequireDefault(require("lodash/uniq"));

var _startCase2 = _interopRequireDefault(require("lodash/startCase"));

var _react = _interopRequireDefault(require("react"));

var _diff = require("../../../../diff");

var TextSymbols = _interopRequireWildcard(require("../symbols"));

var _helpers = require("../helpers");

var _Block = _interopRequireDefault(require("./Block"));

var _Annotation = require("./Annotation");

var _Decorator = _interopRequireDefault(require("./Decorator"));

var _InlineObject = require("./InlineObject");

var _Text = require("./Text");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var decoratorSymbolsStart = TextSymbols.DECORATOR_SYMBOLS.map(set => set[0]);
var decoratorSymbolsEnd = TextSymbols.DECORATOR_SYMBOLS.map(set => set[1]);
var annotationSymbolsStart = TextSymbols.ANNOTATION_SYMBOLS.map(set => set[0]);
var annotationSymbolsEnd = TextSymbols.ANNOTATION_SYMBOLS.map(set => set[1]);
var allSymbolsStart = decoratorSymbolsStart.concat(annotationSymbolsStart);
var allSymbolsEnd = decoratorSymbolsEnd.concat(annotationSymbolsEnd);
var allDecoratorSymbols = decoratorSymbolsStart.concat(decoratorSymbolsEnd);
var markRegex = new RegExp("".concat(allDecoratorSymbols.concat(allSymbolsEnd).join('|')), 'g');

function PortableText(props) {
  var diff = props.diff,
      schemaType = props.schemaType;
  var block = diff.origin.toValue || diff.origin.fromValue;
  var inlineObjects = diff.origin.toValue ? (0, _helpers.getInlineObjects)(diff.origin) : [];

  var renderChild = ptDiffChild => {
    var spanSchemaType = (0, _helpers.getChildSchemaType)(schemaType.fields, ptDiffChild);
    var decoratorTypes = [];

    if (spanSchemaType) {
      decoratorTypes = (0, _helpers.getDecorators)(spanSchemaType);
      var childrenDiff = diff.fields.children;
      var segments = childrenDiff.items[0].diff && childrenDiff.items[0].diff.type === 'object' && childrenDiff.items[0].diff.fields.text.type === 'string' && childrenDiff.items[0].diff.fields.text.segments || [];
      var returnedChildren = [];
      var annotationSegments = {}; // Special case for new empty PT-block (single span child with empty text)

      if (isEmptyTextChange(block, diff) && (diff.origin.action === 'added' || diff.origin.action === 'removed')) {
        var textDiff = (0, _helpers.findChildDiff)(diff.origin, block.children[0]) || diff.origin;

        if (textDiff && textDiff.action !== 'unchanged') {
          return /*#__PURE__*/_react.default.createElement(_diff.DiffCard, {
            annotation: textDiff.annotation,
            as: textDiff.action === 'removed' ? 'del' : 'ins',
            key: "empty-block-".concat(block._key),
            tooltip: {
              description: "".concat((0, _startCase2.default)(textDiff.action), " empty text")
            }
          }, /*#__PURE__*/_react.default.createElement("span", null, TextSymbols.EMPTY_BLOCK_SYMBOL));
        }
      } // Run through all the segments from the PortableTextDiff


      var childToIndex = -1;
      var segIndex = -1;
      var activeAnnotations = [];
      var endedAnnotation;
      var allMarkDefs = (0, _helpers.getAllMarkDefs)(diff.origin);
      segments.forEach(seg => {
        segIndex++;
        var isInline = TextSymbols.INLINE_SYMBOLS.includes(seg.text);
        var isMarkStart = allSymbolsStart.includes(seg.text);
        var isMarkEnd = allSymbolsEnd.includes(seg.text);
        var isChildStart = seg.text === TextSymbols.CHILD_SYMBOL;
        var isRemoved = seg.action === 'removed';

        if (isChildStart) {
          if (!isRemoved) {
            childToIndex++;
          } // No output

        } else if (isMarkStart || isMarkEnd) {
          if (isMarkStart && annotationSymbolsStart.includes(seg.text)) {
            var object = allMarkDefs[annotationSymbolsStart.indexOf(seg.text)];

            if (object) {
              activeAnnotations.push({
                mark: object._key,
                symbols: [seg.text, annotationSymbolsEnd[annotationSymbolsStart.indexOf(seg.text)]],
                object
              });
            }
          }

          if (isMarkEnd && annotationSymbolsEnd.includes(seg.text)) {
            endedAnnotation = activeAnnotations.pop();
          } // No output

        } else if (isInline) {
          var _inlineObjects$indexO;

          // Render inline object
          var indexOfSymbol = TextSymbols.INLINE_SYMBOLS.findIndex(sym => sym === seg.text);
          var key = (_inlineObjects$indexO = inlineObjects[indexOfSymbol]) === null || _inlineObjects$indexO === void 0 ? void 0 : _inlineObjects$indexO._key;
          var originChild = inlineObjects[indexOfSymbol];

          if (key) {
            var objectSchemaType = (0, _helpers.getChildSchemaType)(schemaType.fields, originChild);
            var objectDiff = (0, _helpers.findChildDiff)(diff.origin, originChild);
            returnedChildren.push( /*#__PURE__*/_react.default.createElement(_InlineObject.InlineObject, {
              key: "inline-object-".concat(originChild._key),
              object: originChild,
              path: [{
                _key: block._key
              }, 'children', {
                _key: originChild._key
              }],
              diff: objectDiff,
              schemaType: objectSchemaType
            }));
          }
        } else if (seg.text) {
          var _childDiff$fields, _childDiff$fields2;

          // TODO: find a better way of getting a removed child
          var getChildFromFromValue = () => {
            var _diff$origin$fromValu;

            return (_diff$origin$fromValu = diff.origin.fromValue) === null || _diff$origin$fromValu === void 0 ? void 0 : _diff$origin$fromValu.children.find(cld => cld.text && cld.text.match((0, _helpers.escapeRegExp)(seg.text)));
          };

          var child = block.children[childToIndex] || getChildFromFromValue();
          var childDiff = child && (0, _helpers.findSpanDiffFromChild)(diff.origin, child);

          if (!child) {
            throw new Error('Could not find child');
          }

          var _textDiff = childDiff !== null && childDiff !== void 0 && (_childDiff$fields = childDiff.fields) !== null && _childDiff$fields !== void 0 && _childDiff$fields.text ? childDiff === null || childDiff === void 0 ? void 0 : (_childDiff$fields2 = childDiff.fields) === null || _childDiff$fields2 === void 0 ? void 0 : _childDiff$fields2.text : undefined;

          var text = /*#__PURE__*/_react.default.createElement(_Text.Text, {
            diff: _textDiff,
            key: "text-".concat(child._key, "-").concat(segIndex),
            path: [{
              _key: block._key
            }, 'children', {
              _key: child._key
            }],
            childDiff: childDiff,
            segment: seg
          }, renderTextSegment({
            diff,
            child,
            decoratorTypes,
            seg,
            segIndex,
            spanSchemaType
          })); // Render annotations text changes within the annotation child


          if (activeAnnotations.length > 0) {
            activeAnnotations.forEach(active => {
              annotationSegments[active.mark] = annotationSegments[active.mark] || [];
              annotationSegments[active.mark].push(text);
            });
          }

          if (endedAnnotation) {
            var _key = "annotation-".concat(endedAnnotation.object._key);

            var lastChild = returnedChildren[returnedChildren.length - 1];

            if (lastChild && lastChild.key !== _key) {
              var annotationDiff = (0, _helpers.findAnnotationDiff)(diff.origin, endedAnnotation.mark);

              var _objectSchemaType = endedAnnotation && spanSchemaType.annotations && spanSchemaType.annotations.find(type => endedAnnotation && endedAnnotation.object && type.name === endedAnnotation.object._type);

              returnedChildren.push( /*#__PURE__*/_react.default.createElement(_Annotation.Annotation, {
                object: endedAnnotation.object,
                diff: annotationDiff,
                path: [{
                  _key: block._key
                }, 'children', {
                  _key: child._key
                }],
                schemaType: _objectSchemaType,
                key: _key
              }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, annotationSegments[endedAnnotation.mark])));
            } // delete annotationSegments[endedAnnotation.mark]


            endedAnnotation = undefined;
          }

          if (activeAnnotations.length === 0) {
            returnedChildren.push(text);
          }
        } // end if seg.text

      });
      return /*#__PURE__*/_react.default.createElement('div', {
        key: block._key
      }, ...returnedChildren);
    }

    throw new Error("'span' schemaType not found");
  };

  return /*#__PURE__*/_react.default.createElement(_Block.default, {
    block: diff.displayValue,
    diff: diff
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (diff.displayValue.children || []).map(child => renderChild(child))));
}

function renderTextSegment(_ref) {
  var diff = _ref.diff,
      child = _ref.child,
      decoratorTypes = _ref.decoratorTypes,
      seg = _ref.seg,
      segIndex = _ref.segIndex,
      spanSchemaType = _ref.spanSchemaType;

  // Newlines
  if (seg.text === '\n') {
    return /*#__PURE__*/_react.default.createElement("br", null);
  } // Make sure we render trailing spaces correctly


  var children = /*#__PURE__*/_react.default.createElement("span", {
    key: "text-".concat(segIndex)
  }, seg.text.replace(/ /g, TextSymbols.TRAILING_SPACE_SYMBOL));

  var spanDiff = child && (0, _helpers.findSpanDiffFromChild)(diff.origin, child); // Render decorator diff info

  var activeMarks = child.marks || [];

  if (spanDiff) {
    children = renderDecorators({
      activeMarks,
      decoratorTypes,
      diff,
      children,
      seg,
      segIndex,
      spanDiff,
      spanSchemaType
    });
  } // Render the segment with the active marks


  if (activeMarks && activeMarks.length > 0) {
    activeMarks.forEach(mark => {
      if ((0, _helpers.isDecorator)(mark, spanSchemaType)) {
        children =
        /*#__PURE__*/
        // eslint-disable-next-line react/no-array-index-key
        _react.default.createElement(_Decorator.default, {
          mark: mark,
          key: "decorator-".concat(mark, "-").concat(child._key, "-").concat(segIndex)
        }, children);
      }
    });
  }

  return children;
}

function renderDecorators(_ref2) {
  var _spanDiff$fields;

  var activeMarks = _ref2.activeMarks,
      decoratorTypes = _ref2.decoratorTypes,
      diff = _ref2.diff,
      children = _ref2.children,
      seg = _ref2.seg,
      segIndex = _ref2.segIndex,
      spanDiff = _ref2.spanDiff,
      spanSchemaType = _ref2.spanSchemaType;

  var returned = /*#__PURE__*/_react.default.createElement("span", {
    key: "text-segment-".concat(segIndex)
  }, children);

  var fromPtDiffText = diff.origin.fromValue && diff.fromValue && diff.fromValue.children[0].text || ''; // Always one child
  // There are cases where we have changed marks, but it's an indirect change in the diff data.
  // For example when '<>normal-text</><>bold-text</>' and 'bold' is unbolded. Then 'bold' is added to first span,
  // and 'bold' is removed from the second span: '<>normal-text-bold</><>-text</>'. No marks are changed.
  // We do however want to indicate to the user that someone removed bold from 'bold'
  // In these cases, fallback to the diff annotation information in the span itself.

  var indirectMarksAnnotation = spanDiff && spanDiff.action !== 'unchanged' && spanDiff.annotation || undefined;
  var marksDiff = spanDiff === null || spanDiff === void 0 ? void 0 : (_spanDiff$fields = spanDiff.fields) === null || _spanDiff$fields === void 0 ? void 0 : _spanDiff$fields.marks;
  var marksAnnotation = marksDiff && marksDiff.action !== 'unchanged' && marksDiff.annotation || indirectMarksAnnotation;
  var marksChanged = [];
  var ptDiffChildren = fromPtDiffText.split(TextSymbols.CHILD_SYMBOL).filter(text => !!text).join('');
  var ptDiffMatchString = ptDiffChildren;
  var controlString = ptDiffMatchString.substring(0, ptDiffMatchString.indexOf(seg.text) + seg.text.length);
  var toTest = controlString.substring(0, controlString.indexOf(seg.text));
  var marks = [];
  var matches = [...toTest.matchAll(markRegex)];
  matches.forEach(match => {
    var sym = match[0];
    var set = TextSymbols.DECORATOR_SYMBOLS.concat(TextSymbols.ANNOTATION_SYMBOLS).find(aSet => aSet.indexOf(sym) > -1);

    if (set) {
      var _decoratorTypes;

      var isMarkStart = sym === set[0];
      var mark = ((_decoratorTypes = decoratorTypes[isMarkStart ? decoratorSymbolsStart.indexOf(sym) : decoratorSymbolsEnd.indexOf(sym)]) === null || _decoratorTypes === void 0 ? void 0 : _decoratorTypes.value) || sym; // Annotation marks are uniqe anyway

      var notClosed = toTest.lastIndexOf(sym) > toTest.lastIndexOf(set[1]);

      if (notClosed) {
        marks.push(mark);
      }
    }
  });
  marksChanged = (0, _xor2.default)(activeMarks, (0, _uniq2.default)(marks)); // Only for decorators, annotations are taken care of elsewhere

  if (marksAnnotation && marksChanged.length > 0 && marksChanged.some(m => (0, _helpers.isDecorator)(m, spanSchemaType))) {
    returned = /*#__PURE__*/_react.default.createElement(_diff.DiffCard, {
      annotation: marksAnnotation,
      key: "diffcard-annotation-".concat(segIndex, "-").concat(marksChanged.join('-')),
      as: 'ins',
      tooltip: {
        description: 'Changed formatting'
      }
    }, returned);
  }

  return returned;
}

function isEmptyTextChange(block, diff) {
  return block.children.length === 1 && block.children[0]._type === 'span' && typeof block.children[0].text === 'string' && block.children[0].text === '' && diff.origin.action !== 'unchanged';
}
