// @sanity/form-builder/lib/inputs/PortableText/Objects/renderers/DefaultObjectEditing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultObjectEditing = DefaultObjectEditing;

var _react = _interopRequireDefault(require("react"));

var _presence = require("@sanity/base/presence");

var _default = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _FormBuilderInput = require("../../../../FormBuilderInput");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />

/* eslint-disable react/prop-types */
function DefaultObjectEditing(props) {
  var focusPath = props.focusPath,
      markers = props.markers,
      object = props.object,
      onBlur = props.onBlur,
      onChange = props.onChange,
      onClose = props.onClose,
      onFocus = props.onFocus,
      path = props.path,
      presence = props.presence,
      readOnly = props.readOnly,
      type = props.type;

  var handleChange = patchEvent => onChange(patchEvent, path);

  return /*#__PURE__*/_react.default.createElement(_default.default, {
    isOpen: true // onClickOutside={onClose}
    ,
    onClose: onClose,
    onEscape: onClose,
    showCloseButton: true,
    title: type.title,
    size: "medium"
  }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
    margins: [0, 0, 1, 0]
  }, /*#__PURE__*/_react.default.createElement(_FormBuilderInput.FormBuilderInput, {
    focusPath: focusPath,
    level: 0,
    markers: markers,
    onBlur: onBlur,
    onChange: handleChange,
    onFocus: onFocus,
    path: path,
    presence: presence,
    readOnly: readOnly || type.readOnly,
    type: type,
    value: object
  })));
}
