// @sanity/default-layout/lib/navbar/presenceMenu/PresenceListRow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresenceListRow = PresenceListRow;

var _orderBy2 = _interopRequireDefault(require("lodash/orderBy"));

var _components = require("@sanity/base/components");

var PathUtils = _interopRequireWildcard(require("@sanity/util/paths"));

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

var _router = require("@sanity/base/router");

var _react = _interopRequireDefault(require("react"));

var _PresenceListRow = _interopRequireDefault(require("./PresenceListRow.css"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />

/* eslint-disable react/require-default-props */
function PresenceListRow(props) {
  var _presence$user;

  var presence = props.presence,
      onClose = props.onClose;
  var lastActiveLocation = (0, _orderBy2.default)(presence.locations || [], ['lastActiveAt'], ['desc']).find(location => location.documentId);
  var hasLink = Boolean(lastActiveLocation === null || lastActiveLocation === void 0 ? void 0 : lastActiveLocation.documentId);

  var item = /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceListRow.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceListRow.default.avatar
  }, /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
    user: presence.user,
    size: "medium"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceListRow.default.inner
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceListRow.default.userName
  }, presence.user.displayName)), hasLink && /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceListRow.default.linkIcon
  }, /*#__PURE__*/_react.default.createElement(_linkIcon.default, null)));

  return lastActiveLocation ? /*#__PURE__*/_react.default.createElement(_router.IntentLink, {
    title: (presence === null || presence === void 0 ? void 0 : (_presence$user = presence.user) === null || _presence$user === void 0 ? void 0 : _presence$user.displayName) && "Go to ".concat(presence.user.displayName),
    className: _PresenceListRow.default.intentLink,
    intent: "edit",
    params: {
      id: lastActiveLocation.documentId,
      path: encodeURIComponent(PathUtils.toString(lastActiveLocation.path))
    },
    onClick: onClose
  }, item) : item;
}
