// @sanity/base/lib/datastores/document/document-pair/snapshotPair.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.snapshotPair = void 0;

var _operators = require("rxjs/operators");

var _createMemoizer = require("../utils/createMemoizer");

var _memoizedPair = require("./memoizedPair");

// return true if the event comes with a document snapshot
function isSnapshotEvent(event) {
  return event.type === 'snapshot';
}

function withSnapshots(pair) {
  return {
    snapshots$: pair.events.pipe((0, _operators.filter)(isSnapshotEvent), (0, _operators.map)(event => event.document), (0, _operators.publishReplay)(1), (0, _operators.refCount)()),
    patch: pair.patch,
    create: pair.create,
    createIfNotExists: pair.createIfNotExists,
    createOrReplace: pair.createOrReplace,
    delete: pair.delete,
    mutate: pair.mutate,
    commit: pair.commit
  };
}

var snapshotPair = (0, _createMemoizer.memoize)(idPair => {
  return (0, _memoizedPair.memoizedPair)(idPair).pipe((0, _operators.map)(_ref => {
    var published = _ref.published,
        draft = _ref.draft;
    return {
      published: withSnapshots(published),
      draft: withSnapshots(draft)
    };
  }), (0, _operators.publishReplay)(1), (0, _operators.refCount)());
}, idPair => idPair.publishedId);
exports.snapshotPair = snapshotPair;
