// @sanity/field/lib/diff/components/DiffInspectWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffInspectWrapper = DiffInspectWrapper;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _paths = require("../../paths");

var _FromToArrow = require("./FromToArrow");

var _excluded = ["children", "as", "change"];

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var CodeWrapper = _styledComponents.default.pre(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  overflow-x: auto;\n  position: relative;\n"])));

var Meta = _styledComponents.default.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  right: 0;\n"])));

function DiffInspectWrapper(_ref) {
  var children = _ref.children,
      as = _ref.as,
      change = _ref.change,
      restProps = _objectWithoutProperties(_ref, _excluded);

  var isHovering = (0, _react.useRef)(false);

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      isInspecting = _useState2[0],
      setIsInspecting = _useState2[1];

  var toggleInspect = (0, _react.useCallback)(() => setIsInspecting(state => !state), [setIsInspecting]);
  var handleMouseEnter = (0, _react.useCallback)(() => isHovering.current = true, []);
  var handleMouseLeave = (0, _react.useCallback)(() => isHovering.current = false, [isHovering]);
  (0, _react.useEffect)(() => {
    function onKeyDown(evt) {
      var metaKey = evt.metaKey,
          key = evt.key;

      if (metaKey && key === 'i' && isHovering.current) {
        toggleInspect();
      }
    }

    window.addEventListener('keydown', onKeyDown, false);
    return () => window.removeEventListener('keydown', onKeyDown, false);
  }, [toggleInspect]);
  return /*#__PURE__*/_react.default.createElement(_ui.Box, _extends({
    as: as,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave
  }, restProps), isInspecting ? /*#__PURE__*/_react.default.createElement(DiffInspector, {
    change: change
  }) : children);
}

var MetaLabel = _ref2 => {
  var title = _ref2.title;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 3,
    display: "inline-block",
    as: Meta
  }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 1,
    muted: true
  }, title));
};

function DiffInspector(_ref3) {
  var _change$itemDiff, _change$itemDiff2, _change$itemDiff3;

  var change = _ref3.change;
  return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    tone: "transparent",
    as: CodeWrapper,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(MetaLabel, {
    title: "meta"
  }), /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json",
    size: 1
  }, printMeta({
    path: (0, _paths.pathToString)(change.path),
    fromIndex: (_change$itemDiff = change.itemDiff) === null || _change$itemDiff === void 0 ? void 0 : _change$itemDiff.fromIndex,
    toIndex: (_change$itemDiff2 = change.itemDiff) === null || _change$itemDiff2 === void 0 ? void 0 : _change$itemDiff2.toIndex,
    hasMoved: (_change$itemDiff3 = change.itemDiff) === null || _change$itemDiff3 === void 0 ? void 0 : _change$itemDiff3.hasMoved,
    action: change.diff.action,
    isChanged: change.diff.isChanged
  }))), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: CodeWrapper,
    tone: "critical",
    padding: 3,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(MetaLabel, {
    title: "from"
  }), /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json",
    size: 1
  }, jsonify(change.diff.fromValue))), /*#__PURE__*/_react.default.createElement(_ui.Card, null, /*#__PURE__*/_react.default.createElement(_FromToArrow.FromToArrow, {
    direction: "down",
    align: "center"
  })), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: CodeWrapper,
    tone: "positive",
    padding: 3,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(MetaLabel, {
    title: "to"
  }), /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json",
    size: 1
  }, jsonify(change.diff.toValue))));
}

function jsonify(value) {
  if (typeof value === 'undefined') {
    return 'undefined';
  }

  return JSON.stringify(value, null, 2);
}

function printMeta(keys) {
  var lines = [];
  Object.entries(keys).forEach(_ref4 => {
    var _ref5 = _slicedToArray(_ref4, 2),
        key = _ref5[0],
        value = _ref5[1];

    if (typeof value !== 'undefined' && value !== null) {
      lines.push("".concat(key, ": ").concat(value));
    }
  });
  return lines.join('\n');
}
