// @sanity/field/lib/diff/components/FallbackDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FallbackDiff = void 0;

var _react = _interopRequireDefault(require("react"));

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _ui = require("@sanity/ui");

var _DiffFromTo = require("./DiffFromTo");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var FallbackPreview = _ref => {
  var value = _ref.value,
      schemaType = _ref.schemaType;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_preview.default, {
    type: schemaType,
    value: value,
    layout: "default"
  }));
};

var FallbackDiff = _ref2 => {
  var diff = _ref2.diff,
      schemaType = _ref2.schemaType;
  return /*#__PURE__*/_react.default.createElement(_DiffFromTo.DiffFromTo, {
    diff: diff,
    schemaType: schemaType,
    previewComponent: FallbackPreview,
    layout: "grid"
  });
};

exports.FallbackDiff = FallbackDiff;
