// @sanity/field/lib/diff/components/MetaInfo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetaInfo = MetaInfo;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var MetaText = (0, _styledComponents.default)(_ui.Text)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  color: inherit;\n"])));

function MetaInfo(_ref) {
  var title = _ref.title,
      action = _ref.action,
      icon = _ref.icon,
      children = _ref.children,
      markRemoved = _ref.markRemoved;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 2,
    align: "center"
  }, icon && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(MetaText, {
    size: 4,
    forwardedAs: markRemoved ? 'del' : 'div'
  }, /*#__PURE__*/(0, _react.createElement)(icon))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2,
    paddingLeft: 2
  }, /*#__PURE__*/_react.default.createElement(MetaText, {
    size: 1,
    weight: "semibold",
    forwardedAs: markRemoved ? 'del' : 'h3',
    textOverflow: "ellipsis"
  }, title), action && /*#__PURE__*/_react.default.createElement("div", null, action), /*#__PURE__*/_react.default.createElement(MetaText, {
    size: 0,
    textOverflow: "ellipsis"
  }, children)));
}
