// @sanity/base/lib/__legacy/@sanity/components/utilities/CaptureOutsideClicks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _excluded = ["wrapperElement", "onClickOutside"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// @todo: refactor to functional component
class CaptureOutsideClicks extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "hadMouseDown", false);

    _defineProperty(this, "_wrapperElement", null);

    _defineProperty(this, "handleMouseDown", event => {
      if (this._wrapperElement && this._wrapperElement.contains(event.target)) {
        this.hadMouseDown = true;
      }
    });

    _defineProperty(this, "handleDocumentClick", event => {
      if (this.props.onClickOutside && this._wrapperElement && event.target instanceof Node && !this._wrapperElement.contains(event.target) && !this.hadMouseDown) {
        this.props.onClickOutside(event);
      }

      this.hadMouseDown = false;
    });

    _defineProperty(this, "setWrapperElement", element => {
      this._wrapperElement = element;
    });
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    document.addEventListener('mouseup', this.handleDocumentClick);
    document.addEventListener('mousedown', this.handleMouseDown);
  }

  componentWillUnmount() {
    document.removeEventListener('mouseup', this.handleDocumentClick);
    document.removeEventListener('mousedown', this.handleMouseDown);
  }

  render() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    var _this$props = this.props,
        _this$props$wrapperEl = _this$props.wrapperElement,
        wrapperElement = _this$props$wrapperEl === void 0 ? 'div' : _this$props$wrapperEl,
        _ = _this$props.onClickOutside,
        restProps = _objectWithoutProperties(_this$props, _excluded);

    return /*#__PURE__*/_react.default.createElement(wrapperElement, _objectSpread(_objectSpread({}, restProps), {}, {
      ref: this.setWrapperElement
    }));
  }

}

exports.default = CaptureOutsideClicks;
