// @sanity/base/lib/__legacy/@sanity/components/utilities/ImageLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _loadImage = require("./loadImage");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// @todo: refactor to functional component
class ImageLoader extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      loadedImage: null,
      error: null
    });

    _defineProperty(this, "subscription", void 0);
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    this.loadImage(this.props.src);
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  unsubscribe() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  loadImage(src) {
    this.unsubscribe();
    this.subscription = (0, _loadImage.loadImage)(src).subscribe({
      next: url => this.setState({
        loadedImage: url,
        error: null
      }),
      error: _error => this.setState({
        loadedImage: null,
        error: _error
      })
    });
  } // eslint-disable-next-line camelcase


  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.src !== this.props.src) {
      this.loadImage(nextProps.src);
    }
  }

  render() {
    var children = this.props.children;
    var _this$state = this.state,
        error = _this$state.error,
        loadedImage = _this$state.loadedImage;
    return error || loadedImage ? children({
      image: loadedImage,
      error
    }) : null;
  }

}

exports.default = ImageLoader;
