// @sanity/base/lib/__legacy/@sanity/components/utilities/Stacked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStack = createStack;
exports.default = void 0;

var _nanoPubsub = _interopRequireDefault(require("nano-pubsub"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function createStack() {
  var stack = [];
  var changes = (0, _nanoPubsub.default)();

  function remove(instance) {
    stack = stack.filter(entry => entry !== instance);
    onChange();
  }

  function peek() {
    return stack[stack.length - 1];
  }

  function push(entry) {
    stack.push(entry);
    onChange();
  }

  function onChange() {
    changes.publish(peek());
  }

  return {
    remove,
    peek,
    push,
    subscribe: changes.subscribe
  };
}

var DEFAULT_STACK = createStack(); // @todo: refactor to functional component

class Stacked extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "state", {
      top: null
    });

    _defineProperty(this, "_unsubscribe", void 0);

    var stack = props.stack;

    if (stack && stack.subscribe) {
      this._unsubscribe = stack.subscribe(top => {
        this.setState({
          top
        });
      });
    }
  } // eslint-disable-next-line camelcase


  UNSAFE_componentWillMount() {
    var stack = this.props.stack;
    if (stack && stack.push) stack.push(this);
  }

  componentWillUnmount() {
    var stack = this.props.stack;
    if (this._unsubscribe) this._unsubscribe();

    if (stack && stack.remove) {
      stack.remove(this);
    }
  }

  render() {
    return this.props.children(this.state.top === this);
  }

}

exports.default = Stacked;

_defineProperty(Stacked, "defaultProps", {
  stack: DEFAULT_STACK
});
