// @sanity/logos/lib/cjs/groqLogo.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  GroqLogo: () => GroqLogo
});
var import_color = __toModule(require("@sanity/color"));
var import_react = __toModule(require("react"));
const GroqLogo = (0, import_react.forwardRef)(function GroqLogo2(props, ref) {
  return /* @__PURE__ */ import_react.default.createElement("svg", __spreadProps(__spreadValues({
    "data-sanity-icon": "groq-logo",
    height: "1em",
    viewBox: "0 0 304 64",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), {
    ref
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M32 0L0 32H32V0Z",
    fill: import_color.hues.magenta[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M32 32H0L32 64V32Z",
    fill: import_color.hues.green[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M64 32H32V64L64 32Z",
    fill: import_color.hues.purple[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M112 0L80 32H112V0Z",
    fill: import_color.hues.green[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M112 0H80V32L112 0Z",
    fill: import_color.hues.purple[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M112 32H80L112 64V32Z",
    fill: import_color.hues.yellow[200].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M144 64L112 32V64H144Z",
    fill: import_color.hues.blue[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M80 32V64H112L80 32Z",
    fill: import_color.hues.orange[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M112 0V32H144L112 0Z",
    fill: import_color.hues.magenta[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M192 0L160 32H192V0Z",
    fill: import_color.hues.yellow[200].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M192 32H160L192 64V32Z",
    fill: import_color.hues.orange[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M224 32H192V64L224 32Z",
    fill: import_color.hues.green[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M192 0V32H224L192 0Z",
    fill: import_color.hues.blue[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M272 0L240 32H272V0Z",
    fill: import_color.hues.purple[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M304 32L272 64H304V32Z",
    fill: import_color.hues.magenta[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M272 32H240L272 64V32Z",
    fill: import_color.hues.blue[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M304 32H272V64L304 32Z",
    fill: import_color.hues.yellow[200].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M272 0V32H304L272 0Z",
    fill: import_color.hues.green[300].hex
  }));
});
