// @sanity/logos/lib/cjs/groqMonogram.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  GroqMonogram: () => GroqMonogram
});
var import_color = __toModule(require("@sanity/color"));
var import_react = __toModule(require("react"));
const GroqMonogram = (0, import_react.forwardRef)(function GroqMonogram2(props, ref) {
  return /* @__PURE__ */ import_react.default.createElement("svg", __spreadProps(__spreadValues({
    "data-sanity-icon": "groq-monogram",
    width: "1em",
    height: "1em",
    viewBox: "0 0 128 128",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), {
    ref
  }), /* @__PURE__ */ import_react.default.createElement("rect", {
    width: "128",
    height: "128",
    rx: "8",
    fill: import_color.hues.green[950].hex
  }), /* @__PURE__ */ import_react.default.createElement("rect", {
    y: "32",
    width: "64",
    height: "64",
    fill: import_color.hues.green[800].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M64 32L32 64H64V32Z",
    fill: import_color.hues.magenta[400].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M64 64H32L64 96V64Z",
    fill: import_color.hues.green[300].hex
  }), /* @__PURE__ */ import_react.default.createElement("path", {
    d: "M96 64H64V96L96 64Z",
    fill: import_color.hues.purple[300].hex
  }));
});
