// @sanity/portable-text-editor/lib/utils/weakMaps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KEY_TO_VALUE_ELEMENT = exports.KEY_TO_SLATE_ELEMENT = exports.IS_DRAGGING_BLOCK_TARGET_POSITION = exports.IS_DRAGGING_ELEMENT_RANGE = exports.IS_DRAGGING_ELEMENT_TARGET = exports.IS_DRAGGING_CHILD_ELEMENT = exports.IS_DRAGGING_BLOCK_ELEMENT = exports.IS_DRAGGING = void 0;

/**
 * Dragging
 */
// Is the editor dragging something?
var IS_DRAGGING = new WeakMap(); // Is the editor dragging a element?

exports.IS_DRAGGING = IS_DRAGGING;
var IS_DRAGGING_BLOCK_ELEMENT = new WeakMap();
exports.IS_DRAGGING_BLOCK_ELEMENT = IS_DRAGGING_BLOCK_ELEMENT;
var IS_DRAGGING_CHILD_ELEMENT = new WeakMap(); // When dragging elements, this will be the target element

exports.IS_DRAGGING_CHILD_ELEMENT = IS_DRAGGING_CHILD_ELEMENT;
var IS_DRAGGING_ELEMENT_TARGET = new WeakMap();
exports.IS_DRAGGING_ELEMENT_TARGET = IS_DRAGGING_ELEMENT_TARGET;
var IS_DRAGGING_ELEMENT_RANGE = new WeakMap(); // Target position for dragging over a block

exports.IS_DRAGGING_ELEMENT_RANGE = IS_DRAGGING_ELEMENT_RANGE;
var IS_DRAGGING_BLOCK_TARGET_POSITION = new WeakMap();
exports.IS_DRAGGING_BLOCK_TARGET_POSITION = IS_DRAGGING_BLOCK_TARGET_POSITION;
var KEY_TO_SLATE_ELEMENT = new WeakMap();
exports.KEY_TO_SLATE_ELEMENT = KEY_TO_SLATE_ELEMENT;
var KEY_TO_VALUE_ELEMENT = new WeakMap();
exports.KEY_TO_VALUE_ELEMENT = KEY_TO_VALUE_ELEMENT;
