// @sanity/desk-tool/lib/panes/userComponentPane/UserComponentPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default = _interopRequireDefault(require("part:@sanity/components/panes/default"));

var _UserComponentPane = _interopRequireDefault(require("./UserComponentPane.css"));

var _excluded = ["isSelected", "isCollapsed", "onCollapse", "onExpand", "component", "index", "styles", "title", "type", "menuItems", "menuItemGroups", "renderActions"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function noActionFn() {
  // eslint-disable-next-line no-console
  console.warn('No handler defined for action');
}

var EMPTY_ARRAY = [];
var EMPTY_RECORD = {};

class UserComponentPane extends _react.default.PureComponent {
  constructor(props) {
    super(props);

    _defineProperty(this, "handleAction", item => {
      var handler;

      if (typeof item.action === 'function') {
        handler = item.action;
      } else {
        handler = this.userComponent && this.userComponent.current && this.userComponent.current.actionHandlers && this.userComponent.current.actionHandlers[item.action];
      }

      if (handler) {
        handler(item.params, this);
      } else {
        noActionFn();
      }
    });

    this.userComponent = /*#__PURE__*/_react.default.createRef();
  }

  render() {
    var _this$props = this.props,
        isSelected = _this$props.isSelected,
        isCollapsed = _this$props.isCollapsed,
        onCollapse = _this$props.onCollapse,
        onExpand = _this$props.onExpand,
        component = _this$props.component,
        index = _this$props.index,
        styles = _this$props.styles,
        title = _this$props.title,
        type = _this$props.type,
        menuItems = _this$props.menuItems,
        menuItemGroups = _this$props.menuItemGroups,
        renderActions = _this$props.renderActions,
        rest = _objectWithoutProperties(_this$props, _excluded);

    var hideHeader = !title && !menuItems.length && !renderActions;
    var paneStyles = hideHeader ? {
      header: _UserComponentPane.default.noHeader
    } : {};
    var UserComponent = typeof component === 'function' && component;
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      styles: paneStyles,
      title: title,
      menuItems: menuItems,
      menuItemGroups: menuItemGroups,
      isSelected: isSelected,
      isCollapsed: isCollapsed,
      onCollapse: onCollapse,
      onExpand: onExpand,
      onAction: this.handleAction
    }, UserComponent ? /*#__PURE__*/_react.default.createElement(UserComponent, _extends({
      ref: this.userComponent
    }, rest)) : component);
  }

}

exports.default = UserComponentPane;

_defineProperty(UserComponentPane, "propTypes", {
  styles: _propTypes.default.object,
  // eslint-disable-line react/forbid-prop-types
  title: _propTypes.default.string,
  index: _propTypes.default.number.isRequired,
  type: _propTypes.default.string.isRequired,
  component: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]).isRequired,
  options: _propTypes.default.object,
  // eslint-disable-line react/forbid-prop-types
  isSelected: _propTypes.default.bool.isRequired,
  isCollapsed: _propTypes.default.bool.isRequired,
  onExpand: _propTypes.default.func,
  onCollapse: _propTypes.default.func,
  renderActions: _propTypes.default.func,
  menuItems: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string.isRequired
  })),
  menuItemGroups: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string.isRequired
  }))
});

_defineProperty(UserComponentPane, "defaultProps", {
  title: '',
  options: EMPTY_RECORD,
  menuItems: EMPTY_ARRAY,
  menuItemGroups: EMPTY_ARRAY,
  styles: undefined,
  onExpand: undefined,
  onCollapse: undefined,
  renderActions: undefined
});
