// @sanity/base/lib/preview/components/ObserveForPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactPropsStream = require("react-props-stream");

var _shallowEquals = _interopRequireDefault(require("shallow-equals"));

var _operators = require("rxjs/operators");

var _rxjs = require("rxjs");

var _ = require("../");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isNonNullable(value) {
  return value !== null && value !== undefined;
} // Will track a memo of the value as long as the isActive$ stream emits true,
// and emit the memoized value after it switches to to false
// (disclaimer: there's probably a better way to do this)


function memoizeBy(isActive$) {
  return producer$ => {
    var memo;
    return isActive$.pipe((0, _operators.distinctUntilChanged)(), (0, _operators.switchMap)(isActive => isActive ? producer$.pipe((0, _operators.tap)(v => memo = v)) : (0, _rxjs.of)(memo).pipe((0, _operators.filter)(isNonNullable))));
  };
}

var connect = props$ => {
  var sharedProps$ = props$.pipe((0, _operators.publishReplay)(1), (0, _operators.refCount)());
  var isActive$ = sharedProps$.pipe((0, _operators.map)(props => props.isActive !== false));
  return sharedProps$.pipe((0, _operators.distinctUntilChanged)((props, nextProps) => (0, _shallowEquals.default)(props.value, nextProps.value)), (0, _operators.switchMap)(props => (0, _rxjs.concat)((0, _rxjs.of)({
    isLoading: true,
    type: props.type,
    snapshot: null,
    children: props.children
  }), (0, _.observeForPreview)(props.value, props.type, props.ordering ? {
    ordering: props.ordering
  } : {}).pipe((0, _operators.map)(result => ({
    isLoading: false,
    type: props.type,
    snapshot: result.snapshot,
    children: props.children
  }))))), memoizeBy(isActive$));
};

var _default = (0, _reactPropsStream.withPropsStream)(connect, function ObserveForPreview(props) {
  var type = props.type,
      error = props.error,
      snapshot = props.snapshot,
      isLoading = props.isLoading,
      children = props.children;
  return children({
    error,
    isLoading,
    result: {
      type,
      snapshot
    }
  });
});

exports.default = _default;
