// @sanity/field/lib/types/array/diff/ArrayOfOptionsFieldDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayOfOptionsFieldDiff = void 0;

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _userColor = require("@sanity/base/user-color");

var _types = require("@sanity/types");

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _diff = require("../../../diff");

var _preview2 = require("../../boolean/preview");

var _arrayUtils = require("../util/arrayUtils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var ArrayOfOptionsFieldDiff = _ref => {
  var _options;

  var diff = _ref.diff,
      schemaType = _ref.schemaType;
  var options = (_options = schemaType.options) === null || _options === void 0 ? void 0 : _options.list;
  var colorManager = (0, _userColor.useUserColorManager)();

  if (!Array.isArray(options)) {
    // Shouldn't happen, because the resolver should only resolve here if it does
    return null;
  }

  return /*#__PURE__*/_react.default.createElement("div", null, diff.items.map(item => normalizeItems(item, diff, schemaType)).filter(item => item !== null).sort(sortItems).map((item, index) => {
    var annotation = item.annotation,
        isPresent = item.isPresent,
        value = item.value,
        memberType = item.memberType,
        title = item.title;
    var color = (0, _diff.getAnnotationColor)(colorManager, annotation);
    var action = isPresent ? 'Added' : 'Removed';
    return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center",
      key: getItemKey(diff, index)
    }, /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
      annotations: annotation ? [annotation] : [],
      description: action
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center"
    }, /*#__PURE__*/_react.default.createElement(_preview2.Checkbox, {
      checked: !isPresent,
      color: color
    }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      margin: 2
    }, /*#__PURE__*/_react.default.createElement(_diff.FromToArrow, null)), /*#__PURE__*/_react.default.createElement(_preview2.Checkbox, {
      checked: isPresent,
      color: color
    }))), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center"
    }, /*#__PURE__*/_react.default.createElement(ItemPreview, {
      value: title || value,
      memberType: memberType
    })));
  }));
};

exports.ArrayOfOptionsFieldDiff = ArrayOfOptionsFieldDiff;

function normalizeItems(item, parentDiff, schemaType) {
  if (item.diff.action === 'unchanged') {
    return null;
  }

  var fromValue = parentDiff.fromValue,
      toValue = parentDiff.toValue;
  var value = getValue(item.diff);
  var wasPresent = isInArray(value, fromValue);
  var isPresent = isInArray(value, toValue);

  if (wasPresent === isPresent) {
    return null;
  }

  return {
    title: getItemTitle(value, schemaType),
    memberType: resolveMemberType(getValue(item.diff), schemaType),
    itemIndex: getOptionIndex(value, schemaType),
    annotation: item.annotation,
    isPresent,
    value
  };
}

function sortItems(itemA, itemB) {
  return itemA.itemIndex - itemB.itemIndex;
}

function ItemPreview(_ref2) {
  var value = _ref2.value,
      memberType = _ref2.memberType;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginX: 2,
    marginY: 1
  }, typeof value === 'string' || typeof value === 'number' ? value : /*#__PURE__*/_react.default.createElement(_preview.default, {
    type: memberType,
    value: value,
    layout: "default"
  }));
}

function isInArray(value, parent) {
  var array = parent || [];
  return typeof value === 'object' && value !== null ? array.some(item => (0, _arrayUtils.isEqual)(item, value)) : array.includes(value);
}

function getItemKey(diff, index) {
  var value = diff.toValue || diff.fromValue;
  return (0, _types.isKeyedObject)(value) ? value._key : index;
}

function getValue(diff) {
  return typeof diff.toValue === 'undefined' ? diff.fromValue : diff.toValue;
}

function resolveMemberType(item, schemaType) {
  var itemTypeName = resolveTypeName(item);
  return schemaType.of.find(memberType => memberType.name === itemTypeName);
}

function resolveTypeName(value) {
  var jsType = resolveJSType(value);

  if (jsType !== 'object') {
    return jsType;
  }

  var obj = value;
  return '_type' in obj && obj._type || jsType;
}

function resolveJSType(val) {
  if (val === null) {
    return 'null';
  }

  if (Array.isArray(val)) {
    return 'array';
  }

  return typeof val;
}

function isNamedOption(item) {
  return typeof item === 'object' && item !== null && 'title' in item;
}

function getOptionIndex(item, schemaType) {
  var _options2;

  var list = ((_options2 = schemaType.options) === null || _options2 === void 0 ? void 0 : _options2.list) || [];
  return list.findIndex(opt => (0, _arrayUtils.isEqual)(isNamedOption(opt) ? opt.value : opt, item));
}

function getItemTitle(item, schemaType) {
  var _options3;

  var list = ((_options3 = schemaType.options) === null || _options3 === void 0 ? void 0 : _options3.list) || [];
  var index = getOptionIndex(item, schemaType);
  return index === -1 ? undefined : list[index].title || undefined;
}
