// @sanity/base/lib/__legacy/@sanity/components/progress/ProgressCircle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _circleStyle = _interopRequireDefault(require("part:@sanity/components/progress/circle-style"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var radiusFactor = 1.3;

class ProgressCircle extends _react.default.PureComponent {
  componentDidMount() {
    // eslint-disable-next-line no-console
    console.warn("ProgressCircle from `part:@sanity/components/progress/circle` is deprecated. Please use `import {ProgressCircle} from '@sanity/base/components'` instead");
  }

  render() {
    var _this$props = this.props,
        _this$props$percent = _this$props.percent,
        percent = _this$props$percent === void 0 ? 0 : _this$props$percent,
        text = _this$props.text,
        style = _this$props.style,
        showPercent = _this$props.showPercent,
        isComplete = _this$props.isComplete;
    var radius = 50;
    var strokeWidth = 10;
    var width = radius * 2;
    var height = radius * 2;
    var viewBox = "0 0 ".concat(width, " ").concat(height);
    var dashArray = radius * Math.PI * 2;
    var dashOffset = dashArray - dashArray * percent / 100;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: isComplete ? _circleStyle.default.completed : _circleStyle.default.unCompleted,
      style: style
    }, /*#__PURE__*/_react.default.createElement("svg", {
      className: _circleStyle.default.svg,
      width: width,
      height: height,
      viewBox: viewBox
    }, /*#__PURE__*/_react.default.createElement("circle", {
      className: _circleStyle.default.background,
      cx: radius,
      cy: radius,
      r: radius - strokeWidth,
      strokeWidth: "".concat(strokeWidth, "px")
    }), /*#__PURE__*/_react.default.createElement("circle", {
      className: _circleStyle.default.foreground,
      cx: radius,
      cy: radius,
      r: radius - strokeWidth,
      strokeWidth: "".concat(strokeWidth, "px"),
      style: {
        strokeDasharray: dashArray,
        strokeDashoffset: dashOffset
      }
    }), /*#__PURE__*/_react.default.createElement("circle", {
      className: _circleStyle.default.activeCircle,
      cx: radius,
      cy: radius,
      r: radius - 1,
      style: {
        strokeDasharray: Math.PI,
        strokeDashoffset: Math.PI
      }
    }), showPercent && /*#__PURE__*/_react.default.createElement("text", {
      className: _circleStyle.default.percent,
      x: radius,
      y: text ? radius - 5 : radius,
      dy: ".4em",
      textAnchor: "middle"
    }, "".concat(Math.round(percent), "%")), text && /*#__PURE__*/_react.default.createElement("text", {
      className: _circleStyle.default.status,
      x: radius,
      y: radius * radiusFactor - 5,
      dy: ".4em",
      textAnchor: "middle"
    }, text)));
  }

}

exports.default = ProgressCircle;
