// @sanity/default-layout/lib/addons/Sidecar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var sidecar = _interopRequireWildcard(require("part:@sanity/default-layout/sidecar?"));

var _sidecarDatastore = require("part:@sanity/default-layout/sidecar-datastore");

var _Sidecar = _interopRequireDefault(require("./Sidecar.css"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var isSidecarEnabled = null;
var SidecarLayout = null;

if (sidecar) {
  isSidecarEnabled = sidecar.isSidecarEnabled;
  SidecarLayout = sidecar.SidecarLayout;
}

// eslint-disable-next-line @typescript-eslint/ban-types
class Sidecar extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      isOpen: true,
      isVisible: true
    });

    _defineProperty(this, "subscription", null);

    _defineProperty(this, "handleRemoveSidecar", () => {
      this.setState({
        isVisible: false
      });
    });

    _defineProperty(this, "handleDismissSidecar", () => {
      var transitionDuration = 500; // from the isOpen class

      setTimeout(() => {
        this.handleRemoveSidecar();
      }, transitionDuration);
    });
  }

  componentDidMount() {
    if (isSidecarEnabled && isSidecarEnabled()) {
      this.subscription = _sidecarDatastore.isSidecarOpenSetting.listen().subscribe(isOpen => {
        this.setState({
          isOpen: isOpen !== false
        });
      });
    }

    if (!this.state.isOpen) {
      this.handleRemoveSidecar();
    }
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  componentDidUpdate() {
    // eslint-disable-next-line react/no-did-update-set-state
    return this.state.isOpen ? this.setState({
      isVisible: true
    }) : this.handleDismissSidecar();
  }

  render() {
    var _this$state = this.state,
        isOpen = _this$state.isOpen,
        isVisible = _this$state.isVisible;

    if (!isVisible || !(isSidecarEnabled && isSidecarEnabled())) {
      return null;
    }

    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "pane"
    }, /*#__PURE__*/_react.default.createElement(_ui.Layer, {
      className: (0, _classnames.default)(_Sidecar.default.root, isOpen && _Sidecar.default.isOpen)
    }, isVisible && /*#__PURE__*/_react.default.createElement(SidecarLayout, null)));
  }

}

var _default = Sidecar;
exports.default = _default;
