// @sanity/base/lib/__legacy/@sanity/components/panes/TabbedPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _splitHorizontalIcon = _interopRequireDefault(require("part:@sanity/base/split-horizontal-icon"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _tab = _interopRequireDefault(require("part:@sanity/components/tabs/tab"));

var _tabList = _interopRequireDefault(require("part:@sanity/components/tabs/tab-list"));

var _DefaultPane = _interopRequireDefault(require("./DefaultPane"));

var _TabbedPane = _interopRequireDefault(require("./TabbedPane.css"));

var _excluded = ["activeView", "idPrefix", "views", "onSetActiveView", "onSplitPane", "onCloseView"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class TabbedPane extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "renderHeaderViewMenu", () => {
      var _this$props = this.props,
          _this$props$views = _this$props.views,
          views = _this$props$views === void 0 ? [] : _this$props$views,
          onSplitPane = _this$props.onSplitPane,
          onCloseView = _this$props.onCloseView,
          isClosable = _this$props.isClosable;
      var isSplittable = Boolean(onSplitPane); // Do not render view menu when there are only 1 view

      if (views.length <= 1) return null;
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _TabbedPane.default.tabsContainer
      }, this.renderTabs(), /*#__PURE__*/_react.default.createElement("div", {
        className: _TabbedPane.default.headerPaneActions
      }, isSplittable && views.length > 1 && /*#__PURE__*/_react.default.createElement("button", {
        type: "button",
        onClick: onSplitPane,
        title: "Split pane right"
      }, /*#__PURE__*/_react.default.createElement("div", {
        tabIndex: -1
      }, /*#__PURE__*/_react.default.createElement(_splitHorizontalIcon.default, null))), isSplittable && isClosable && /*#__PURE__*/_react.default.createElement("button", {
        type: "button",
        onClick: onCloseView,
        title: "Close pane"
      }, /*#__PURE__*/_react.default.createElement("div", {
        tabIndex: -1
      }, /*#__PURE__*/_react.default.createElement(_closeIcon.default, null)))));
    });
  }

  renderTabs() {
    var _this$props2 = this.props,
        idPrefix = _this$props2.idPrefix,
        _this$props2$views = _this$props2.views,
        views = _this$props2$views === void 0 ? [] : _this$props2$views,
        activeView = _this$props2.activeView,
        onSetActiveView = _this$props2.onSetActiveView;

    if (views.length <= 1) {
      return null;
    }

    var tabPanelId = "".concat(idPrefix, "tabpanel");
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _TabbedPane.default.headerTabsContainer
    }, /*#__PURE__*/_react.default.createElement(_tabList.default, null, views.map((view, index) => /*#__PURE__*/_react.default.createElement(_tab.default, {
      id: "".concat(idPrefix, "tab-").concat(view.id),
      isActive: activeView === view.id,
      key: view.id,
      label: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, view.title),
      icon: view.icon,
      onClick: () => onSetActiveView(index === 0 ? null : view.id),
      "aria-controls": tabPanelId
    }))));
  }

  render() {
    var _this$props3 = this.props,
        activeView = _this$props3.activeView,
        idPrefix = _this$props3.idPrefix,
        _this$props3$views = _this$props3.views,
        views = _this$props3$views === void 0 ? [] : _this$props3$views,
        onSetActiveView = _this$props3.onSetActiveView,
        onSplitPane = _this$props3.onSplitPane,
        onCloseView = _this$props3.onCloseView,
        rest = _objectWithoutProperties(_this$props3, _excluded);

    var hasTabs = views.length > 1;
    return /*#__PURE__*/_react.default.createElement(_DefaultPane.default, _extends({
      hasTabs: hasTabs,
      index: 0,
      renderHeaderViewMenu: this.renderHeaderViewMenu,
      tabIdPrefix: idPrefix,
      viewId: activeView
    }, rest));
  }

}

var _default = TabbedPane;
exports.default = _default;
