// @sanity/base/lib/datastores/presence/message-transports/bifurTransport.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createBifurTransport = void 0;

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var handleIncomingMessage = event => {
  if (event.type === 'rollCall') {
    return {
      type: 'rollCall',
      userId: event.i,
      sessionId: event.session
    };
  }

  if (event.type === 'state') {
    var _event$m = event.m,
        sessionId = _event$m.sessionId,
        locations = _event$m.locations;
    return {
      type: 'state',
      userId: event.i,
      sessionId: sessionId,
      timestamp: new Date().toISOString(),
      locations
    };
  }

  if (event.type === 'disconnect') {
    return {
      type: 'disconnect',
      userId: event.i,
      sessionId: event.m.session,
      timestamp: new Date().toISOString()
    };
  }

  throw new Error("Got unknown presence event: ".concat(JSON.stringify(event)));
};

var createBifurTransport = (bifur, sessionId) => {
  var incomingEvents$ = bifur.request('presence').pipe((0, _operators.map)(handleIncomingMessage));

  var dispatchMessage = message => {
    if (message.type === 'rollCall') {
      return bifur.request('presence_rollcall', {
        session: sessionId
      });
    }

    if (message.type === 'state') {
      return bifur.request('presence_announce', {
        data: {
          locations: message.locations,
          sessionId
        }
      });
    }

    if (message.type === 'disconnect') {
      return bifur.request('presence_disconnect', {
        session: sessionId
      });
    }

    return _rxjs.EMPTY;
  };

  return [incomingEvents$.pipe((0, _operators.share)()), dispatchMessage];
};

exports.createBifurTransport = createBifurTransport;
