// @sanity/form-builder/lib/sanity/uploads/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createUploadEvent = createUploadEvent;
exports.createInitialUploadEvent = createInitialUploadEvent;
exports.CLEANUP_EVENT = void 0;

var _patches = require("../../patch/patches");

var _constants = require("./constants");

var UNSET_UPLOAD_PATCH = (0, _patches.unset)([_constants.UPLOAD_STATUS_KEY]);

function createUploadEvent() {
  var patches = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  return {
    type: 'uploadEvent',
    patches
  };
}

var CLEANUP_EVENT = createUploadEvent([UNSET_UPLOAD_PATCH]);
exports.CLEANUP_EVENT = CLEANUP_EVENT;

function createInitialUploadEvent(file) {
  var now = new Date().toISOString();
  var value = {
    progress: 2,
    initiated: now,
    updated: now,
    file: {
      name: file.name,
      type: file.type
    }
  };
  return createUploadEvent([(0, _patches.setIfMissing)({
    [_constants.UPLOAD_STATUS_KEY]: value
  }, []), (0, _patches.set)(value, [_constants.UPLOAD_STATUS_KEY])]);
}
