// @sanity/portable-text-editor/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  PortableTextEditor: true,
  keyGenerator: true,
  compactPatches: true,
  getPortableTextFeatures: true,
  PortableTextEditable: true,
  usePortableTextEditor: true,
  usePortableTextEditorSelection: true
};
Object.defineProperty(exports, "PortableTextEditor", {
  enumerable: true,
  get: function get() {
    return _PortableTextEditor.PortableTextEditor;
  }
});
Object.defineProperty(exports, "keyGenerator", {
  enumerable: true,
  get: function get() {
    return _PortableTextEditor.defaultKeyGenerator;
  }
});
Object.defineProperty(exports, "compactPatches", {
  enumerable: true,
  get: function get() {
    return _patches.compactPatches;
  }
});
Object.defineProperty(exports, "getPortableTextFeatures", {
  enumerable: true,
  get: function get() {
    return _getPortableTextFeatures.getPortableTextFeatures;
  }
});
Object.defineProperty(exports, "PortableTextEditable", {
  enumerable: true,
  get: function get() {
    return _Editable.PortableTextEditable;
  }
});
Object.defineProperty(exports, "usePortableTextEditor", {
  enumerable: true,
  get: function get() {
    return _usePortableTextEditor.usePortableTextEditor;
  }
});
Object.defineProperty(exports, "usePortableTextEditorSelection", {
  enumerable: true,
  get: function get() {
    return _usePortableTextEditorSelection.usePortableTextEditorSelection;
  }
});

var _PortableTextEditor = require("./editor/PortableTextEditor");

var _editor = require("./types/editor");

Object.keys(_editor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _editor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _editor[key];
    }
  });
});

var _portableText = require("./types/portableText");

Object.keys(_portableText).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _portableText[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _portableText[key];
    }
  });
});

var _schema = require("./types/schema");

Object.keys(_schema).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _schema[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _schema[key];
    }
  });
});

var _options = require("./types/options");

Object.keys(_options).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _options[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _options[key];
    }
  });
});

var _patches = require("./utils/patches");

var _getPortableTextFeatures = require("./utils/getPortableTextFeatures");

var _Editable = require("./editor/Editable");

var _usePortableTextEditor = require("./editor/hooks/usePortableTextEditor");

var _usePortableTextEditorSelection = require("./editor/hooks/usePortableTextEditorSelection");
